// n8n-workflow/dist/cjs/versioned-node-type.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.VersionedNodeType = void 0;
    class VersionedNodeType {
        currentVersion;
        nodeVersions;
        description;
        constructor(nodeVersions, description) {
            this.nodeVersions = nodeVersions;
            this.currentVersion = description.defaultVersion ?? this.getLatestVersion();
            this.description = description;
        }
        getLatestVersion() {
            return Math.max(...Object.keys(this.nodeVersions).map(Number));
        }
        getNodeType(version) {
            if (version) {
                return this.nodeVersions[version];
            }
            else {
                return this.nodeVersions[this.currentVersion];
            }
        }
    }
    exports.VersionedNodeType = VersionedNodeType;
});
