// n8n-core/dist/execution-engine/node-execution-context/utils/request-helpers/axios-utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFormDataObject = exports.getBeforeRedirectFn = exports.getHostFromRequestObject = void 0;
exports.throwIfDomainNotAllowed = throwIfDomainNotAllowed;
exports.tryParseUrl = tryParseUrl;
exports.isIgnoreStatusErrorConfig = isIgnoreStatusErrorConfig;
exports.searchForHeader = searchForHeader;
exports.digestAuthAxiosConfig = digestAuthAxiosConfig;
exports.generateContentLengthHeader = generateContentLengthHeader;
exports.getUrlFromProxyConfig = getUrlFromProxyConfig;
exports.buildTargetUrl = buildTargetUrl;
exports.setAxiosAgents = setAxiosAgents;
const backend_common_1 = require("@n8n/backend-common");
const di_1 = require("@n8n/di");
const axios_1 = __importDefault(require("axios"));
const crypto_1 = __importDefault(require("crypto"));
const form_data_1 = __importDefault(require("form-data"));
const n8n_workflow_1 = require("n8n-workflow");
const http_proxy_1 = require("../../../../http-proxy");
function throwIfDomainNotAllowed(configOrUrl, allowedDomains) {
    const url = typeof configOrUrl === 'string' ? configOrUrl : axios_1.default.getUri(configOrUrl);
    if (allowedDomains && !(0, n8n_workflow_1.isDomainAllowed)(url, { allowedDomains })) {
        throw new n8n_workflow_1.ApplicationError(`Domain not allowed: This credential is restricted from accessing ${url}. Only the following domains are allowed: ${allowedDomains}`);
    }
}
function tryParseUrl(url) {
    try {
        return new URL(url);
    }
    catch {
        return null;
    }
}
function isIgnoreStatusErrorConfig(ignoreHttpStatusErrors) {
    return (typeof ignoreHttpStatusErrors === 'object' &&
        ignoreHttpStatusErrors !== null &&
        'ignore' in ignoreHttpStatusErrors &&
        ignoreHttpStatusErrors.ignore === true);
}
function searchForHeader(config, headerName) {
    if (config.headers === undefined) {
        return undefined;
    }
    const headerNames = Object.keys(config.headers);
    headerName = headerName.toLowerCase();
    return headerNames.find((thisHeader) => thisHeader.toLowerCase() === headerName);
}
const getHostFromRequestObject = (requestObject) => {
    try {
        const url = (requestObject.url ?? requestObject.uri);
        return new URL(url, requestObject.baseURL).hostname;
    }
    catch (error) {
        return null;
    }
};
exports.getHostFromRequestObject = getHostFromRequestObject;
const getBeforeRedirectFn = (agentOptions, axiosConfig, proxyConfig, sendCredentialsOnCrossOriginRedirect, allowedDomains, ssrfBridge) => (redirectedRequest) => {
    throwIfDomainNotAllowed(redirectedRequest.href, allowedDomains);
    if (ssrfBridge) {
        ssrfBridge.validateRedirectSync(redirectedRequest.href);
    }
    const redirectAgentOptions = {
        ...agentOptions,
        servername: redirectedRequest.hostname,
    };
    const customProxyUrl = proxyConfig ? getUrlFromProxyConfig(proxyConfig) : null;
    const effectiveRedirectOptions = ssrfBridge && !customProxyUrl
        ? { ...redirectAgentOptions, lookup: ssrfBridge.createSecureLookup() }
        : redirectAgentOptions;
    const targetUrl = redirectedRequest.href;
    const httpAgent = (0, http_proxy_1.createHttpProxyAgent)(customProxyUrl, targetUrl, effectiveRedirectOptions);
    const httpsAgent = (0, http_proxy_1.createHttpsProxyAgent)(customProxyUrl, targetUrl, effectiveRedirectOptions);
    redirectedRequest.agent = redirectedRequest.href.startsWith('https://')
        ? httpsAgent
        : httpAgent;
    redirectedRequest.agents = { http: httpAgent, https: httpsAgent };
    const originalUrl = axiosConfig.baseURL
        ? new URL(axiosConfig.url ?? '', axiosConfig.baseURL)
        : new URL(axiosConfig.url ?? '');
    const originalOrigin = originalUrl.origin;
    const targetOrigin = new URL(targetUrl).origin;
    if (originalOrigin === targetOrigin || sendCredentialsOnCrossOriginRedirect) {
        if (axiosConfig.headers?.Authorization) {
            redirectedRequest.headers.Authorization = axiosConfig.headers.Authorization;
        }
        if (axiosConfig.auth) {
            redirectedRequest.auth = `${axiosConfig.auth.username}:${axiosConfig.auth.password}`;
        }
    }
};
exports.getBeforeRedirectFn = getBeforeRedirectFn;
function digestAuthAxiosConfig(axiosConfig, response, auth) {
    const authDetails = response.headers['www-authenticate']
        .split(',')
        .map((v) => v.split('='));
    if (authDetails) {
        const nonceCount = '000000001';
        const cnonce = crypto_1.default.randomBytes(24).toString('hex');
        const realm = authDetails
            .find((el) => el[0].toLowerCase().indexOf('realm') > -1)[1]
            .replace(/"/g, '');
        const opaqueKV = authDetails.find((el) => el[0].toLowerCase().indexOf('opaque') > -1);
        const opaque = opaqueKV ? opaqueKV[1].replace(/"/g, '') : undefined;
        const nonce = authDetails
            .find((el) => el[0].toLowerCase().indexOf('nonce') > -1)[1]
            .replace(/"/g, '');
        const ha1 = crypto_1.default
            .createHash('md5')
            .update(`${auth?.username}:${realm}:${auth?.password}`)
            .digest('hex');
        const url = new URL(axios_1.default.getUri(axiosConfig));
        const path = url.pathname + url.search;
        const ha2 = crypto_1.default
            .createHash('md5')
            .update(`${axiosConfig.method ?? 'GET'}:${path}`)
            .digest('hex');
        const response = crypto_1.default
            .createHash('md5')
            .update(`${ha1}:${nonce}:${nonceCount}:${cnonce}:auth:${ha2}`)
            .digest('hex');
        let authorization = `Digest username="${auth?.username}",realm="${realm}",` +
            `nonce="${nonce}",uri="${path}",qop="auth",algorithm="MD5",` +
            `response="${response}",nc="${nonceCount}",cnonce="${cnonce}"`;
        if (opaque) {
            authorization += `,opaque="${opaque}"`;
        }
        if (axiosConfig.headers) {
            axiosConfig.headers.authorization = authorization;
        }
        else {
            axiosConfig.headers = { authorization };
        }
    }
    return axiosConfig;
}
const pushFormDataValue = (form, key, value) => {
    if (value?.hasOwnProperty('value') && value.hasOwnProperty('options')) {
        form.append(key, value.value, value.options);
    }
    else {
        form.append(key, value);
    }
};
const createFormDataObject = (data) => {
    const formData = new form_data_1.default();
    const keys = Object.keys(data);
    keys.forEach((key) => {
        const formField = data[key];
        if (formField instanceof Array) {
            formField.forEach((item) => {
                pushFormDataValue(formData, key, item);
            });
        }
        else {
            pushFormDataValue(formData, key, formField);
        }
    });
    return formData;
};
exports.createFormDataObject = createFormDataObject;
async function generateContentLengthHeader(config) {
    if (!(config.data instanceof form_data_1.default)) {
        return;
    }
    try {
        const length = await new Promise((res, rej) => {
            config.data.getLength((error, dataLength) => {
                if (error)
                    rej(error);
                else
                    res(dataLength);
            });
        });
        config.headers = {
            ...config.headers,
            'content-length': length,
        };
    }
    catch (error) {
        di_1.Container.get(backend_common_1.Logger).error('Unable to calculate form data length', { error });
    }
}
function getUrlFromProxyConfig(proxyConfig) {
    if (typeof proxyConfig === 'string') {
        const isValidUrl = !!tryParseUrl(proxyConfig);
        return isValidUrl ? proxyConfig : null;
    }
    if (!proxyConfig?.host)
        return null;
    const { protocol, host, port, auth } = proxyConfig;
    const safeProtocol = protocol?.endsWith(':') ? protocol.slice(0, -1) : (protocol ?? 'http');
    try {
        const url = new URL(`${safeProtocol}://${host}`);
        if (port !== undefined)
            url.port = String(port);
        if (auth?.username) {
            url.username = auth.username;
            url.password = auth.password ?? '';
        }
        return url.href;
    }
    catch {
        return null;
    }
}
function buildTargetUrl(url, baseURL) {
    if (!url)
        return undefined;
    try {
        return baseURL ? new URL(url, baseURL).href : url;
    }
    catch {
        return undefined;
    }
}
function setAxiosAgents(config, agentOptions, proxyConfig, secureLookup) {
    if (config.httpAgent || config.httpsAgent)
        return;
    const customProxyUrl = getUrlFromProxyConfig(proxyConfig);
    const targetUrl = buildTargetUrl(config.url, config.baseURL);
    if (!targetUrl)
        return;
    const effectiveOptions = secureLookup && !customProxyUrl ? { ...agentOptions, lookup: secureLookup } : agentOptions;
    config.httpAgent = (0, http_proxy_1.createHttpProxyAgent)(customProxyUrl, targetUrl, effectiveOptions);
    config.httpsAgent = (0, http_proxy_1.createHttpsProxyAgent)(customProxyUrl, targetUrl, effectiveOptions);
}
