// @n8n/workflow-sdk/dist/ast-interpreter/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnknownIdentifierError = exports.SecurityError = exports.UnsupportedNodeError = exports.InterpreterError = void 0;
class InterpreterError extends Error {
    constructor(message, location, sourceCode) {
        super(message);
        this.name = 'InterpreterError';
        this.location = location;
        this.sourceCode = sourceCode;
        if (location && sourceCode) {
            this.message = `${message}\n\n${this.generateCodeFrame()}`;
        }
    }
    generateCodeFrame() {
        if (!this.location || !this.sourceCode)
            return '';
        const lines = this.sourceCode.split('\n');
        const { line, column } = this.location.start;
        const lineIndex = line - 1;
        if (lineIndex < 0 || lineIndex >= lines.length)
            return '';
        const contextLines = 2;
        const startLine = Math.max(0, lineIndex - contextLines);
        const endLine = Math.min(lines.length - 1, lineIndex + contextLines);
        const frameLines = [];
        const gutterWidth = String(endLine + 1).length;
        for (let i = startLine; i <= endLine; i++) {
            const lineNum = String(i + 1).padStart(gutterWidth, ' ');
            const prefix = i === lineIndex ? '> ' : '  ';
            frameLines.push(`${prefix}${lineNum} | ${lines[i]}`);
            if (i === lineIndex) {
                const indicator = ' '.repeat(column) + '^';
                frameLines.push(`  ${' '.repeat(gutterWidth)} | ${indicator}`);
            }
        }
        return frameLines.join('\n');
    }
}
exports.InterpreterError = InterpreterError;
class UnsupportedNodeError extends InterpreterError {
    constructor(nodeType, location, sourceCode) {
        super(`Unsupported syntax: '${nodeType}' is not allowed in SDK code`, location, sourceCode);
        this.name = 'UnsupportedNodeError';
        this.nodeType = nodeType;
    }
}
exports.UnsupportedNodeError = UnsupportedNodeError;
class SecurityError extends InterpreterError {
    constructor(pattern, location, sourceCode) {
        super(`Security violation: '${pattern}' is not allowed`, location, sourceCode);
        this.name = 'SecurityError';
        this.pattern = pattern;
    }
}
exports.SecurityError = SecurityError;
class UnknownIdentifierError extends InterpreterError {
    constructor(identifier, location, sourceCode) {
        super(`Unknown identifier: '${identifier}' is not defined`, location, sourceCode);
        this.name = 'UnknownIdentifierError';
        this.identifier = identifier;
    }
}
exports.UnknownIdentifierError = UnknownIdentifierError;
