// @n8n/workflow-sdk/dist/ast-interpreter/validators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ALLOWED_METHODS = exports.AUTO_RENAMEABLE_SDK_FUNCTIONS = exports.ALLOWED_SDK_FUNCTIONS = void 0;
exports.isAutoRenameableSDKFunction = isAutoRenameableSDKFunction;
exports.validateNodeType = validateNodeType;
exports.validateIdentifier = validateIdentifier;
exports.validateCallExpression = validateCallExpression;
exports.validateMemberExpression = validateMemberExpression;
exports.isAllowedSDKFunction = isAllowedSDKFunction;
exports.isAllowedMethod = isAllowedMethod;
exports.getSafeJSONMethod = getSafeJSONMethod;
exports.getSafeStringMethod = getSafeStringMethod;
const errors_1 = require("./errors");
exports.ALLOWED_SDK_FUNCTIONS = new Set([
    'workflow',
    'node',
    'trigger',
    'sticky',
    'placeholder',
    'newCredential',
    'ifElse',
    'switchCase',
    'merge',
    'splitInBatches',
    'nextBatch',
    'languageModel',
    'memory',
    'tool',
    'outputParser',
    'embedding',
    'embeddings',
    'vectorStore',
    'retriever',
    'documentLoader',
    'textSplitter',
    'reranker',
    'fromAi',
]);
exports.AUTO_RENAMEABLE_SDK_FUNCTIONS = new Set([
    'languageModel',
    'memory',
    'tool',
    'outputParser',
    'embedding',
    'embeddings',
    'vectorStore',
    'retriever',
    'documentLoader',
    'textSplitter',
    'reranker',
]);
function isAutoRenameableSDKFunction(name) {
    return exports.AUTO_RENAMEABLE_SDK_FUNCTIONS.has(name);
}
exports.ALLOWED_METHODS = new Set([
    'add',
    'to',
    'input',
    'output',
    'onError',
    'onTrue',
    'onFalse',
    'onCase',
    'onEachBatch',
    'onDone',
    'connect',
    'toJSON',
    'validate',
]);
const DANGEROUS_GLOBALS = new Set([
    'eval',
    'Function',
    'require',
    'import',
    'process',
    'global',
    'globalThis',
    'window',
    'document',
    'setTimeout',
    'setInterval',
    'setImmediate',
    'clearTimeout',
    'clearInterval',
    'clearImmediate',
    '__dirname',
    '__filename',
    'module',
    'exports',
    'Buffer',
    'Reflect',
    'Proxy',
    'Object',
    'Array',
    'String',
    'Number',
    'Boolean',
    'Symbol',
    'BigInt',
    'JSON',
    'Math',
    'Date',
    'RegExp',
    'Promise',
    'WeakRef',
    'WeakMap',
    'WeakSet',
    'Map',
    'Set',
    'ArrayBuffer',
    'SharedArrayBuffer',
    'DataView',
    'Atomics',
    'WebAssembly',
    'fetch',
    'XMLHttpRequest',
    'queueMicrotask',
    'structuredClone',
    'Error',
    'TypeError',
    'RangeError',
    'console',
]);
const ALLOWED_NODE_TYPES = new Set([
    'Program',
    'VariableDeclaration',
    'VariableDeclarator',
    'ExpressionStatement',
    'ExportDefaultDeclaration',
    'CallExpression',
    'MemberExpression',
    'ObjectExpression',
    'ArrayExpression',
    'Identifier',
    'Literal',
    'TemplateLiteral',
    'TemplateElement',
    'SpreadElement',
    'Property',
    'AssignmentExpression',
    'UnaryExpression',
    'BinaryExpression',
    'LogicalExpression',
    'ConditionalExpression',
]);
const FORBIDDEN_NODE_TYPES = {
    ArrowFunctionExpression: 'Arrow functions are not allowed. Use fromAi() directly instead of ($) => $.fromAi()',
    FunctionExpression: 'Function expressions are not allowed in SDK code',
    FunctionDeclaration: 'Function declarations are not allowed in SDK code',
    ClassDeclaration: 'Class declarations are not allowed in SDK code',
    ClassExpression: 'Class expressions are not allowed in SDK code',
    ForStatement: 'For loops are not allowed in SDK code',
    ForInStatement: 'For-in loops are not allowed in SDK code',
    ForOfStatement: 'For-of loops are not allowed in SDK code',
    WhileStatement: 'While loops are not allowed in SDK code',
    DoWhileStatement: 'Do-while loops are not allowed in SDK code',
    TryStatement: 'Try-catch is not allowed in SDK code',
    ThrowStatement: 'Throw statements are not allowed in SDK code',
    WithStatement: 'With statements are not allowed in SDK code',
    UpdateExpression: 'Update expressions (++, --) are not allowed in SDK code',
    NewExpression: 'new expressions are not allowed. Use SDK factory functions instead.',
    ImportDeclaration: 'Import declarations are not allowed in SDK code',
    ImportExpression: 'Dynamic imports are not allowed in SDK code',
    ExportNamedDeclaration: 'Named exports are not allowed. Use export default only.',
    ExportAllDeclaration: 'Re-exports are not allowed in SDK code',
    AwaitExpression: 'Await expressions are not allowed in SDK code',
    YieldExpression: 'Yield expressions are not allowed in SDK code',
};
function validateNodeType(node, sourceCode) {
    if (FORBIDDEN_NODE_TYPES[node.type]) {
        throw new errors_1.UnsupportedNodeError(`${node.type}: ${FORBIDDEN_NODE_TYPES[node.type]}`, node.loc ?? undefined, sourceCode);
    }
    if (!ALLOWED_NODE_TYPES.has(node.type)) {
        throw new errors_1.UnsupportedNodeError(node.type, node.loc ?? undefined, sourceCode);
    }
}
function validateIdentifier(name, _allowedVariables, node, sourceCode) {
    if (DANGEROUS_GLOBALS.has(name)) {
        throw new errors_1.SecurityError(`Access to '${name}' is not allowed`, node.loc ?? undefined, sourceCode);
    }
}
function validateCallExpression(node, sourceCode) {
    if (node.callee.type === 'Identifier') {
        const name = node.callee.name;
        if (name === 'eval') {
            throw new errors_1.SecurityError('eval()', node.loc ?? undefined, sourceCode);
        }
        if (name === 'Function') {
            throw new errors_1.SecurityError('Function()', node.loc ?? undefined, sourceCode);
        }
        if (name === 'require') {
            throw new errors_1.SecurityError('require()', node.loc ?? undefined, sourceCode);
        }
    }
    if (node.callee.type === 'MemberExpression') {
        const memberExpr = node.callee;
        if (memberExpr.property.type === 'Identifier' && memberExpr.property.name === 'constructor') {
            throw new errors_1.SecurityError('constructor access', node.loc ?? undefined, sourceCode);
        }
    }
}
function validateMemberExpression(node, sourceCode) {
    if (node.computed) {
        if (node.property.type !== 'Literal') {
            throw new errors_1.SecurityError('Dynamic property access is not allowed. Use static property names.', node.loc ?? undefined, sourceCode);
        }
    }
    const propName = node.property.type === 'Identifier'
        ? node.property.name
        : node.property.type === 'Literal' && typeof node.property.value === 'string'
            ? node.property.value
            : undefined;
    if (propName !== undefined &&
        (propName === '__proto__' || propName === 'prototype' || propName === 'constructor')) {
        throw new errors_1.SecurityError(`Access to '${propName}' is not allowed`, node.loc ?? undefined, sourceCode);
    }
}
function isAllowedSDKFunction(name) {
    return exports.ALLOWED_SDK_FUNCTIONS.has(name);
}
function isAllowedMethod(name) {
    return exports.ALLOWED_METHODS.has(name);
}
const SAFE_JSON_METHODS = {
    stringify: (...args) => JSON.stringify(args[0], args[1], args[2]),
};
function getSafeJSONMethod(objectName, methodName) {
    if (objectName !== 'JSON')
        return undefined;
    return SAFE_JSON_METHODS[methodName];
}
const SAFE_STRING_METHODS = {
    repeat: (str, count) => str.repeat(count),
};
function getSafeStringMethod(value, methodName) {
    if (typeof value !== 'string')
        return undefined;
    const factory = SAFE_STRING_METHODS[methodName];
    if (!factory)
        return undefined;
    return (...args) => factory(value, ...args);
}
