// @n8n/expression-runtime/dist/cjs/runtime/reset.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "luxon", "../extensions/extend", "../extensions/function-extensions", "./safe-globals", "./lazy-proxy"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.resetDataProxies = resetDataProxies;
    const luxon_1 = require("luxon");
    const extend_1 = require("../extensions/extend");
    const function_extensions_1 = require("../extensions/function-extensions");
    const safe_globals_1 = require("./safe-globals");
    const lazy_proxy_1 = require("./lazy-proxy");
    // Pre-create safe error subclass wrappers (reused across evaluations)
    const SafeTypeError = (0, safe_globals_1.createSafeErrorSubclass)(TypeError);
    const SafeSyntaxError = (0, safe_globals_1.createSafeErrorSubclass)(SyntaxError);
    const SafeEvalError = (0, safe_globals_1.createSafeErrorSubclass)(EvalError);
    const SafeRangeError = (0, safe_globals_1.createSafeErrorSubclass)(RangeError);
    const SafeReferenceError = (0, safe_globals_1.createSafeErrorSubclass)(ReferenceError);
    const SafeURIError = (0, safe_globals_1.createSafeErrorSubclass)(URIError);
    // ============================================================================
    // Reset Function for Data Proxies
    // ============================================================================
    function fetchPrimitive(key) {
        try {
            return globalThis.__getValueAtPath.applySync(null, [[key]], {
                arguments: { copy: true },
                result: { copy: true },
            });
        }
        catch {
            return undefined;
        }
    }
    /**
     * Reset workflow data proxies before each evaluation.
     *
     * This function is called from the bridge before executing each expression
     * to clear proxy caches and initialize fresh workflow data references.
     *
     * Pattern:
     * 1. Create lazy proxies for complex properties ($json, $binary, etc.)
     * 2. Fetch primitives directly ($runIndex, $itemIndex)
     * 3. Create function wrappers for callable properties ($items, etc.)
     * 4. Expose all properties to globalThis for expression access
     *
     * Called from bridge: context.evalSync('resetDataProxies()')
     */
    function resetDataProxies(timezone) {
        if (timezone && !luxon_1.IANAZone.isValidZone(timezone)) {
            throw new Error(`Invalid timezone: "${timezone}"`);
        }
        luxon_1.Settings.defaultZone = timezone ?? 'system';
        // Clear existing __data object
        globalThis.__data = {};
        // __sanitize must be on __data because PrototypeSanitizer generates:
        // obj[this.__sanitize(expr)] where 'this' is __data (via .call(__data) wrapping)
        // Use a non-writable property descriptor so override attempts throw instead of silently succeeding.
        Object.defineProperty(globalThis.__data, '__sanitize', {
            get: () => safe_globals_1.__sanitize,
            set: () => {
                throw new safe_globals_1.ExpressionError('Cannot override "__sanitize" due to security concerns');
            },
            enumerable: false,
            configurable: false,
        });
        // Verify callbacks are available
        // Note: ivm.Reference may not be typeof 'function', check for existence
        if (!globalThis.__getValueAtPath) {
            throw new Error('__getValueAtPath callback not registered');
        }
        // -------------------------------------------------------------------------
        // Create lazy proxies for complex workflow properties
        // -------------------------------------------------------------------------
        globalThis.__data.$json = (0, lazy_proxy_1.createDeepLazyProxy)(['$json']);
        globalThis.__data.$binary = (0, lazy_proxy_1.createDeepLazyProxy)(['$binary']);
        globalThis.__data.$input = (0, lazy_proxy_1.createDeepLazyProxy)(['$input']);
        globalThis.__data.$node = (0, lazy_proxy_1.createDeepLazyProxy)(['$node']);
        globalThis.__data.$parameter = (0, lazy_proxy_1.createDeepLazyProxy)(['$parameter']);
        globalThis.__data.$workflow = (0, lazy_proxy_1.createDeepLazyProxy)(['$workflow']);
        globalThis.__data.$prevNode = (0, lazy_proxy_1.createDeepLazyProxy)(['$prevNode']);
        globalThis.__data.$data = (0, lazy_proxy_1.createDeepLazyProxy)(['$data']);
        globalThis.__data.$env = (0, lazy_proxy_1.createDeepLazyProxy)(['$env']);
        globalThis.__data.process = (0, lazy_proxy_1.createDeepLazyProxy)(['process']);
        globalThis.__data.$execution = (0, lazy_proxy_1.createDeepLazyProxy)(['$execution']);
        globalThis.__data.$vars = (0, lazy_proxy_1.createDeepLazyProxy)(['$vars']);
        globalThis.__data.$secrets = (0, lazy_proxy_1.createDeepLazyProxy)(['$secrets']);
        // -------------------------------------------------------------------------
        // Create DateTime values inside the isolate (not lazy-loaded from host,
        // because host-side DateTime objects lose their prototype crossing the
        // boundary). The isolate has its own luxon with the correct timezone
        // already set via Settings.defaultZone above.
        // -------------------------------------------------------------------------
        globalThis.__data.$now = luxon_1.DateTime.now();
        globalThis.__data.$today = luxon_1.DateTime.now().set({ hour: 0, minute: 0, second: 0, millisecond: 0 });
        // -------------------------------------------------------------------------
        // Fetch primitives directly (no lazy loading needed for simple values)
        // -------------------------------------------------------------------------
        globalThis.__data.$runIndex = fetchPrimitive('$runIndex');
        globalThis.__data.$itemIndex = fetchPrimitive('$itemIndex');
        globalThis.__data.$executionId = fetchPrimitive('$executionId');
        globalThis.__data.$resumeWebhookUrl = fetchPrimitive('$resumeWebhookUrl');
        globalThis.__data.$webhookId = fetchPrimitive('$webhookId');
        globalThis.__data.$nodeId = fetchPrimitive('$nodeId');
        globalThis.__data.$nodeVersion = fetchPrimitive('$nodeVersion');
        // -------------------------------------------------------------------------
        // Expose workflow data to globalThis for expression access
        // -------------------------------------------------------------------------
        globalThis.$json = globalThis.__data.$json;
        globalThis.$binary = globalThis.__data.$binary;
        globalThis.$input = globalThis.__data.$input;
        globalThis.$node = globalThis.__data.$node;
        globalThis.$parameter = globalThis.__data.$parameter;
        globalThis.$workflow = globalThis.__data.$workflow;
        globalThis.$prevNode = globalThis.__data.$prevNode;
        globalThis.$runIndex = globalThis.__data.$runIndex;
        globalThis.$itemIndex = globalThis.__data.$itemIndex;
        globalThis.$data = globalThis.__data.$data;
        globalThis.$env = globalThis.__data.$env;
        globalThis.$now = globalThis.__data.$now;
        globalThis.$today = globalThis.__data.$today;
        globalThis.$execution = globalThis.__data.$execution;
        globalThis.$vars = globalThis.__data.$vars;
        globalThis.$secrets = globalThis.__data.$secrets;
        globalThis.$executionId = globalThis.__data.$executionId;
        globalThis.$resumeWebhookUrl = globalThis.__data.$resumeWebhookUrl;
        globalThis.$webhookId = globalThis.__data.$webhookId;
        globalThis.$nodeId = globalThis.__data.$nodeId;
        globalThis.$nodeVersion = globalThis.__data.$nodeVersion;
        // Expose standalone functions (min, max, average, numberList, zip, $ifEmpty, etc.)
        Object.assign(globalThis.__data, function_extensions_1.extendedFunctions);
        // -------------------------------------------------------------------------
        // Handle function properties (check if value is function metadata)
        // -------------------------------------------------------------------------
        // Check if $items exists and is a function
        if (globalThis.__callFunctionAtPath) {
            try {
                const itemsValue = globalThis.__getValueAtPath.applySync(null, [['$items']], {
                    arguments: { copy: true },
                    result: { copy: true },
                });
                // If it's function metadata, create wrapper
                if (itemsValue && typeof itemsValue === 'object' && itemsValue.__isFunction) {
                    globalThis.$items = function (...args) {
                        const result = globalThis.__callFunctionAtPath.applySync(null, [['$items'], ...args], {
                            arguments: { copy: true },
                            result: { copy: true },
                        });
                        (0, lazy_proxy_1.throwIfErrorSentinel)(result);
                        return result;
                    };
                    globalThis.__data.$items = globalThis.$items;
                }
                else {
                    // Not a function - set to undefined or the value itself
                    globalThis.$items = itemsValue;
                    globalThis.__data.$items = itemsValue;
                }
            }
            catch (error) {
                // Property doesn't exist
                globalThis.$items = undefined;
                globalThis.__data.$items = undefined;
            }
        }
        // -------------------------------------------------------------------------
        // Expose globals on __data so tournament's "x in this ? this.x : global.x"
        // pattern resolves them correctly (tournament checks __data before global)
        // -------------------------------------------------------------------------
        globalThis.__data.DateTime = globalThis.DateTime;
        globalThis.__data.Duration = globalThis.Duration;
        globalThis.__data.Interval = globalThis.Interval;
        // Expose extend/extendOptional on __data so tournament's "x in this ? this.x : global.x"
        // pattern resolves them correctly when the VM checks __data first
        globalThis.__data.extend = extend_1.extend;
        globalThis.__data.extendOptional = extend_1.extendOptional;
        // Wire builtins so tournament's VariablePolyfill resolves them from __data
        initializeBuiltins(globalThis.__data);
        // $item(itemIndex) returns a sub-proxy for the specified item (legacy syntax)
        globalThis.__data.$item = function (itemIndex) {
            const indexStr = String(itemIndex);
            return {
                $json: (0, lazy_proxy_1.createDeepLazyProxy)(['$item', indexStr, '$json']),
                $binary: (0, lazy_proxy_1.createDeepLazyProxy)(['$item', indexStr, '$binary']),
            };
        };
        globalThis.$ = function (nodeName) {
            return (0, lazy_proxy_1.createDeepLazyProxy)(['$', nodeName]);
        };
        globalThis.__data.$ = globalThis.$;
    }
    // Matches initializeGlobalContext() lines 262-318 in packages/workflow/src/expression.ts
    const DENYLISTED_GLOBALS = [
        'document',
        'global',
        'window',
        'Window',
        'globalThis',
        'self',
        'alert',
        'prompt',
        'confirm',
        'eval',
        'uneval',
        'setTimeout',
        'setInterval',
        'setImmediate',
        'clearImmediate',
        'queueMicrotask',
        'Function',
        'require',
        'module',
        'Buffer',
        '__dirname',
        '__filename',
        'fetch',
        'XMLHttpRequest',
        'Promise',
        'Generator',
        'GeneratorFunction',
        'AsyncFunction',
        'AsyncGenerator',
        'AsyncGeneratorFunction',
        'WebAssembly',
        'Reflect',
        'Proxy',
        'escape',
        'unescape',
    ];
    /**
     * Wire builtins onto __data so tournament's VariablePolyfill resolves them.
     *
     * Tournament transforms `Object` → `("Object" in this ? this : window).Object`.
     * `this` = __data. Without these entries on __data, builtins fall through to
     * `window` which doesn't exist in the isolate, causing a ReferenceError.
     *
     * Mirrors Expression.initializeGlobalContext() in packages/workflow/src/expression.ts.
     */
    function initializeBuiltins(data) {
        // ── Denylist: dangerous globals → empty objects ──
        for (const key of DENYLISTED_GLOBALS) {
            data[key] = {};
        }
        data.__lookupGetter__ = undefined;
        data.__lookupSetter__ = undefined;
        data.__defineGetter__ = undefined;
        data.__defineSetter__ = undefined;
        // ── Allowlist: safe versions of builtins ──
        // Object — use SafeObject wrapper from the runtime bundle
        data.Object = globalThis.SafeObject;
        // Error types — use SafeError and safe subclass wrappers
        data.Error = globalThis.SafeError;
        data.TypeError = SafeTypeError;
        data.SyntaxError = SafeSyntaxError;
        data.EvalError = SafeEvalError;
        data.RangeError = SafeRangeError;
        data.ReferenceError = SafeReferenceError;
        data.URIError = SafeURIError;
        // Arrays
        data.Array = Array;
        data.Int8Array = Int8Array;
        data.Uint8Array = Uint8Array;
        data.Uint8ClampedArray = Uint8ClampedArray;
        data.Int16Array = Int16Array;
        data.Uint16Array = Uint16Array;
        data.Int32Array = Int32Array;
        data.Uint32Array = Uint32Array;
        data.Float32Array = Float32Array;
        data.Float64Array = Float64Array;
        data.BigInt64Array = typeof BigInt64Array !== 'undefined' ? BigInt64Array : {};
        data.BigUint64Array = typeof BigUint64Array !== 'undefined' ? BigUint64Array : {};
        // Collections
        data.Map = Map;
        data.WeakMap = WeakMap;
        data.Set = Set;
        data.WeakSet = WeakSet;
        // Internationalization
        data.Intl = typeof Intl !== 'undefined' ? Intl : {};
        // Text
        data.String = String;
        // Numbers
        data.Number = Number;
        data.BigInt = typeof BigInt !== 'undefined' ? BigInt : {};
        data.Infinity = Infinity;
        data.parseFloat = parseFloat;
        data.parseInt = parseInt;
        // Structured data
        data.JSON = JSON;
        data.ArrayBuffer = typeof ArrayBuffer !== 'undefined' ? ArrayBuffer : {};
        data.SharedArrayBuffer = typeof SharedArrayBuffer !== 'undefined' ? SharedArrayBuffer : {};
        data.Atomics = typeof Atomics !== 'undefined' ? Atomics : {};
        data.DataView = typeof DataView !== 'undefined' ? DataView : {};
        // Encoding
        data.encodeURI = encodeURI;
        data.encodeURIComponent = encodeURIComponent;
        data.decodeURI = decodeURI;
        data.decodeURIComponent = decodeURIComponent;
        // Other
        data.Boolean = Boolean;
        data.Symbol = Symbol;
    }
});
