// @n8n/workflow-sdk/dist/workflow-builder/node-builders/node-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isInputTarget = isInputTarget;
exports.isOutputSelector = isOutputSelector;
exports.isIfElseBuilder = isIfElseBuilder;
exports.isSwitchCaseBuilder = isSwitchCaseBuilder;
exports.node = node;
exports.ifElse = ifElse;
exports.merge = merge;
exports.switchCase = switchCase;
exports.trigger = trigger;
exports.sticky = sticky;
exports.placeholder = placeholder;
exports.newCredential = newCredential;
exports.cloneNodeWithId = cloneNodeWithId;
const uuid_1 = require("uuid");
const node_types_1 = require("../../constants/node-types");
const base_1 = require("../../types/base");
const type_guards_1 = require("../type-guards");
function isInputTarget(value) {
    return (typeof value === 'object' &&
        value !== null &&
        '_isInputTarget' in value &&
        value._isInputTarget);
}
function isOutputSelector(value) {
    return (typeof value === 'object' &&
        value !== null &&
        '_isOutputSelector' in value &&
        value._isOutputSelector);
}
function getCompositeOutputNode(value) {
    if ((0, type_guards_1.isSwitchCaseComposite)(value)) {
        return value.switchNode;
    }
    if (isSwitchCaseBuilder(value)) {
        return value.switchNode;
    }
    if ((0, type_guards_1.isIfElseComposite)(value)) {
        return value.ifNode;
    }
    if ((0, type_guards_1.isSplitInBatchesBuilder)(value)) {
        return (0, type_guards_1.extractSplitInBatchesBuilder)(value).sibNode;
    }
    return null;
}
function generateNodeName(type) {
    const parts = type.split('.');
    const nodeName = parts[parts.length - 1];
    return nodeName
        .replace(/([a-z])([A-Z])/g, '$1 $2')
        .replace(/([A-Z]+)([A-Z][a-z])/g, '$1 $2')
        .replace(/^./, (str) => str.toUpperCase())
        .replace(/Http/g, 'HTTP')
        .replace(/Api/g, 'API')
        .replace(/Url/g, 'URL')
        .replace(/Id/g, 'ID')
        .replace(/Json/g, 'JSON')
        .replace(/Xml/g, 'XML')
        .replace(/Sql/g, 'SQL')
        .replace(/Ssh/g, 'SSH')
        .replace(/Ftp/g, 'FTP')
        .replace(/Aws/g, 'AWS')
        .replace(/Gcp/g, 'GCP');
}
class NodeInstanceImpl {
    constructor(type, version, config, id, name, connections) {
        this._connections = [];
        this.type = type;
        this.version = version;
        this.config = { ...config };
        this.id = id ?? (0, uuid_1.v4)();
        this.name = name ?? config.name ?? generateNodeName(type);
        this._connections = connections ?? [];
    }
    update(config) {
        const mergedConfig = {
            ...this.config,
            ...config,
            parameters: config.parameters ?? this.config.parameters,
            credentials: config.credentials ?? this.config.credentials,
        };
        return new NodeInstanceImpl(this.type, this.version, mergedConfig, this.id, this.name, this._connections);
    }
    to(target, outputIndex = 0) {
        if (isInputTarget(target)) {
            this._connections.push({
                target: target.node,
                outputIndex,
                targetInputIndex: target.inputIndex,
            });
            return new NodeChainImpl(this, target.node, [this, target.node]);
        }
        const targets = Array.isArray(target)
            ? target
            : [target];
        for (const t of targets) {
            this._connections.push({ target: t, outputIndex });
        }
        const flattenTarget = (t) => {
            if ((0, base_1.isNodeChain)(t)) {
                return t.allNodes;
            }
            if ((0, type_guards_1.isSplitInBatchesBuilder)(t)) {
                return [t];
            }
            if (isIfElseBuilder(t)) {
                return [t];
            }
            if (isSwitchCaseBuilder(t)) {
                return [t];
            }
            return [t];
        };
        const allTargetNodes = targets.flatMap(flattenTarget);
        const lastTarget = targets[targets.length - 1];
        return new NodeChainImpl(this, lastTarget, [this, ...allTargetNodes]);
    }
    input(index) {
        return {
            _isInputTarget: true,
            node: this,
            inputIndex: index,
        };
    }
    output(index) {
        return new OutputSelectorImpl(this, index);
    }
    onTrue(target) {
        if (!(0, node_types_1.isIfNodeType)(this.type)) {
            throw new Error(`.onTrue() is only available on IF nodes (${node_types_1.NODE_TYPES.IF})`);
        }
        const builder = new IfElseBuilderImpl(this);
        return builder.onTrue(target);
    }
    onFalse(target) {
        if (!(0, node_types_1.isIfNodeType)(this.type)) {
            throw new Error(`.onFalse() is only available on IF nodes (${node_types_1.NODE_TYPES.IF})`);
        }
        const builder = new IfElseBuilderImpl(this);
        return builder.onFalse(target);
    }
    onCase(index, target) {
        if (!(0, node_types_1.isSwitchNodeType)(this.type)) {
            throw new Error(`.onCase() is only available on Switch nodes (${node_types_1.NODE_TYPES.SWITCH})`);
        }
        const builder = new SwitchCaseBuilderImpl(this);
        return builder.onCase(index, target);
    }
    onError(handler) {
        if (isInputTarget(handler)) {
            this._connections.push({
                target: handler.node,
                outputIndex: 0,
                targetInputIndex: handler.inputIndex,
                connectionType: 'error',
            });
        }
        else {
            this._connections.push({ target: handler, outputIndex: 0, connectionType: 'error' });
        }
        return this;
    }
    getConnections() {
        return [...this._connections];
    }
}
class TriggerInstanceImpl extends NodeInstanceImpl {
    constructor() {
        super(...arguments);
        this.isTrigger = true;
    }
}
class NodeChainImpl {
    constructor(head, tail, allNodes) {
        this._isChain = true;
        this.head = head;
        this.tail = tail;
        this.allNodes = allNodes;
    }
    get type() {
        return this.tail.type;
    }
    get version() {
        return this.tail.version;
    }
    get config() {
        return this.tail.config;
    }
    get id() {
        return this.tail.id;
    }
    get name() {
        return this.tail.name;
    }
    get _outputType() {
        return this.tail._outputType;
    }
    update(config) {
        return this.tail.update(config);
    }
    to(target, outputIndex = 0) {
        if (isInputTarget(target)) {
            this.tail.to(target, outputIndex);
            return new NodeChainImpl(this.head, target.node, [...this.allNodes, target.node]);
        }
        const targets = Array.isArray(target)
            ? target
            : [target];
        const flattenTarget = (t) => {
            if ((0, base_1.isNodeChain)(t)) {
                return t.allNodes;
            }
            if ((0, type_guards_1.isSplitInBatchesBuilder)(t)) {
                return [t];
            }
            if (isIfElseBuilder(t)) {
                return [t];
            }
            if (isSwitchCaseBuilder(t)) {
                return [t];
            }
            return [t];
        };
        const allTargetNodes = targets.flatMap(flattenTarget);
        const outputNode = getCompositeOutputNode(this.tail);
        if (outputNode) {
            if (typeof outputNode.to === 'function') {
                outputNode.to(targets, outputIndex);
            }
            const lastTarget = targets[targets.length - 1];
            return new NodeChainImpl(this.head, lastTarget, [...this.allNodes, ...allTargetNodes]);
        }
        this.tail.to(targets, outputIndex);
        const lastTarget = targets[targets.length - 1];
        return new NodeChainImpl(this.head, lastTarget, [...this.allNodes, ...allTargetNodes]);
    }
    input(index) {
        return this.tail.input(index);
    }
    output(index) {
        const compositeNode = getCompositeOutputNode(this.tail);
        if (compositeNode) {
            return compositeNode.output(index);
        }
        return this.tail.output(index);
    }
    onTrue(target) {
        if (!this.tail.onTrue) {
            throw new Error(`.onTrue() is only available on IF nodes (${node_types_1.NODE_TYPES.IF})`);
        }
        return this.tail.onTrue(target);
    }
    onFalse(target) {
        if (!this.tail.onFalse) {
            throw new Error(`.onFalse() is only available on IF nodes (${node_types_1.NODE_TYPES.IF})`);
        }
        return this.tail.onFalse(target);
    }
    onCase(index, target) {
        if (!this.tail.onCase) {
            throw new Error(`.onCase() is only available on Switch nodes (${node_types_1.NODE_TYPES.SWITCH})`);
        }
        const builder = this.tail.onCase(index, target);
        builder.sourceChain = this;
        return builder;
    }
    onError(handler) {
        this.tail.onError(handler);
        return this;
    }
    getConnections() {
        const allConnections = [];
        for (const node of this.allNodes) {
            if (node === null) {
                continue;
            }
            if (typeof node.getConnections === 'function') {
                allConnections.push(...node.getConnections());
            }
        }
        return allConnections;
    }
}
class OutputSelectorImpl {
    constructor(node, outputIndex) {
        this._isOutputSelector = true;
        this.node = node;
        this.outputIndex = outputIndex;
    }
    to(target) {
        return this.node.to(target, this.outputIndex);
    }
}
function extractNodesFromTarget(target) {
    if (target === null)
        return [];
    if (Array.isArray(target)) {
        const nodes = [];
        for (const t of target) {
            nodes.push(...extractNodesFromTarget(t));
        }
        return nodes;
    }
    if ((0, base_1.isNodeChain)(target)) {
        const nodes = [];
        for (const chainNode of target.allNodes) {
            nodes.push(...extractNodesFromTarget(chainNode));
        }
        return nodes;
    }
    if (target !== null &&
        typeof target === 'object' &&
        '_isIfElseBuilder' in target &&
        target._isIfElseBuilder) {
        const builder = target;
        const nodes = [builder.ifNode];
        nodes.push(...extractNodesFromTarget(builder.trueBranch));
        nodes.push(...extractNodesFromTarget(builder.falseBranch));
        return nodes;
    }
    if (target !== null &&
        typeof target === 'object' &&
        '_isSwitchCaseBuilder' in target &&
        target._isSwitchCaseBuilder) {
        const builder = target;
        const nodes = [builder.switchNode];
        for (const caseTarget of builder.caseMapping.values()) {
            nodes.push(...extractNodesFromTarget(caseTarget));
        }
        return nodes;
    }
    if (target !== null &&
        typeof target === 'object' &&
        'type' in target &&
        'version' in target &&
        'config' in target) {
        return [target];
    }
    return [];
}
function isIfElseBuilder(value) {
    return (value !== null &&
        typeof value === 'object' &&
        '_isIfElseBuilder' in value &&
        value._isIfElseBuilder);
}
function isSwitchCaseBuilder(value) {
    return (value !== null &&
        typeof value === 'object' &&
        '_isSwitchCaseBuilder' in value &&
        value._isSwitchCaseBuilder);
}
class IfElseBuilderImpl {
    constructor(ifNode) {
        this._isIfElseBuilder = true;
        this.trueBranch = null;
        this.falseBranch = null;
        this._allBranchNodes = [];
        this.ifNode = ifNode;
    }
    onTrue(target) {
        this.trueBranch = target;
        this._updateAllBranchNodes();
        return this;
    }
    onFalse(target) {
        this.falseBranch = target;
        this._updateAllBranchNodes();
        return this;
    }
    onError(target) {
        this.errorBranch = target;
        this._updateAllBranchNodes();
        return this;
    }
    to(target, outputIndex = 0) {
        return this.ifNode.to(target, outputIndex);
    }
    _updateAllBranchNodes() {
        const allNodes = [];
        for (const node of extractNodesFromTarget(this.trueBranch)) {
            if (!allNodes.some((n) => n.name === node.name)) {
                allNodes.push(node);
            }
        }
        for (const node of extractNodesFromTarget(this.falseBranch)) {
            if (!allNodes.some((n) => n.name === node.name)) {
                allNodes.push(node);
            }
        }
        if (this.errorBranch) {
            for (const node of extractNodesFromTarget(this.errorBranch)) {
                if (!allNodes.some((n) => n.name === node.name)) {
                    allNodes.push(node);
                }
            }
        }
        this._allBranchNodes = allNodes;
    }
}
class SwitchCaseBuilderImpl {
    constructor(switchNode) {
        this._isSwitchCaseBuilder = true;
        this.caseMapping = new Map();
        this._allCaseNodes = [];
        this.switchNode = switchNode;
    }
    onCase(index, target) {
        this.caseMapping.set(index, target);
        this._updateAllCaseNodes();
        return this;
    }
    to(target, outputIndex = 0) {
        return this.switchNode.to(target, outputIndex);
    }
    _updateAllCaseNodes() {
        const allNodes = [];
        for (const target of this.caseMapping.values()) {
            for (const node of extractNodesFromTarget(target)) {
                if (!allNodes.some((n) => n.name === node.name)) {
                    allNodes.push(node);
                }
            }
        }
        this._allCaseNodes = allNodes;
    }
}
function node(input) {
    const versionStr = String(input.version);
    const config = input.output
        ? { ...input.config, output: input.output }
        : input.config;
    return new NodeInstanceImpl(input.type, versionStr, config);
}
function ifElse(input) {
    return node({
        type: 'n8n-nodes-base.if',
        version: input.version,
        config: input.config ?? {},
    });
}
function merge(input) {
    return node({
        type: 'n8n-nodes-base.merge',
        version: input.version,
        config: input.config ?? {},
    });
}
function switchCase(input) {
    return node({
        type: 'n8n-nodes-base.switch',
        version: input.version,
        config: input.config ?? {},
    });
}
function trigger(input) {
    const versionStr = String(input.version);
    const config = input.output
        ? { ...input.config, output: input.output }
        : input.config;
    return new TriggerInstanceImpl(input.type, versionStr, config);
}
const DEFAULT_NODE_WIDTH = 200;
const DEFAULT_NODE_HEIGHT = 100;
const STICKY_PADDING = 50;
function calculateNodesBoundingBox(nodes) {
    if (nodes.length === 0)
        return null;
    const normalizedNodes = nodes
        .map((item) => {
        if ((0, type_guards_1.isSplitInBatchesBuilder)(item)) {
            return (0, type_guards_1.extractSplitInBatchesBuilder)(item).sibNode;
        }
        if (isIfElseBuilder(item)) {
            return item.ifNode;
        }
        if (isSwitchCaseBuilder(item)) {
            return item.switchNode;
        }
        if (item.config) {
            return item;
        }
        return null;
    })
        .filter((n) => n !== null);
    if (normalizedNodes.length === 0)
        return null;
    let minX = Infinity;
    let minY = Infinity;
    let maxX = -Infinity;
    let maxY = -Infinity;
    for (const node of normalizedNodes) {
        const pos = node.config.position ?? [0, 0];
        const x = pos[0];
        const y = pos[1];
        minX = Math.min(minX, x);
        minY = Math.min(minY, y);
        maxX = Math.max(maxX, x + DEFAULT_NODE_WIDTH);
        maxY = Math.max(maxY, y + DEFAULT_NODE_HEIGHT);
    }
    return {
        position: [minX - STICKY_PADDING, minY - STICKY_PADDING],
        width: maxX - minX + STICKY_PADDING * 2,
        height: maxY - minY + STICKY_PADDING * 2,
    };
}
class StickyNoteInstance {
    constructor(content, nodes = [], stickyConfig = {}) {
        this.type = 'n8n-nodes-base.stickyNote';
        this.version = 'v1';
        this.id = (0, uuid_1.v4)();
        this.name = stickyConfig.name ?? `Sticky Note ${this.id.slice(0, 8)}`;
        const boundingBox = nodes.length > 0 ? calculateNodesBoundingBox(nodes) : null;
        this.config = {
            name: this.name,
            position: stickyConfig.position ?? boundingBox?.position,
            parameters: {
                content,
                ...(stickyConfig.color !== undefined && { color: stickyConfig.color }),
                ...((stickyConfig.width ?? boundingBox?.width) !== undefined && {
                    width: stickyConfig.width ?? boundingBox?.width,
                }),
                ...((stickyConfig.height ?? boundingBox?.height) !== undefined && {
                    height: stickyConfig.height ?? boundingBox?.height,
                }),
            },
        };
    }
    update(config) {
        const newContent = config.parameters?.content ?? this.config.parameters?.content;
        const newConfig = {
            position: config.position ?? this.config.position,
            color: config.parameters?.color ?? this.config.parameters?.color,
            width: config.parameters?.width ?? this.config.parameters?.width,
            height: config.parameters?.height ?? this.config.parameters?.height,
            name: config.name ?? this.name,
        };
        return new StickyNoteInstance(newContent, [], newConfig);
    }
    input(_index) {
        throw new Error('Sticky notes do not support input connections');
    }
    output(_index) {
        throw new Error('Sticky notes do not support output connections');
    }
    to(_target, _outputIndex) {
        throw new Error('Sticky notes do not support connections');
    }
    onError(_handler) {
        throw new Error('Sticky notes do not support error handlers');
    }
    getConnections() {
        return [];
    }
}
function sticky(content, nodesOrConfig, config) {
    if (nodesOrConfig !== undefined && !Array.isArray(nodesOrConfig)) {
        return new StickyNoteInstance(content, [], nodesOrConfig);
    }
    const nodes = nodesOrConfig ?? [];
    return new StickyNoteInstance(content, nodes, config ?? {});
}
class PlaceholderImpl {
    constructor(hint) {
        this.__placeholder = true;
        this.hint = hint;
    }
    toString() {
        return `<__PLACEHOLDER_VALUE__${this.hint}__>`;
    }
    toJSON() {
        return this.toString();
    }
}
function placeholder(hint) {
    return new PlaceholderImpl(hint);
}
class NewCredentialImpl {
    constructor(name, id) {
        this.__newCredential = true;
        this.name = name;
        this.id = id;
    }
    toJSON() {
        if (this.id !== undefined) {
            return { id: this.id, name: this.name };
        }
        return undefined;
    }
}
function newCredential(name, id) {
    return new NewCredentialImpl(name, id);
}
function cloneNodeWithId(instance, newId) {
    const connections = typeof instance.getConnections === 'function' ? instance.getConnections() : [];
    const isTrigger = 'isTrigger' in instance && instance.isTrigger === true;
    if (isTrigger) {
        return new TriggerInstanceImpl(instance.type, instance.version, instance.config, newId, instance.name, connections);
    }
    return new NodeInstanceImpl(instance.type, instance.version, instance.config, newId, instance.name, connections);
}
