// @n8n/workflow-sdk/dist/workflow-builder/node-builders/subnode-builders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.embeddings = void 0;
exports.languageModel = languageModel;
exports.memory = memory;
exports.tool = tool;
exports.outputParser = outputParser;
exports.embedding = embedding;
exports.vectorStore = vectorStore;
exports.retriever = retriever;
exports.documentLoader = documentLoader;
exports.textSplitter = textSplitter;
exports.reranker = reranker;
exports.fromAi = fromAi;
const uuid_1 = require("uuid");
const expression_1 = require("../../expression");
function generateNodeName(type) {
    const parts = type.split('.');
    const nodeName = parts[parts.length - 1];
    return nodeName
        .replace(/([a-z])([A-Z])/g, '$1 $2')
        .replace(/([A-Z]+)([A-Z][a-z])/g, '$1 $2')
        .replace(/^./, (str) => str.toUpperCase())
        .replace(/^Lm/, 'LM')
        .replace(/^Llm/, 'LLM')
        .replace(/OpenAi/g, 'OpenAI')
        .replace(/Api/g, 'API');
}
class SubnodeInstanceImpl {
    constructor(type, version, config, subnodeType, id, name) {
        this.type = type;
        this.version = version;
        this.config = { ...config };
        this.id = id ?? (0, uuid_1.v4)();
        this.name = name ?? config.name ?? generateNodeName(type);
        this._subnodeType = subnodeType;
    }
    update(config) {
        const mergedConfig = {
            ...this.config,
            ...config,
            parameters: config.parameters ?? this.config.parameters,
            credentials: config.credentials ?? this.config.credentials,
        };
        return new SubnodeInstanceImpl(this.type, this.version, mergedConfig, this._subnodeType, this.id, this.name);
    }
    input(_index) {
        throw new Error('Subnode input connections are managed by parent node SubnodeConfig');
    }
    output(_index) {
        throw new Error('Subnode output connections are managed by parent node SubnodeConfig');
    }
    then(_target, _outputIndex) {
        throw new Error('Subnode connections are managed by parent node SubnodeConfig');
    }
    to(_target, _outputIndex) {
        throw new Error('Subnode connections are managed by parent node SubnodeConfig');
    }
    onError(_handler) {
        throw new Error('Subnode error handling is managed by parent node SubnodeConfig');
    }
    getConnections() {
        return [];
    }
}
function languageModel(input) {
    const versionStr = String(input.version);
    return new SubnodeInstanceImpl(input.type, versionStr, input.config, 'ai_languageModel');
}
function memory(input) {
    const versionStr = String(input.version);
    return new SubnodeInstanceImpl(input.type, versionStr, input.config, 'ai_memory');
}
function tool(input) {
    const versionStr = String(input.version);
    return new SubnodeInstanceImpl(input.type, versionStr, input.config, 'ai_tool');
}
function outputParser(input) {
    const versionStr = String(input.version);
    return new SubnodeInstanceImpl(input.type, versionStr, input.config, 'ai_outputParser');
}
function embedding(input) {
    const versionStr = String(input.version);
    return new SubnodeInstanceImpl(input.type, versionStr, input.config, 'ai_embedding');
}
function vectorStore(input) {
    const versionStr = String(input.version);
    return new SubnodeInstanceImpl(input.type, versionStr, input.config, 'ai_vectorStore');
}
function retriever(input) {
    const versionStr = String(input.version);
    return new SubnodeInstanceImpl(input.type, versionStr, input.config, 'ai_retriever');
}
function documentLoader(input) {
    const versionStr = String(input.version);
    return new SubnodeInstanceImpl(input.type, versionStr, input.config, 'ai_document');
}
function textSplitter(input) {
    const versionStr = String(input.version);
    return new SubnodeInstanceImpl(input.type, versionStr, input.config, 'ai_textSplitter');
}
function reranker(input) {
    const versionStr = String(input.version);
    return new SubnodeInstanceImpl(input.type, versionStr, input.config, 'ai_reranker');
}
exports.embeddings = embedding;
function fromAi(key, description, type, defaultValue) {
    return (0, expression_1.createFromAIExpression)(key, description, type, defaultValue);
}
