// n8n-core/dist/storage.config.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageConfig = exports.EXECUTION_DATA_STORAGE_MODES = void 0;
const backend_common_1 = require("@n8n/backend-common");
const config_1 = require("@n8n/config");
const node_fs_1 = require("node:fs");
const node_path_1 = __importDefault(require("node:path"));
const zod_1 = require("zod");
const instance_settings_1 = require("./instance-settings");
const storage_path_conflict_error_1 = require("./storage-path-conflict.error");
exports.EXECUTION_DATA_STORAGE_MODES = ['database', 'filesystem'];
const modeSchema = zod_1.z.enum(exports.EXECUTION_DATA_STORAGE_MODES);
let StorageConfig = class StorageConfig {
    get modeTag() {
        return this.mode === 'database' ? 'db' : 'fs';
    }
    constructor(instanceSettings, logger) {
        this.mode = 'database';
        this.instanceSettings = instanceSettings;
        this.logger = logger;
        this.storagePath = node_path_1.default.join(instanceSettings.n8nFolder, 'storage');
    }
    sanitize() {
        const storagePath = process.env.N8N_STORAGE_PATH;
        const binaryDataStoragePath = process.env.N8N_BINARY_DATA_STORAGE_PATH;
        if (storagePath && binaryDataStoragePath && storagePath !== binaryDataStoragePath) {
            throw storage_path_conflict_error_1.StoragePathError.conflict();
        }
        this.migrateStorageDir();
    }
    migrateStorageDir() {
        if (this.instanceSettings.fsStorageMigrated)
            return;
        if (process.env.N8N_STORAGE_PATH)
            return;
        if (process.env.N8N_BINARY_DATA_STORAGE_PATH)
            return;
        const { n8nFolder } = this.instanceSettings;
        const oldPath = node_path_1.default.join(n8nFolder, 'binaryData');
        if (!(0, node_fs_1.existsSync)(oldPath))
            return;
        const shouldMigrate = process.env.N8N_MIGRATE_FS_STORAGE_PATH === 'true';
        if (!shouldMigrate) {
            this.logger.warn(`Deprecation warning: The storage directory "${oldPath}" will be renamed to "${node_path_1.default.join(n8nFolder, 'storage')}" in n8n v3. To migrate now, set N8N_MIGRATE_FS_STORAGE_PATH=true. If you have a volume mounted at the old path, update your mount configuration after migration.`);
            this.storagePath = oldPath;
            return;
        }
        const newPath = node_path_1.default.join(n8nFolder, 'storage');
        if ((0, node_fs_1.existsSync)(newPath)) {
            throw storage_path_conflict_error_1.StoragePathError.taken(oldPath, newPath);
        }
        try {
            (0, node_fs_1.renameSync)(oldPath, newPath);
            this.instanceSettings.markFsStorageMigrated();
        }
        catch (error) {
            if (this.isIgnorableRenameError(error))
                return;
            throw error;
        }
    }
    isIgnorableRenameError(error) {
        return (error !== null &&
            typeof error === 'object' &&
            'code' in error &&
            (error.code === 'ENOENT' || error.code === 'EEXIST'));
    }
};
exports.StorageConfig = StorageConfig;
__decorate([
    (0, config_1.Env)('N8N_EXECUTION_DATA_STORAGE_MODE', modeSchema),
    __metadata("design:type", void 0)
], StorageConfig.prototype, "mode", void 0);
__decorate([
    (0, config_1.Env)('N8N_STORAGE_PATH'),
    __metadata("design:type", String)
], StorageConfig.prototype, "storagePath", void 0);
exports.StorageConfig = StorageConfig = __decorate([
    config_1.Config,
    __metadata("design:paramtypes", [instance_settings_1.InstanceSettings, backend_common_1.Logger])
], StorageConfig);
