// @n8n/workflow-sdk/dist/workflow-builder/plugins/composite-handlers/branch-handler-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTargetNodeId = getTargetNodeId;
exports.getTargetNodeName = getTargetNodeName;
exports.collectFromTarget = collectFromTarget;
exports.addBranchTargetNodes = addBranchTargetNodes;
exports.processBranchForComposite = processBranchForComposite;
exports.processBranchForBuilder = processBranchForBuilder;
exports.fixupBranchConnectionTargets = fixupBranchConnectionTargets;
const base_1 = require("../../../types/base");
const node_builder_1 = require("../../node-builders/node-builder");
const type_guards_1 = require("../../type-guards");
function getTargetNodeId(target) {
    if (target === null || target === undefined)
        return undefined;
    if ((0, base_1.isNodeChain)(target)) {
        return target.head.id;
    }
    if ((0, type_guards_1.isIfElseComposite)(target)) {
        return target.ifNode.id;
    }
    if ((0, type_guards_1.isSwitchCaseComposite)(target)) {
        return target.switchNode.id;
    }
    if ((0, node_builder_1.isIfElseBuilder)(target)) {
        return target.ifNode.id;
    }
    if ((0, node_builder_1.isSwitchCaseBuilder)(target)) {
        return target.switchNode.id;
    }
    if ((0, type_guards_1.isSplitInBatchesBuilder)(target)) {
        const builder = (0, type_guards_1.extractSplitInBatchesBuilder)(target);
        return builder.sibNode.id;
    }
    if (typeof target.id === 'string') {
        return target.id;
    }
    return undefined;
}
function getTargetNodeName(target) {
    if (target === null || target === undefined)
        return undefined;
    if ((0, base_1.isNodeChain)(target)) {
        return target.head.name;
    }
    if ((0, type_guards_1.isIfElseComposite)(target)) {
        return target.ifNode.name;
    }
    if ((0, type_guards_1.isSwitchCaseComposite)(target)) {
        return target.switchNode.name;
    }
    if ((0, node_builder_1.isIfElseBuilder)(target)) {
        return target.ifNode.name;
    }
    if ((0, node_builder_1.isSwitchCaseBuilder)(target)) {
        return target.switchNode.name;
    }
    if ((0, type_guards_1.isSplitInBatchesBuilder)(target)) {
        const builder = (0, type_guards_1.extractSplitInBatchesBuilder)(target);
        return builder.sibNode.name;
    }
    if (typeof target.name === 'string') {
        return target.name;
    }
    return undefined;
}
function collectFromTarget(target, collector) {
    if (target === null || target === undefined)
        return;
    if (Array.isArray(target)) {
        for (const n of target) {
            if (n !== null && n !== undefined) {
                collector(n);
            }
        }
    }
    else {
        collector(target);
    }
}
function addBranchTargetNodes(target, ctx) {
    if (target === null || target === undefined)
        return;
    if (Array.isArray(target)) {
        for (const t of target) {
            addBranchTargetNodes(t, ctx);
        }
        return;
    }
    ctx.addBranchToGraph(target);
}
function processBranchForComposite(branch, outputIndex, ctx, mainConns) {
    if (branch === null || branch === undefined) {
        return;
    }
    if (Array.isArray(branch)) {
        const targets = [];
        for (const branchNode of branch) {
            if (branchNode === null)
                continue;
            const branchHead = ctx.addBranchToGraph(branchNode);
            targets.push({ node: branchHead, type: 'main', index: 0 });
        }
        if (targets.length > 0) {
            mainConns.set(outputIndex, targets);
        }
    }
    else {
        const branchHead = ctx.addBranchToGraph(branch);
        mainConns.set(outputIndex, [{ node: branchHead, type: 'main', index: 0 }]);
    }
}
function processBranchForBuilder(branch, outputIndex, mainConns, targetNodeIds) {
    if (branch === null || branch === undefined) {
        return;
    }
    if (Array.isArray(branch)) {
        const targets = [];
        const ids = [];
        for (const t of branch) {
            const targetName = getTargetNodeName(t);
            if (targetName) {
                targets.push({ node: targetName, type: 'main', index: 0 });
                const id = getTargetNodeId(t);
                if (id)
                    ids.push(id);
            }
        }
        if (targets.length > 0) {
            mainConns.set(outputIndex, targets);
            if (targetNodeIds && ids.length > 0) {
                targetNodeIds.set(outputIndex, ids);
            }
        }
    }
    else {
        const targetName = getTargetNodeName(branch);
        if (targetName) {
            mainConns.set(outputIndex, [{ node: targetName, type: 'main', index: 0 }]);
            const id = getTargetNodeId(branch);
            if (targetNodeIds && id) {
                targetNodeIds.set(outputIndex, [id]);
            }
        }
    }
}
function fixupBranchConnectionTargets(mainConns, targetNodeIds, nameMapping) {
    if (nameMapping.size === 0)
        return;
    for (const [outputIndex, ids] of targetNodeIds) {
        const targets = mainConns.get(outputIndex);
        if (!targets)
            continue;
        for (let i = 0; i < ids.length && i < targets.length; i++) {
            const actualName = nameMapping.get(ids[i]);
            if (actualName && actualName !== targets[i].node) {
                targets[i].node = actualName;
            }
        }
    }
}
