// @n8n/workflow-sdk/dist/workflow-builder/plugins/composite-handlers/split-in-batches-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitInBatchesHandler = void 0;
const base_1 = require("../../../types/base");
function isSplitInBatchesBuilderShape(value) {
    if (value === null || typeof value !== 'object')
        return false;
    return 'sibNode' in value && '_doneNodes' in value && '_eachNodes' in value;
}
const processingSibBuilders = new WeakSet();
exports.splitInBatchesHandler = {
    id: 'core:split-in-batches',
    name: 'Split In Batches Handler',
    priority: 100,
    canHandle(input) {
        return isSplitInBatchesBuilderShape(input);
    },
    getHeadNodeName(input) {
        return { name: input.sibNode.name, id: input.sibNode.id };
    },
    collectPinData(input, collector) {
        collector(input.sibNode);
    },
    addNodes(input, ctx) {
        if (processingSibBuilders.has(input)) {
            return input.sibNode.name;
        }
        processingSibBuilders.add(input);
        try {
            const hasNamedSyntax = '_doneTarget' in input || '_eachTarget' in input;
            if (hasNamedSyntax) {
                return processNamedSyntax(input, ctx);
            }
            const hasFluentApi = (input._doneBatches && input._doneBatches.length > 0) ||
                (input._eachBatches && input._eachBatches.length > 0);
            if (hasFluentApi) {
                return processFluentApi(input, ctx);
            }
            const sibConns = new Map();
            sibConns.set('main', new Map());
            ctx.nodes.set(input.sibNode.name, {
                instance: input.sibNode,
                connections: sibConns,
            });
            return input.sibNode.name;
        }
        finally {
            processingSibBuilders.delete(input);
        }
    },
};
function processNamedSyntax(input, ctx) {
    const sibMainConns = new Map();
    if (input._doneTarget !== null && input._doneTarget !== undefined) {
        const doneTarget = input._doneTarget;
        if (Array.isArray(doneTarget)) {
            const targets = [];
            for (const target of doneTarget) {
                const targetHead = ctx.addBranchToGraph(target);
                targets.push({ node: targetHead, type: 'main', index: 0 });
            }
            sibMainConns.set(0, targets);
        }
        else {
            const targetHead = ctx.addBranchToGraph(doneTarget);
            sibMainConns.set(0, [{ node: targetHead, type: 'main', index: 0 }]);
        }
    }
    if (input._eachTarget !== null && input._eachTarget !== undefined) {
        const eachTarget = input._eachTarget;
        if (Array.isArray(eachTarget)) {
            const targets = [];
            for (const target of eachTarget) {
                const targetHead = ctx.addBranchToGraph(target);
                targets.push({ node: targetHead, type: 'main', index: 0 });
            }
            sibMainConns.set(1, targets);
        }
        else {
            const targetHead = ctx.addBranchToGraph(eachTarget);
            sibMainConns.set(1, [{ node: targetHead, type: 'main', index: 0 }]);
        }
    }
    const sibConns = new Map();
    sibConns.set('main', sibMainConns);
    ctx.nodes.set(input.sibNode.name, {
        instance: input.sibNode,
        connections: sibConns,
    });
    return input.sibNode.name;
}
function processFluentApi(input, ctx) {
    const sibMainConns = new Map();
    let lastEachNode = null;
    if (input._doneTarget !== null && input._doneTarget !== undefined) {
        const doneTarget = input._doneTarget;
        if (Array.isArray(doneTarget)) {
            const targets = [];
            for (const target of doneTarget) {
                const targetHead = ctx.addBranchToGraph(target);
                targets.push({ node: targetHead, type: 'main', index: 0 });
            }
            sibMainConns.set(0, targets);
        }
        else {
            const targetHead = ctx.addBranchToGraph(doneTarget);
            sibMainConns.set(0, [{ node: targetHead, type: 'main', index: 0 }]);
        }
    }
    else if (input._doneBatches && input._doneBatches.length > 0) {
        const { targets } = processBatches(input._doneBatches, ctx);
        if (targets.length > 0) {
            sibMainConns.set(0, targets);
        }
    }
    if (input._eachTarget !== null && input._eachTarget !== undefined) {
        const eachTarget = input._eachTarget;
        if (Array.isArray(eachTarget)) {
            const targets = [];
            for (const target of eachTarget) {
                const targetHead = ctx.addBranchToGraph(target);
                targets.push({ node: targetHead, type: 'main', index: 0 });
                if ((0, base_1.isNodeChain)(target)) {
                    lastEachNode = target.tail?.name ?? targetHead;
                }
                else {
                    lastEachNode = targetHead;
                }
            }
            sibMainConns.set(1, targets);
        }
        else {
            const targetHead = ctx.addBranchToGraph(eachTarget);
            sibMainConns.set(1, [{ node: targetHead, type: 'main', index: 0 }]);
            if ((0, base_1.isNodeChain)(eachTarget)) {
                lastEachNode = eachTarget.tail?.name ?? targetHead;
            }
            else {
                lastEachNode = targetHead;
            }
        }
    }
    else if (input._eachBatches && input._eachBatches.length > 0) {
        const { targets, lastNode } = processBatches(input._eachBatches, ctx);
        if (targets.length > 0) {
            sibMainConns.set(1, targets);
        }
        lastEachNode = lastNode;
    }
    const sibConns = new Map();
    sibConns.set('main', sibMainConns);
    ctx.nodes.set(input.sibNode.name, {
        instance: input.sibNode,
        connections: sibConns,
    });
    if (input._hasLoop && lastEachNode) {
        const lastEachGraphNode = ctx.nodes.get(lastEachNode);
        if (lastEachGraphNode) {
            const lastEachMainConns = lastEachGraphNode.connections.get('main') ?? new Map();
            const existingConns = lastEachMainConns.get(0) ?? [];
            lastEachMainConns.set(0, [
                ...existingConns,
                { node: input.sibNode.name, type: 'main', index: 0 },
            ]);
            lastEachGraphNode.connections.set('main', lastEachMainConns);
        }
    }
    return input.sibNode.name;
}
function processBatches(batches, ctx) {
    const targets = [];
    let lastNode = null;
    let previousBatchNodes = [];
    for (let i = 0; i < batches.length; i++) {
        const batch = batches[i];
        const currentBatchNodes = [];
        if (Array.isArray(batch)) {
            for (const node of batch) {
                const nodeName = ctx.addBranchToGraph(node);
                currentBatchNodes.push(nodeName);
                if (i === 0) {
                    targets.push({ node: nodeName, type: 'main', index: 0 });
                }
            }
        }
        else {
            const nodeName = ctx.addBranchToGraph(batch);
            currentBatchNodes.push(nodeName);
            if (i === 0) {
                targets.push({ node: nodeName, type: 'main', index: 0 });
            }
        }
        if (previousBatchNodes.length > 0 && currentBatchNodes.length > 0) {
            for (const prevNode of previousBatchNodes) {
                const prevGraphNode = ctx.nodes.get(prevNode);
                if (prevGraphNode) {
                    const prevMainConns = prevGraphNode.connections.get('main') ?? new Map();
                    const existingConns = prevMainConns.get(0) ?? [];
                    const newConns = currentBatchNodes.map((n) => ({
                        node: n,
                        type: 'main',
                        index: 0,
                    }));
                    prevMainConns.set(0, [...existingConns, ...newConns]);
                    prevGraphNode.connections.set('main', prevMainConns);
                }
            }
        }
        previousBatchNodes = currentBatchNodes;
        if (currentBatchNodes.length > 0) {
            lastNode = currentBatchNodes[currentBatchNodes.length - 1];
        }
    }
    return { targets, lastNode };
}
