// @n8n/workflow-sdk/dist/workflow-builder/plugins/composite-handlers/switch-case-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.switchCaseHandler = void 0;
const branch_handler_utils_1 = require("./branch-handler-utils");
const node_builder_1 = require("../../node-builders/node-builder");
const type_guards_1 = require("../../type-guards");
exports.switchCaseHandler = {
    id: 'core:switch-case',
    name: 'Switch/Case Handler',
    priority: 100,
    canHandle(input) {
        return (0, type_guards_1.isSwitchCaseComposite)(input) || (0, node_builder_1.isSwitchCaseBuilder)(input);
    },
    getHeadNodeName(input) {
        if ((0, node_builder_1.isSwitchCaseBuilder)(input)) {
            return { name: input.switchNode.name, id: input.switchNode.id };
        }
        const composite = input;
        return { name: composite.switchNode.name, id: composite.switchNode.id };
    },
    collectPinData(input, collector) {
        collector(input.switchNode);
        if ('caseMapping' in input && input.caseMapping instanceof Map) {
            for (const [, target] of input.caseMapping) {
                (0, branch_handler_utils_1.collectFromTarget)(target, collector);
            }
        }
        else if ('cases' in input && Array.isArray(input.cases)) {
            for (const caseNode of input.cases) {
                (0, branch_handler_utils_1.collectFromTarget)(caseNode, collector);
            }
        }
    },
    addNodes(input, ctx) {
        const builderWithChain = input;
        if (builderWithChain.sourceChain) {
            ctx.addBranchToGraph(builderWithChain.sourceChain);
        }
        const switchMainConns = new Map();
        if ('caseMapping' in input && input.caseMapping instanceof Map) {
            const builder = input;
            const targetNodeIds = new Map();
            for (const [caseIndex, target] of builder.caseMapping) {
                (0, branch_handler_utils_1.processBranchForBuilder)(target, caseIndex, switchMainConns, targetNodeIds);
            }
            const existingNode = ctx.nodes.get(builder.switchNode.name);
            if (existingNode) {
                const existingMainConns = existingNode.connections.get('main') ?? new Map();
                for (const [outputIndex, targets] of switchMainConns) {
                    const existingTargets = existingMainConns.get(outputIndex) ?? [];
                    for (const target of targets) {
                        const alreadyExists = existingTargets.some((t) => t.node === target.node && t.index === target.index);
                        if (!alreadyExists) {
                            existingTargets.push(target);
                        }
                    }
                    existingMainConns.set(outputIndex, existingTargets);
                }
                existingNode.connections.set('main', existingMainConns);
            }
            else {
                const switchConns = new Map();
                switchConns.set('main', switchMainConns);
                ctx.nodes.set(builder.switchNode.name, {
                    instance: builder.switchNode,
                    connections: switchConns,
                });
            }
            for (const [, target] of builder.caseMapping) {
                (0, branch_handler_utils_1.addBranchTargetNodes)(target, ctx);
            }
            if (ctx.nameMapping) {
                (0, branch_handler_utils_1.fixupBranchConnectionTargets)(switchMainConns, targetNodeIds, ctx.nameMapping);
            }
            return builder.switchNode.name;
        }
        if ('cases' in input && Array.isArray(input.cases)) {
            input.cases.forEach((caseNode, index) => {
                (0, branch_handler_utils_1.processBranchForComposite)(caseNode, index, ctx, switchMainConns);
            });
        }
        const switchConns = new Map();
        switchConns.set('main', switchMainConns);
        ctx.nodes.set(input.switchNode.name, {
            instance: input.switchNode,
            connections: switchConns,
        });
        return input.switchNode.name;
    },
};
