// n8n-workflow/dist/cjs/extensions/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./expression-extension"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.ExpressionExtensions = exports.extendTransform = exports.hasNativeMethod = exports.hasExpressionExtension = exports.extendOptional = exports.extend = void 0;
    var expression_extension_1 = require("./expression-extension");
    Object.defineProperty(exports, "extend", { enumerable: true, get: function () { return expression_extension_1.extend; } });
    Object.defineProperty(exports, "extendOptional", { enumerable: true, get: function () { return expression_extension_1.extendOptional; } });
    Object.defineProperty(exports, "hasExpressionExtension", { enumerable: true, get: function () { return expression_extension_1.hasExpressionExtension; } });
    Object.defineProperty(exports, "hasNativeMethod", { enumerable: true, get: function () { return expression_extension_1.hasNativeMethod; } });
    Object.defineProperty(exports, "extendTransform", { enumerable: true, get: function () { return expression_extension_1.extendTransform; } });
    Object.defineProperty(exports, "ExpressionExtensions", { enumerable: true, get: function () { return expression_extension_1.EXTENSION_OBJECTS; } });
});
