// @n8n/workflow-sdk/dist/codegen/parse-workflow-code.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseWorkflowCode = parseWorkflowCode;
exports.parseWorkflowCodeToBuilder = parseWorkflowCodeToBuilder;
const ast_interpreter_1 = require("../ast-interpreter");
const expression_1 = require("../expression");
const workflow_builder_1 = require("../workflow-builder");
const next_batch_1 = require("../workflow-builder/control-flow-builders/next-batch");
const split_in_batches_1 = require("../workflow-builder/control-flow-builders/split-in-batches");
const node_builder_1 = require("../workflow-builder/node-builders/node-builder");
const subnode_builders_1 = require("../workflow-builder/node-builders/subnode-builders");
const N8N_RUNTIME_VARIABLES = [
    '$today',
    '$now',
    '$json',
    '$input',
    '$binary',
    '$execution',
    '$workflow',
    '$vars',
    '$node',
    '$item',
    '$position',
    '$runIndex',
    '$prevNode',
    '$env',
    '$itemIndex',
    '$parameter',
    '$jmespath',
    '$if',
    '$min',
    '$max',
];
function escapeN8nVariablesInTemplateLiterals(code) {
    let result = code;
    for (const varName of N8N_RUNTIME_VARIABLES) {
        const escapedVarName = varName.replace(/\$/g, '\\$');
        const pattern = new RegExp('(?<!\\\\)\\$\\{' + escapedVarName, 'g');
        result = result.replace(pattern, '\\${' + varName);
    }
    result = result.replace(/(?<!\\)\$\{\$\(/g, '\\${$(');
    result = result.replace(/(?<!\\)\$\{\{/g, '\\${{');
    return result;
}
function readDoubleQuotedString(code, start) {
    let result = '"';
    let i = start + 1;
    while (i < code.length) {
        if (code[i] === '\\' && i + 1 < code.length) {
            result += code[i] + code[i + 1];
            i += 2;
        }
        else if (code[i] === '"') {
            result += '"';
            i++;
            break;
        }
        else {
            result += code[i];
            i++;
        }
    }
    return [result, i];
}
function readTemplateLiteral(code, start) {
    let result = '`';
    let i = start + 1;
    let depth = 0;
    let inNestedTemplateLiteral = false;
    while (i < code.length) {
        if (code[i] === '\\' && i + 1 < code.length) {
            if (code[i + 1] === '`') {
                inNestedTemplateLiteral = !inNestedTemplateLiteral;
            }
            result += code[i] + code[i + 1];
            i += 2;
        }
        else if (code[i] === '$' && i + 1 < code.length && code[i + 1] === '{') {
            if (inNestedTemplateLiteral && depth === 0) {
                result += '\\${';
            }
            else {
                result += '${';
                depth++;
            }
            i += 2;
        }
        else if (code[i] === '}' && depth > 0) {
            result += '}';
            i++;
            depth--;
        }
        else if (code[i] === '`' && depth === 0) {
            result += '`';
            i++;
            break;
        }
        else {
            result += code[i];
            i++;
        }
    }
    return [result, i];
}
function readAndFixSingleQuotedString(code, start) {
    let result = "'";
    let i = start + 1;
    while (i < code.length) {
        if (code[i] === '\\' && i + 1 < code.length) {
            result += code[i] + code[i + 1];
            i += 2;
            continue;
        }
        if (code[i] === '$' &&
            code[i + 1] === '(' &&
            code[i + 2] === '\\' &&
            code[i + 3] === '\\' &&
            code[i + 4] === "'") {
            result += "$(\\'";
            i += 5;
            while (i < code.length) {
                if (code[i] === '\\' &&
                    code[i + 1] === '\\' &&
                    code[i + 2] === "'" &&
                    code[i + 3] === ')') {
                    result += "\\')";
                    i += 4;
                    break;
                }
                else if (code[i] === '\\' && i + 1 < code.length) {
                    result += code[i] + code[i + 1];
                    i += 2;
                }
                else {
                    result += code[i];
                    i++;
                }
            }
            continue;
        }
        if (code[i] === '$' && code[i + 1] === '(' && code[i + 2] === "'") {
            result += "$(\\'";
            i += 3;
            while (i < code.length) {
                if (code[i] === '\\' && i + 1 < code.length) {
                    result += code[i] + code[i + 1];
                    i += 2;
                }
                else if (code[i] === "'" && code[i + 1] === ')') {
                    result += "\\')";
                    i += 2;
                    break;
                }
                else {
                    result += code[i];
                    i++;
                }
            }
            continue;
        }
        if (code[i] === "'") {
            const prevChar = result.length > 1 ? result[result.length - 1] : '';
            const nextChar = i + 1 < code.length ? code[i + 1] : '';
            if (/[a-zA-Z]/.test(prevChar) && /[a-z]/.test(nextChar)) {
                result += "\\'";
                i++;
                continue;
            }
            result += "'";
            i++;
            break;
        }
        result += code[i];
        i++;
    }
    return [result, i];
}
function escapeNodeReferencesInSingleQuotedStrings(code) {
    let result = '';
    let i = 0;
    while (i < code.length) {
        if (code[i] === '"') {
            const [str, newI] = readDoubleQuotedString(code, i);
            result += str;
            i = newI;
            continue;
        }
        if (code[i] === '`') {
            const [str, newI] = readTemplateLiteral(code, i);
            result += str;
            i = newI;
            continue;
        }
        if (code[i] === "'") {
            const [str, newI] = readAndFixSingleQuotedString(code, i);
            result += str;
            i = newI;
            continue;
        }
        result += code[i];
        i++;
    }
    return result;
}
function escapeN8nVariables(code) {
    let result = escapeNodeReferencesInSingleQuotedStrings(code);
    result = escapeN8nVariablesInTemplateLiterals(result);
    return result;
}
function unescapeJsonEscapeSequences(code) {
    if (code.includes('\n') && !hasLiteralBackslashNOutsideStrings(code)) {
        return code;
    }
    if (!hasLiteralBackslashNOutsideStrings(code)) {
        return code;
    }
    return unescapeOutsideStrings(code);
}
function hasLiteralBackslashNOutsideStrings(code) {
    let i = 0;
    while (i < code.length) {
        const char = code[i];
        if (char === '"') {
            i++;
            while (i < code.length && code[i] !== '"') {
                if (code[i] === '\\' && i + 1 < code.length)
                    i += 2;
                else
                    i++;
            }
            i++;
            continue;
        }
        if (char === "'") {
            i++;
            while (i < code.length && code[i] !== "'") {
                if (code[i] === '\\' && i + 1 < code.length)
                    i += 2;
                else
                    i++;
            }
            i++;
            continue;
        }
        if (char === '`') {
            i++;
            let depth = 0;
            while (i < code.length) {
                if (code[i] === '\\' && i + 1 < code.length) {
                    i += 2;
                }
                else if (code[i] === '$' && code[i + 1] === '{') {
                    depth++;
                    i += 2;
                }
                else if (code[i] === '}' && depth > 0) {
                    depth--;
                    i++;
                }
                else if (code[i] === '`' && depth === 0) {
                    i++;
                    break;
                }
                else {
                    i++;
                }
            }
            continue;
        }
        if (char === '\\' && i + 1 < code.length && code[i + 1] === 'n') {
            return true;
        }
        i++;
    }
    return false;
}
function unescapeOutsideStrings(code) {
    let result = '';
    let i = 0;
    while (i < code.length) {
        const char = code[i];
        if (char === '"') {
            const start = i;
            i++;
            while (i < code.length && code[i] !== '"') {
                if (code[i] === '\\' && i + 1 < code.length)
                    i += 2;
                else
                    i++;
            }
            i++;
            result += code.slice(start, i);
            continue;
        }
        if (char === "'") {
            const start = i;
            i++;
            while (i < code.length && code[i] !== "'") {
                if (code[i] === '\\' && i + 1 < code.length)
                    i += 2;
                else
                    i++;
            }
            i++;
            result += code.slice(start, i);
            continue;
        }
        if (char === '`') {
            const start = i;
            i++;
            let depth = 0;
            while (i < code.length) {
                if (code[i] === '\\' && i + 1 < code.length) {
                    i += 2;
                }
                else if (code[i] === '$' && code[i + 1] === '{') {
                    depth++;
                    i += 2;
                }
                else if (code[i] === '}' && depth > 0) {
                    depth--;
                    i++;
                }
                else if (code[i] === '`' && depth === 0) {
                    i++;
                    break;
                }
                else {
                    i++;
                }
            }
            result += code.slice(start, i);
            continue;
        }
        if (char === '\\' && i + 1 < code.length) {
            const nextChar = code[i + 1];
            switch (nextChar) {
                case 'n':
                    result += '\n';
                    i += 2;
                    break;
                case 'r':
                    result += '\r';
                    i += 2;
                    break;
                case 't':
                    result += '\t';
                    i += 2;
                    break;
                case '\\':
                    result += '\\';
                    i += 2;
                    break;
                case '"':
                    result += '"';
                    i += 2;
                    break;
                default:
                    result += char;
                    i++;
            }
            continue;
        }
        result += char;
        i++;
    }
    return result;
}
const sdkFunctions = {
    workflow: workflow_builder_1.workflow,
    node: node_builder_1.node,
    trigger: node_builder_1.trigger,
    sticky: node_builder_1.sticky,
    placeholder: node_builder_1.placeholder,
    newCredential: node_builder_1.newCredential,
    ifElse: node_builder_1.ifElse,
    switchCase: node_builder_1.switchCase,
    merge: node_builder_1.merge,
    splitInBatches: split_in_batches_1.splitInBatches,
    nextBatch: next_batch_1.nextBatch,
    languageModel: subnode_builders_1.languageModel,
    memory: subnode_builders_1.memory,
    tool: subnode_builders_1.tool,
    outputParser: subnode_builders_1.outputParser,
    embedding: subnode_builders_1.embedding,
    embeddings: subnode_builders_1.embeddings,
    vectorStore: subnode_builders_1.vectorStore,
    retriever: subnode_builders_1.retriever,
    documentLoader: subnode_builders_1.documentLoader,
    textSplitter: subnode_builders_1.textSplitter,
    reranker: subnode_builders_1.reranker,
    fromAi: subnode_builders_1.fromAi,
    expr: expression_1.expr,
};
function parseWorkflowCode(code) {
    const unescapedCode = unescapeJsonEscapeSequences(code);
    const executableCode = escapeN8nVariables(unescapedCode);
    try {
        const wf = (0, ast_interpreter_1.interpretSDKCode)(executableCode, sdkFunctions);
        return wf.toJSON();
    }
    catch (error) {
        if (error instanceof ast_interpreter_1.SecurityError) {
            throw new SyntaxError(`Failed to parse workflow code: ${error.message}. ` +
                'This code contains patterns that are not allowed for security reasons.');
        }
        if (error instanceof ast_interpreter_1.InterpreterError) {
            if (error.message.includes('reserved SDK function name')) {
                throw new SyntaxError(`Failed to parse workflow code: ${error.message}`);
            }
            throw new SyntaxError(`Failed to parse workflow code: ${error.message}. ` +
                'Common causes include unclosed template literals, missing commas, or unbalanced brackets.');
        }
        throw error;
    }
}
function parseWorkflowCodeToBuilder(code) {
    const unescapedCode = unescapeJsonEscapeSequences(code);
    const executableCode = escapeN8nVariables(unescapedCode);
    try {
        return (0, ast_interpreter_1.interpretSDKCode)(executableCode, sdkFunctions);
    }
    catch (error) {
        if (error instanceof ast_interpreter_1.SecurityError) {
            throw new SyntaxError(`Failed to parse workflow code: ${error.message}. ` +
                'This code contains patterns that are not allowed for security reasons.');
        }
        if (error instanceof ast_interpreter_1.InterpreterError) {
            if (error.message.includes('reserved SDK function name')) {
                throw new SyntaxError(`Failed to parse workflow code: ${error.message}`);
            }
            throw new SyntaxError(`Failed to parse workflow code: ${error.message}. ` +
                'Common causes include unclosed template literals, missing commas, or unbalanced brackets.');
        }
        throw error;
    }
}
