// @n8n/workflow-sdk/dist/codegen/subnode-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatCredentials = formatCredentials;
exports.formatValue = formatValue;
exports.generateSubnodeCall = generateSubnodeCall;
exports.generateSubnodesConfig = generateSubnodesConfig;
const constants_1 = require("./constants");
const node_type_utils_1 = require("./node-type-utils");
const string_utils_1 = require("./string-utils");
function formatCredentials(credentials) {
    if (typeof credentials === 'string') {
        return `'${(0, string_utils_1.escapeString)(credentials)}'`;
    }
    const entries = Object.entries(credentials).map(([key, value]) => {
        if (value.name !== undefined || value.id !== undefined) {
            if (value.name !== undefined) {
                if (value.id !== undefined) {
                    return `${(0, string_utils_1.formatKey)(key)}: newCredential('${(0, string_utils_1.escapeString)(value.name)}', '${(0, string_utils_1.escapeString)(value.id)}')`;
                }
                return `${(0, string_utils_1.formatKey)(key)}: newCredential('${(0, string_utils_1.escapeString)(value.name)}')`;
            }
            if (value.id !== undefined) {
                return `${(0, string_utils_1.formatKey)(key)}: { id: '${(0, string_utils_1.escapeString)(value.id)}' }`;
            }
        }
        return `${(0, string_utils_1.formatKey)(key)}: ${formatValue(value)}`;
    });
    return `{ ${entries.join(', ')} }`;
}
function getVarName(nodeName, ctx) {
    return (ctx.nodeNameToVarName.get(nodeName) ?? nodeName.toLowerCase().replace(/[^a-zA-Z0-9]/g, '_'));
}
function containsExpressionAnnotation(value, ctx) {
    if (!ctx?.expressionAnnotations || ctx.expressionAnnotations.size === 0)
        return false;
    if (typeof value === 'string') {
        return ctx.expressionAnnotations.has(value);
    }
    if (Array.isArray(value)) {
        return value.some((v) => containsExpressionAnnotation(v, ctx));
    }
    if (typeof value === 'object' && value !== null) {
        return Object.values(value).some((v) => containsExpressionAnnotation(v, ctx));
    }
    return false;
}
function formatValue(value, ctx) {
    if (value === null)
        return 'null';
    if (value === undefined)
        return 'undefined';
    if (typeof value === 'string') {
        if ((0, string_utils_1.isPlaceholderValue)(value)) {
            const hint = (0, string_utils_1.extractPlaceholderHint)(value);
            return `placeholder('${(0, string_utils_1.escapeString)(hint)}')`;
        }
        if (value.startsWith('=')) {
            const inner = value.slice(1);
            const formatted = `expr('${(0, string_utils_1.escapeString)(inner)}')`;
            if (ctx?.expressionAnnotations?.has(value)) {
                const annotation = ctx.expressionAnnotations.get(value);
                return `/** @example ${annotation} */\n${formatted}`;
            }
            return formatted;
        }
        const formatted = `'${(0, string_utils_1.escapeString)(value)}'`;
        if (ctx?.expressionAnnotations?.has(value)) {
            const annotation = ctx.expressionAnnotations.get(value);
            return `/** @example ${annotation} */\n${formatted}`;
        }
        return formatted;
    }
    if (typeof value === 'number' || typeof value === 'boolean')
        return String(value);
    if (Array.isArray(value)) {
        const innerCtx = ctx ? { ...ctx, indent: (ctx.indent ?? 0) + 1 } : ctx;
        const formattedElements = value.map((v) => formatValue(v, innerCtx));
        if (containsExpressionAnnotation(value, ctx)) {
            const baseIndent = '  '.repeat(ctx?.indent ?? 0);
            const elementIndent = '  '.repeat((ctx?.indent ?? 0) + 1);
            return `[\n${formattedElements.map((e) => `${elementIndent}${e}`).join(',\n')}\n${baseIndent}]`;
        }
        return `[${formattedElements.join(', ')}]`;
    }
    if (typeof value === 'object') {
        const entries = Object.entries(value);
        if (entries.length === 0)
            return '{}';
        const innerCtx = ctx ? { ...ctx, indent: (ctx.indent ?? 0) + 1 } : ctx;
        const formattedEntries = entries.map(([k, v]) => ({
            key: (0, string_utils_1.formatKey)(k),
            value: formatValue(v, innerCtx),
        }));
        if (containsExpressionAnnotation(value, ctx)) {
            const baseIndent = '  '.repeat(ctx?.indent ?? 0);
            const propIndent = '  '.repeat((ctx?.indent ?? 0) + 1);
            const lines = formattedEntries.map((e, i) => {
                const isLast = i === formattedEntries.length - 1;
                const valueLines = e.value.split('\n');
                const isBlockCommentPrefixed = valueLines.length > 1 &&
                    valueLines[0].trim().startsWith('/**') &&
                    !valueLines[valueLines.length - 1].trim().match(/^[}\]]$/);
                if (isBlockCommentPrefixed) {
                    const commentLines = valueLines.slice(0, -1).map((l) => `${propIndent}${l}`);
                    const valueLine = `${propIndent}${e.key}: ${valueLines[valueLines.length - 1]}`;
                    return [...commentLines, isLast ? valueLine : `${valueLine},`].join('\n');
                }
                if (valueLines.length > 1) {
                    const indentedValue = valueLines
                        .map((line, lineIdx) => (lineIdx === 0 ? line : `${propIndent}${line}`))
                        .join('\n');
                    const propLine = `${propIndent}${e.key}: ${indentedValue}`;
                    return isLast ? propLine : `${propLine},`;
                }
                const propLine = `${propIndent}${e.key}: ${e.value}`;
                return isLast ? propLine : `${propLine},`;
            });
            return `{\n${lines.join('\n')}\n${baseIndent}}`;
        }
        return `{ ${formattedEntries.map((e) => `${e.key}: ${e.value}`).join(', ')} }`;
    }
    return String(value);
}
function generateSubnodeConfigParts(node, ctx, options) {
    const configParts = [];
    const defaultName = (0, node_type_utils_1.generateDefaultNodeName)(node.type);
    if (node.json.name && node.json.name !== defaultName) {
        configParts.push(`name: '${(0, string_utils_1.escapeString)(node.json.name)}'`);
    }
    if (node.json.parameters && Object.keys(node.json.parameters).length > 0) {
        configParts.push(`parameters: ${formatValue(node.json.parameters, ctx)}`);
    }
    if (node.json.credentials && Object.keys(node.json.credentials).length > 0) {
        configParts.push(`credentials: ${formatCredentials(node.json.credentials)}`);
    }
    const pos = node.json.position;
    if (pos && (pos[0] !== 0 || pos[1] !== 0)) {
        configParts.push(`position: [${pos[0]}, ${pos[1]}]`);
    }
    const nestedConfig = generateSubnodesConfig(node, ctx, options);
    if (nestedConfig) {
        configParts.push(`subnodes: ${nestedConfig}`);
    }
    return configParts;
}
function generateSubnodeCall(node, builderName, ctx, options) {
    const parts = [];
    parts.push(`type: '${node.type}'`);
    parts.push(`version: ${node.json.typeVersion}`);
    const configParts = generateSubnodeConfigParts(node, ctx, options);
    if (configParts.length > 0) {
        parts.push(`config: { ${configParts.join(', ')} }`);
    }
    else {
        parts.push('config: {}');
    }
    return `${builderName}({ ${parts.join(', ')} })`;
}
function generateSubnodesConfig(node, ctx, options) {
    if (node.subnodes.length === 0) {
        return null;
    }
    if (options.useVarRefs) {
        return generateSubnodesConfigWithVarRefs(node, ctx);
    }
    else {
        return generateSubnodesConfigInline(node, ctx, options);
    }
}
function generateSubnodesConfigInline(node, ctx, options) {
    const grouped = new Map();
    for (const sub of node.subnodes) {
        const subnodeNode = ctx.graph.nodes.get(sub.subnodeName);
        if (!subnodeNode)
            continue;
        const existing = grouped.get(sub.connectionType) ?? [];
        existing.push({ node: subnodeNode, index: sub.index });
        grouped.set(sub.connectionType, existing);
    }
    const entries = [];
    for (const [connType, subnodeEntries] of grouped) {
        const configKey = constants_1.AI_CONNECTION_TO_CONFIG_KEY[connType];
        const builderName = constants_1.AI_CONNECTION_TO_BUILDER[connType];
        if (subnodeEntries.length === 0)
            continue;
        const calls = subnodeEntries.map((e) => generateSubnodeCall(e.node, builderName, ctx, options));
        if (constants_1.AI_ALWAYS_ARRAY_TYPES.has(connType)) {
            entries.push(`${configKey}: [${calls.join(', ')}]`);
        }
        else if (constants_1.AI_OPTIONAL_ARRAY_TYPES.has(connType)) {
            if (subnodeEntries.length === 1) {
                entries.push(`${configKey}: ${calls[0]}`);
            }
            else if (connType === 'ai_languageModel' &&
                subnodeEntries.every((e) => e.index === subnodeEntries[0].index)) {
                entries.push(`${configKey}: [[${calls.join(', ')}]]`);
            }
            else {
                entries.push(`${configKey}: [${calls.join(', ')}]`);
            }
        }
        else {
            entries.push(`${configKey}: ${calls[0]}`);
        }
    }
    if (entries.length === 0) {
        return null;
    }
    return `{ ${entries.join(', ')} }`;
}
function generateSubnodesConfigWithVarRefs(node, ctx) {
    const grouped = new Map();
    for (const sub of node.subnodes) {
        const varName = getVarName(sub.subnodeName, ctx);
        const existing = grouped.get(sub.connectionType) ?? [];
        existing.push({ varName, index: sub.index });
        grouped.set(sub.connectionType, existing);
    }
    const entries = [];
    for (const [connType, varEntries] of grouped) {
        const configKey = constants_1.AI_CONNECTION_TO_CONFIG_KEY[connType];
        if (varEntries.length === 0)
            continue;
        const varNames = varEntries.map((e) => e.varName);
        if (constants_1.AI_ALWAYS_ARRAY_TYPES.has(connType)) {
            entries.push(`${configKey}: [${varNames.join(', ')}]`);
        }
        else if (constants_1.AI_OPTIONAL_ARRAY_TYPES.has(connType)) {
            if (varEntries.length === 1) {
                entries.push(`${configKey}: ${varNames[0]}`);
            }
            else if (connType === 'ai_languageModel' &&
                varEntries.every((e) => e.index === varEntries[0].index)) {
                entries.push(`${configKey}: [[${varNames.join(', ')}]]`);
            }
            else {
                entries.push(`${configKey}: [${varNames.join(', ')}]`);
            }
        }
        else {
            entries.push(`${configKey}: ${varNames[0]}`);
        }
    }
    if (entries.length === 0) {
        return null;
    }
    return `{ ${entries.join(', ')} }`;
}
