// @n8n/workflow-sdk/dist/workflow-builder/plugins/validators/disconnected-node-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.disconnectedNodeValidator = void 0;
const node_types_1 = require("../../../constants/node-types");
const base_1 = require("../../../types/base");
const trigger_detection_1 = require("../../../utils/trigger-detection");
const type_guards_1 = require("../../type-guards");
const types_1 = require("../types");
const AI_CONNECTION_TYPES = [
    'ai_languageModel',
    'ai_memory',
    'ai_tool',
    'ai_outputParser',
    'ai_embedding',
    'ai_vectorStore',
    'ai_retriever',
    'ai_document',
    'ai_textSplitter',
    'ai_reranker',
];
function isConnectedSubnode(graphNode) {
    for (const [connType, outputMap] of graphNode.connections) {
        if (AI_CONNECTION_TYPES.includes(connType)) {
            for (const [_outputIndex, targets] of outputMap) {
                if (targets.length > 0) {
                    return true;
                }
            }
        }
    }
    return false;
}
function getTargetNodeName(target) {
    if (target === null || target === undefined) {
        return undefined;
    }
    if ((0, type_guards_1.isIfElseComposite)(target)) {
        const ifElse = target;
        return ifElse.ifNode.name;
    }
    if ((0, type_guards_1.isSwitchCaseComposite)(target) ||
        (typeof target === 'object' && 'switchNode' in target && 'caseMapping' in target)) {
        const switchCase = target;
        return switchCase.switchNode.name;
    }
    if ((0, type_guards_1.isSplitInBatchesBuilder)(target)) {
        const sib = target;
        return sib.sibNode.name;
    }
    if (typeof target === 'object' && 'name' in target) {
        return target.name;
    }
    if (typeof target === 'string') {
        return target;
    }
    if (typeof target === 'number' || typeof target === 'boolean') {
        return String(target);
    }
    return undefined;
}
function findNodesWithIncomingConnections(ctx, registry) {
    const nodesWithIncoming = new Set();
    for (const [_name, graphNode] of ctx.nodes) {
        const mainConns = graphNode.connections.get('main');
        if (mainConns) {
            for (const [_outputIndex, targets] of mainConns) {
                for (const target of targets) {
                    if (typeof target === 'object' && 'node' in target) {
                        nodesWithIncoming.add(target.node);
                    }
                }
            }
        }
        if (typeof graphNode.instance.getConnections === 'function') {
            const connections = graphNode.instance.getConnections();
            for (const conn of connections) {
                if ((0, base_1.isNodeChain)(conn.target)) {
                    nodesWithIncoming.add(conn.target.head.name);
                }
                else if (registry) {
                    const compositeHeadName = registry.resolveCompositeHeadName(conn.target);
                    if (compositeHeadName !== undefined) {
                        nodesWithIncoming.add(compositeHeadName);
                    }
                    else {
                        const nodeName = getTargetNodeName(conn.target);
                        if (nodeName !== undefined) {
                            nodesWithIncoming.add(nodeName);
                        }
                    }
                }
                else {
                    const nodeName = getTargetNodeName(conn.target);
                    if (nodeName !== undefined) {
                        nodesWithIncoming.add(nodeName);
                    }
                }
            }
        }
    }
    return nodesWithIncoming;
}
exports.disconnectedNodeValidator = {
    id: 'core:disconnected-node',
    name: 'Disconnected Node Validator',
    priority: 10,
    validateNode: () => [],
    validateWorkflow(ctx) {
        if (ctx.validationOptions?.allowDisconnectedNodes) {
            return [];
        }
        const issues = [];
        const nodesWithIncoming = findNodesWithIncomingConnections(ctx);
        for (const [mapKey, graphNode] of ctx.nodes) {
            const originalName = graphNode.instance.name;
            if ((0, trigger_detection_1.isTriggerNodeType)(graphNode.instance.type)) {
                continue;
            }
            if ((0, node_types_1.isStickyNoteType)(graphNode.instance.type)) {
                continue;
            }
            if (isConnectedSubnode(graphNode)) {
                continue;
            }
            if (!nodesWithIncoming.has(mapKey)) {
                const renamed = (0, types_1.isAutoRenamed)(mapKey, originalName);
                const displayName = renamed ? mapKey : originalName;
                const origForWarning = renamed ? originalName : undefined;
                const nodeRef = (0, types_1.formatNodeRef)(displayName, origForWarning, graphNode.instance.type);
                issues.push({
                    code: 'DISCONNECTED_NODE',
                    message: `${nodeRef} is not connected to any input. It will not receive data.`,
                    severity: 'warning',
                    violationLevel: 'major',
                    nodeName: displayName,
                    originalName: origForWarning,
                });
            }
        }
        return issues;
    },
};
