// @n8n/workflow-sdk/dist/workflow-builder/plugins/validators/expression-path-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.expressionPathValidator = void 0;
const base_1 = require("../../../types/base");
const string_utils_1 = require("../../string-utils");
const validation_helpers_1 = require("../../validation-helpers");
function resolveTargetNodeName(target) {
    if (!target)
        return undefined;
    if (typeof target === 'object' &&
        'node' in target &&
        typeof target.node === 'object') {
        const nodeTarget = target.node;
        return nodeTarget?.name;
    }
    if ((0, base_1.isNodeChain)(target)) {
        return target.head.name;
    }
    if (typeof target === 'object' && 'name' in target) {
        return target.name;
    }
    return undefined;
}
function findPredecessors(nodeName, nodes) {
    const predecessors = [];
    for (const [sourceNodeName, graphNode] of nodes) {
        const mainConns = graphNode.connections.get('main');
        if (mainConns) {
            for (const [_outputIndex, targets] of mainConns) {
                for (const target of targets) {
                    if (target.node === nodeName) {
                        predecessors.push(sourceNodeName);
                    }
                }
            }
        }
        if (typeof graphNode.instance.getConnections === 'function') {
            const connections = graphNode.instance.getConnections();
            for (const conn of connections) {
                const targetName = resolveTargetNodeName(conn.target);
                if (targetName === nodeName) {
                    predecessors.push(sourceNodeName);
                }
            }
        }
    }
    return [...new Set(predecessors)];
}
function validateJsonPath(nodeName, paramPath, fieldPath, predecessors, outputShapes, issues) {
    const validIn = [];
    const invalidIn = [];
    for (const pred of predecessors) {
        const shape = outputShapes.get(pred);
        if (shape) {
            if ((0, validation_helpers_1.hasPath)(shape, fieldPath)) {
                validIn.push(pred);
            }
            else {
                invalidIn.push(pred);
            }
        }
    }
    const fieldPathStr = fieldPath.join('.');
    if (invalidIn.length > 0 && validIn.length === 0) {
        issues.push({
            code: 'INVALID_EXPRESSION_PATH',
            message: `'${nodeName}' parameter '${paramPath}' uses $json.${fieldPathStr} but no predecessor outputs this field.`,
            severity: 'warning',
            nodeName,
            parameterPath: paramPath,
        });
    }
    else if (invalidIn.length > 0 && validIn.length > 0) {
        issues.push({
            code: 'PARTIAL_EXPRESSION_PATH',
            message: `'${nodeName}' parameter '${paramPath}' uses $json.${fieldPathStr} - exists in [${validIn.join(', ')}] but NOT in [${invalidIn.join(', ')}].`,
            severity: 'warning',
            nodeName,
            parameterPath: paramPath,
        });
    }
}
function validateNodePath(nodeName, paramPath, referencedNode, fieldPath, outputShapes, issues) {
    const shape = outputShapes.get(referencedNode);
    if (shape && !(0, validation_helpers_1.hasPath)(shape, fieldPath)) {
        issues.push({
            code: 'INVALID_EXPRESSION_PATH',
            message: `'${nodeName}' parameter '${paramPath}' uses $('${referencedNode}').item.json.${fieldPath.join('.')} but '${referencedNode}' doesn't output this field.`,
            severity: 'warning',
            nodeName,
            parameterPath: paramPath,
        });
    }
}
function unwrapItemJson(item) {
    if ('json' in item && typeof item.json === 'object' && item.json !== null) {
        return item.json;
    }
    return item;
}
exports.expressionPathValidator = {
    id: 'core:expression-path',
    name: 'Expression Path Validator',
    priority: 20,
    validateNode(_node, _graphNode, _ctx) {
        return [];
    },
    validateWorkflow(ctx) {
        const issues = [];
        const outputShapes = new Map();
        for (const [mapKey, graphNode] of ctx.nodes) {
            const output = graphNode.instance.config?.output;
            if (output && output.length > 0) {
                outputShapes.set(mapKey, unwrapItemJson(output[0]));
            }
        }
        for (const [mapKey, graphNode] of ctx.nodes) {
            if (!outputShapes.has(mapKey)) {
                const nodePinData = graphNode.instance.config?.pinData;
                if (nodePinData && nodePinData.length > 0) {
                    outputShapes.set(mapKey, unwrapItemJson(nodePinData[0]));
                }
            }
        }
        if (ctx.pinData) {
            for (const [nodeName, pinData] of Object.entries(ctx.pinData)) {
                if (!outputShapes.has(nodeName) && pinData.length > 0) {
                    outputShapes.set(nodeName, unwrapItemJson(pinData[0]));
                }
            }
        }
        if (outputShapes.size === 0) {
            return issues;
        }
        for (const [mapKey, graphNode] of ctx.nodes) {
            const params = graphNode.instance.config?.parameters;
            if (!params)
                continue;
            const expressions = (0, validation_helpers_1.extractExpressions)(params);
            const predecessors = findPredecessors(mapKey, ctx.nodes);
            for (const { expression, path } of expressions) {
                const parsed = (0, validation_helpers_1.parseExpression)(expression);
                const filteredFieldPath = (0, string_utils_1.filterMethodsFromPath)(parsed.fieldPath);
                if (parsed.type === '$json' && filteredFieldPath.length > 0) {
                    validateJsonPath(mapKey, path, filteredFieldPath, predecessors, outputShapes, issues);
                }
                if (parsed.type === '$node' && parsed.nodeName && filteredFieldPath.length > 0) {
                    validateNodePath(mapKey, path, parsed.nodeName, filteredFieldPath, outputShapes, issues);
                }
            }
        }
        return issues;
    },
};
