// @n8n/workflow-sdk/dist/workflow-builder/plugins/validators/expression-prefix-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.expressionPrefixValidator = void 0;
const node_types_1 = require("../../../constants/node-types");
const validation_helpers_1 = require("../../validation-helpers");
exports.expressionPrefixValidator = {
    id: 'core:expression-prefix',
    name: 'Expression Prefix Validator',
    priority: 30,
    validateNode(node, _graphNode, _ctx) {
        const issues = [];
        if ((0, node_types_1.isStickyNoteType)(node.type)) {
            return issues;
        }
        if (node.type === 'n8n-nodes-base.html') {
            return issues;
        }
        const params = node.config?.parameters;
        if (!params) {
            return issues;
        }
        const prefixIssues = (0, validation_helpers_1.findMissingExpressionPrefixes)(params);
        for (const { path } of prefixIssues) {
            issues.push({
                code: 'MISSING_EXPRESSION_PREFIX',
                message: `'${node.name}' has parameter "${path}" containing {{ $... }} without '=' prefix. n8n expressions must start with '=' like '={{ $json.field }}'.`,
                severity: 'warning',
                nodeName: node.name,
                parameterPath: path,
            });
        }
        return issues;
    },
};
