// @n8n/workflow-sdk/dist/workflow-builder/plugins/validators/merge-node-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeNodeValidator = void 0;
const node_builder_1 = require("../../node-builders/node-builder");
const types_1 = require("../types");
exports.mergeNodeValidator = {
    id: 'core:merge-node',
    name: 'Merge Node Validator',
    nodeTypes: ['n8n-nodes-base.merge'],
    priority: 40,
    validateNode(node, graphNode, ctx) {
        const issues = [];
        const mapKey = (0, types_1.findMapKey)(graphNode, ctx);
        const originalName = node.name;
        const connectedInputIndices = new Set();
        for (const [_name, otherNode] of ctx.nodes) {
            const mainConns = otherNode.connections.get('main');
            if (mainConns) {
                for (const [_outputIndex, targets] of mainConns) {
                    for (const target of targets) {
                        if (target.node === mapKey) {
                            connectedInputIndices.add(target.index);
                        }
                    }
                }
            }
            if (typeof otherNode.instance.getConnections === 'function') {
                const conns = otherNode.instance.getConnections();
                for (const conn of conns) {
                    const targetNode = (0, node_builder_1.isInputTarget)(conn.target) ? conn.target.node : conn.target;
                    const targetNodeName = typeof targetNode === 'object' && 'name' in targetNode ? targetNode.name : undefined;
                    if (targetNode === node || targetNodeName === originalName) {
                        const targetIndex = (0, node_builder_1.isInputTarget)(conn.target)
                            ? conn.target.inputIndex
                            : (conn.targetInputIndex ?? 0);
                        connectedInputIndices.add(targetIndex);
                    }
                }
            }
        }
        const inputCount = connectedInputIndices.size;
        if (inputCount < 2) {
            issues.push({
                code: 'MERGE_SINGLE_INPUT',
                message: `'${mapKey}' has only ${inputCount} input connection(s). Merge nodes require at least 2 inputs.`,
                severity: 'warning',
                violationLevel: 'major',
                nodeName: mapKey,
            });
        }
        return issues;
    },
};
