// @n8n/workflow-sdk/dist/workflow-builder/plugins/validators/tool-node-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toolNodeValidator = void 0;
const validation_helpers_1 = require("../../validation-helpers");
exports.toolNodeValidator = {
    id: 'core:tool-node',
    name: 'Tool Node Validator',
    priority: 50,
    validateNode(node, _graphNode, ctx) {
        const issues = [];
        if (!(0, validation_helpers_1.isToolNode)(node.type)) {
            return issues;
        }
        const provider = ctx.validationOptions?.nodeTypesProvider;
        if (provider) {
            const nodeType = provider.getByNameAndVersion(node.type, Number(node.version));
            const properties = nodeType?.description?.properties;
            if (properties !== undefined && properties.length === 0) {
                return issues;
            }
        }
        if (validation_helpers_1.TOOLS_WITHOUT_PARAMETERS.has(node.type)) {
            return issues;
        }
        const params = node.config?.parameters;
        if (!params || Object.keys(params).length === 0) {
            issues.push({
                code: 'TOOL_NO_PARAMETERS',
                message: `'${node.name}' has no parameters set.`,
                severity: 'warning',
                nodeName: node.name,
            });
        }
        return issues;
    },
};
