// n8n-core/dist/execution-engine/node-execution-context/utils/file-system-helper-functions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFileSystemHelperFunctions = void 0;
const backend_common_1 = require("@n8n/backend-common");
const config_1 = require("@n8n/config");
const di_1 = require("@n8n/di");
const n8n_workflow_1 = require("n8n-workflow");
const node_fs_1 = require("node:fs");
const promises_1 = require("node:fs/promises");
const node_os_1 = require("node:os");
const node_path_1 = require("node:path");
const constants_1 = require("../../../constants");
const instance_settings_1 = require("../../../instance-settings");
const getAllowedPaths = () => {
    const { restrictFileAccessTo } = di_1.Container.get(config_1.SecurityConfig);
    if (restrictFileAccessTo === '')
        return [];
    const allowedPaths = restrictFileAccessTo
        .split(';')
        .map((path) => path.trim())
        .filter((path) => path)
        .map((path) => (path.startsWith('~') ? path.replace('~', (0, node_os_1.homedir)()) : path));
    return allowedPaths;
};
async function resolvePath(path) {
    const pathStr = path.toString();
    try {
        return (await (0, promises_1.realpath)(pathStr));
    }
    catch (error) {
        if (error instanceof Error && 'code' in error && error.code === 'ENOENT') {
            const dir = (0, node_path_1.dirname)(pathStr);
            const file = (0, node_path_1.basename)(pathStr);
            const resolvedDir = await (0, promises_1.realpath)(dir);
            return (0, node_path_1.join)(resolvedDir, file);
        }
        throw error;
    }
}
function isFilePatternBlocked(resolvedFilePath) {
    const { blockFilePatterns } = di_1.Container.get(config_1.SecurityConfig);
    const normalizedPath = node_path_1.posix.normalize(resolvedFilePath.replace(/\\/g, '/'));
    return blockFilePatterns
        .split(';')
        .map((pattern) => pattern.trim())
        .filter((pattern) => pattern)
        .some((pattern) => {
        try {
            return new RegExp(pattern, 'mi').test(normalizedPath);
        }
        catch {
            return true;
        }
    });
}
function isFilePathBlocked(resolvedFilePath) {
    const allowedPaths = getAllowedPaths();
    const blockFileAccessToN8nFiles = process.env[constants_1.BLOCK_FILE_ACCESS_TO_N8N_FILES] !== 'false';
    const restrictedPaths = blockFileAccessToN8nFiles ? getN8nRestrictedPaths() : [];
    if (restrictedPaths.some((restrictedPath) => (0, backend_common_1.isContainedWithin)(restrictedPath, resolvedFilePath))) {
        return true;
    }
    if (isFilePatternBlocked(resolvedFilePath)) {
        return true;
    }
    if (allowedPaths.length) {
        return !allowedPaths.some((allowedPath) => (0, backend_common_1.isContainedWithin)(allowedPath, resolvedFilePath));
    }
    return false;
}
const getFileSystemHelperFunctions = (node) => ({
    async createReadStream(resolvedFilePath) {
        const pathIdentity = await (0, promises_1.stat)(resolvedFilePath);
        if (isFilePathBlocked(resolvedFilePath)) {
            const allowedPaths = getAllowedPaths();
            const message = allowedPaths.length ? ` Allowed paths: ${allowedPaths.join(', ')}` : '';
            throw new n8n_workflow_1.NodeOperationError(node, `Access to the file is not allowed.${message}`, {
                level: 'warning',
            });
        }
        try {
            await (0, promises_1.access)(resolvedFilePath);
        }
        catch (error) {
            throw error.code === 'ENOENT'
                ?
                    new n8n_workflow_1.NodeOperationError(node, error, {
                        message: `The file "${String(resolvedFilePath)}" could not be accessed.`,
                        level: 'warning',
                    })
                : error;
        }
        let fileHandle;
        try {
            fileHandle = await (0, promises_1.open)(resolvedFilePath, node_fs_1.constants.O_RDONLY | node_fs_1.constants.O_NOFOLLOW);
        }
        catch (error) {
            if ('code' in error && error.code === 'ELOOP') {
                throw new n8n_workflow_1.NodeOperationError(node, error instanceof Error ? error : '', {
                    message: 'Symlinks are not allowed.',
                    level: 'warning',
                });
            }
            else {
                throw error;
            }
        }
        try {
            const fileHandleIdentity = await fileHandle.stat();
            if (fileHandleIdentity.dev !== pathIdentity.dev ||
                fileHandleIdentity.ino !== pathIdentity.ino) {
                throw new n8n_workflow_1.NodeOperationError(node, 'The file has changed and cannot be accessed.', {
                    level: 'warning',
                });
            }
            return fileHandle.createReadStream();
        }
        catch (error) {
            await fileHandle.close();
            throw error;
        }
    },
    getStoragePath() {
        return (0, backend_common_1.safeJoinPath)(di_1.Container.get(instance_settings_1.InstanceSettings).n8nFolder, `storage/${node.type}`);
    },
    async writeContentToFile(resolvedFilePath, content, flag) {
        let pathIdentity;
        let fileExists = true;
        try {
            pathIdentity = await (0, promises_1.stat)(resolvedFilePath);
        }
        catch (error) {
            if ('code' in error && error.code === 'ENOENT') {
                fileExists = false;
            }
            else {
                throw error;
            }
        }
        if (isFilePathBlocked(resolvedFilePath)) {
            throw new n8n_workflow_1.NodeOperationError(node, `The file "${String(resolvedFilePath)}" is not writable.`, {
                level: 'warning',
            });
        }
        const shouldTruncate = flag === undefined || (flag & node_fs_1.constants.O_TRUNC) === node_fs_1.constants.O_TRUNC;
        const userFlags = flag ?? 0;
        const openFlags = node_fs_1.constants.O_WRONLY |
            node_fs_1.constants.O_CREAT |
            node_fs_1.constants.O_NOFOLLOW |
            (userFlags & ~node_fs_1.constants.O_TRUNC);
        let fileHandle;
        try {
            fileHandle = await (0, promises_1.open)(resolvedFilePath, openFlags);
        }
        catch (error) {
            if ('code' in error && error.code === 'ELOOP') {
                throw new n8n_workflow_1.NodeOperationError(node, error instanceof Error ? error : '', {
                    message: 'Symlinks are not allowed.',
                    level: 'warning',
                });
            }
            else {
                throw error;
            }
        }
        try {
            const fileHandleIdentity = await fileHandle.stat();
            if (fileExists && pathIdentity) {
                if (fileHandleIdentity.dev !== pathIdentity.dev ||
                    fileHandleIdentity.ino !== pathIdentity.ino) {
                    throw new n8n_workflow_1.NodeOperationError(node, 'The file has changed and cannot be written.', {
                        level: 'warning',
                    });
                }
            }
            else {
                pathIdentity = await (0, promises_1.stat)(resolvedFilePath);
                if (fileHandleIdentity.dev !== pathIdentity.dev ||
                    fileHandleIdentity.ino !== pathIdentity.ino) {
                    throw new n8n_workflow_1.NodeOperationError(node, 'The file was created but its identity does not match and cannot be written.', {
                        level: 'warning',
                    });
                }
            }
            if (!fileHandleIdentity.isFile()) {
                throw new n8n_workflow_1.NodeOperationError(node, 'The path is not a regular file.', {
                    level: 'warning',
                });
            }
            if (shouldTruncate) {
                await fileHandle.truncate(0);
            }
            if (typeof content === 'string' || Buffer.isBuffer(content)) {
                await fileHandle.writeFile(content, { encoding: 'binary' });
            }
            else {
                const writeStream = fileHandle.createWriteStream({ encoding: 'binary' });
                await new Promise((resolve, reject) => {
                    content.pipe(writeStream);
                    writeStream.on('finish', resolve);
                    writeStream.on('error', reject);
                });
            }
        }
        finally {
            await fileHandle.close();
        }
    },
    resolvePath,
    isFilePathBlocked,
});
exports.getFileSystemHelperFunctions = getFileSystemHelperFunctions;
function getN8nRestrictedPaths() {
    const { n8nFolder, staticCacheDir } = di_1.Container.get(instance_settings_1.InstanceSettings);
    const restrictedPaths = [n8nFolder, staticCacheDir];
    if (process.env[constants_1.CONFIG_FILES]) {
        restrictedPaths.push(...process.env[constants_1.CONFIG_FILES].split(','));
    }
    if (process.env[constants_1.CUSTOM_EXTENSION_ENV]) {
        const customExtensionFolders = process.env[constants_1.CUSTOM_EXTENSION_ENV].split(';');
        restrictedPaths.push(...customExtensionFolders);
    }
    if (process.env[constants_1.BINARY_DATA_STORAGE_PATH]) {
        restrictedPaths.push(process.env[constants_1.BINARY_DATA_STORAGE_PATH]);
    }
    if (process.env[constants_1.UM_EMAIL_TEMPLATES_INVITE]) {
        restrictedPaths.push(process.env[constants_1.UM_EMAIL_TEMPLATES_INVITE]);
    }
    if (process.env[constants_1.UM_EMAIL_TEMPLATES_PWRESET]) {
        restrictedPaths.push(process.env[constants_1.UM_EMAIL_TEMPLATES_PWRESET]);
    }
    return restrictedPaths;
}
