// n8n-core/dist/execution-engine/node-execution-context/utils/get-additional-keys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAdditionalKeys = getAdditionalKeys;
exports.getNonWorkflowAdditionalKeys = getNonWorkflowAdditionalKeys;
const constants_1 = require("../../../constants");
const custom_data_1 = require("./custom-data");
const get_secrets_proxy_1 = require("./get-secrets-proxy");
function appendResumeToken(url, token) {
    const urlObj = new URL(url);
    urlObj.searchParams.set(constants_1.WAITING_TOKEN_QUERY_PARAM, token);
    return urlObj.toString();
}
function getAdditionalKeys(additionalData, mode, runExecutionData) {
    const executionId = additionalData.executionId ?? constants_1.PLACEHOLDER_EMPTY_EXECUTION_ID;
    let resumeUrl = `${additionalData.webhookWaitingBaseUrl}/${executionId}`;
    let resumeFormUrl = `${additionalData.formWaitingBaseUrl}/${executionId}`;
    if (runExecutionData?.resumeToken) {
        resumeUrl = appendResumeToken(resumeUrl, runExecutionData.resumeToken);
        resumeFormUrl = appendResumeToken(resumeFormUrl, runExecutionData.resumeToken);
    }
    return {
        $execution: {
            id: executionId,
            mode: mode === 'manual' ? 'test' : 'production',
            resumeUrl,
            resumeFormUrl,
            customData: runExecutionData
                ? (0, custom_data_1.createExecutionCustomData)({ runExecutionData, mode })
                : undefined,
        },
        $vars: additionalData.variables,
        $secrets: (0, get_secrets_proxy_1.getSecretsProxy)(additionalData),
        $executionId: executionId,
        $resumeWebhookUrl: resumeUrl,
    };
}
function getNonWorkflowAdditionalKeys(additionalData) {
    return {
        $vars: additionalData.variables,
        $secrets: (0, get_secrets_proxy_1.getSecretsProxy)(additionalData),
    };
}
