// @n8n/permissions/dist/constants.ee.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EXTERNAL_SECRETS_SYSTEM_ROLES_ENABLED_SETTING = exports.PERSONAL_SPACE_SHARING_SETTING = exports.PERSONAL_SPACE_PUBLISHING_SETTING = exports.PROJECT_CHAT_USER_ROLE_SLUG = exports.PROJECT_VIEWER_ROLE_SLUG = exports.PROJECT_EDITOR_ROLE_SLUG = exports.PROJECT_ADMIN_ROLE_SLUG = exports.PROJECT_OWNER_ROLE_SLUG = exports.API_KEY_RESOURCES = exports.RESOURCES = exports.DEFAULT_OPERATIONS = void 0;
exports.DEFAULT_OPERATIONS = ['create', 'read', 'update', 'delete', 'list'];
exports.RESOURCES = {
    aiAssistant: ['manage'],
    annotationTag: [...exports.DEFAULT_OPERATIONS],
    auditLogs: ['manage'],
    banner: ['dismiss'],
    community: ['register'],
    communityPackage: ['install', 'uninstall', 'update', 'list', 'manage'],
    credential: ['share', 'unshare', 'shareGlobally', 'move', ...exports.DEFAULT_OPERATIONS],
    externalSecretsProvider: ['sync', ...exports.DEFAULT_OPERATIONS],
    externalSecret: ['list'],
    eventBusDestination: ['test', ...exports.DEFAULT_OPERATIONS],
    ldap: ['sync', 'manage'],
    license: ['manage'],
    logStreaming: ['manage'],
    orchestration: ['read', 'list'],
    project: [...exports.DEFAULT_OPERATIONS],
    saml: ['manage'],
    securityAudit: ['generate'],
    securitySettings: ['manage'],
    sourceControl: ['pull', 'push', 'manage'],
    tag: [...exports.DEFAULT_OPERATIONS],
    user: [
        'resetPassword',
        'changeRole',
        'enforceMfa',
        'generateInviteLink',
        ...exports.DEFAULT_OPERATIONS,
    ],
    variable: [...exports.DEFAULT_OPERATIONS],
    projectVariable: [...exports.DEFAULT_OPERATIONS],
    workersView: ['manage'],
    workflow: [
        'share',
        'unshare',
        'execute',
        'execute-chat',
        'move',
        'activate',
        'deactivate',
        'publish',
        'unpublish',
        'updateRedactionSetting',
        ...exports.DEFAULT_OPERATIONS,
    ],
    folder: [...exports.DEFAULT_OPERATIONS, 'move'],
    insights: ['list', 'read'],
    oidc: ['manage'],
    provisioning: ['manage'],
    dataTable: [...exports.DEFAULT_OPERATIONS, 'readRow', 'writeRow', 'listProject'],
    execution: ['delete', 'read', 'retry', 'list', 'get', 'reveal'],
    workflowTags: ['update', 'list'],
    role: ['manage'],
    mcp: ['manage', 'oauth'],
    mcpApiKey: ['create', 'rotate'],
    chatHub: ['manage', 'message'],
    chatHubAgent: [...exports.DEFAULT_OPERATIONS],
    breakingChanges: ['list'],
    apiKey: ['manage'],
    credentialResolver: [...exports.DEFAULT_OPERATIONS],
    instanceAi: ['message', 'manage', 'gateway'],
    roleMappingRule: [...exports.DEFAULT_OPERATIONS],
};
exports.API_KEY_RESOURCES = {
    tag: [...exports.DEFAULT_OPERATIONS],
    workflow: [...exports.DEFAULT_OPERATIONS, 'move', 'activate', 'deactivate'],
    variable: ['create', 'update', 'delete', 'list'],
    securityAudit: ['generate'],
    project: ['create', 'update', 'delete', 'list'],
    user: ['read', 'list', 'create', 'changeRole', 'delete', 'enforceMfa'],
    execution: ['delete', 'read', 'retry', 'list', 'get', 'stop'],
    credential: ['create', 'update', 'move', 'delete', 'list'],
    sourceControl: ['pull'],
    workflowTags: ['update', 'list'],
    executionTags: ['update', 'list'],
    communityPackage: ['install', 'uninstall', 'update', 'list'],
    dataTable: ['create', 'read', 'update', 'delete', 'list'],
    dataTableRow: ['create', 'read', 'update', 'delete', 'upsert'],
    insights: ['read'],
};
exports.PROJECT_OWNER_ROLE_SLUG = 'project:personalOwner';
exports.PROJECT_ADMIN_ROLE_SLUG = 'project:admin';
exports.PROJECT_EDITOR_ROLE_SLUG = 'project:editor';
exports.PROJECT_VIEWER_ROLE_SLUG = 'project:viewer';
exports.PROJECT_CHAT_USER_ROLE_SLUG = 'project:chatUser';
exports.PERSONAL_SPACE_PUBLISHING_SETTING = {
    key: 'security.personalSpacePublishing',
    scopes: ['workflow:publish'],
};
exports.PERSONAL_SPACE_SHARING_SETTING = {
    key: 'security.personalSpaceSharing',
    scopes: ['workflow:share', 'credential:share'],
};
exports.EXTERNAL_SECRETS_SYSTEM_ROLES_ENABLED_SETTING = {
    key: 'externalSecrets.systemRolesEnabled',
    roleScopeMap: {
        [exports.PROJECT_ADMIN_ROLE_SLUG]: [
            'externalSecretsProvider:create',
            'externalSecretsProvider:read',
            'externalSecretsProvider:update',
            'externalSecretsProvider:delete',
            'externalSecretsProvider:list',
            'externalSecretsProvider:sync',
            'externalSecret:list',
        ],
        [exports.PROJECT_EDITOR_ROLE_SLUG]: [
            'externalSecretsProvider:read',
            'externalSecretsProvider:list',
            'externalSecret:list',
        ],
    },
};
