// n8n-core/dist/execution-engine/requests-response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleRequest = handleRequest;
exports.isEngineRequest = isEngineRequest;
exports.makeEngineResponse = makeEngineResponse;
const di_1 = require("@n8n/di");
const n8n_workflow_1 = require("n8n-workflow");
const error_reporter_1 = require("../errors/error-reporter");
function buildParentOutputData(json, parentRunIndex, parentOutputIndex, parentSourceNode) {
    return [
        [
            {
                json,
                pairedItem: {
                    item: parentRunIndex,
                    input: parentOutputIndex,
                    sourceOverwrite: {
                        previousNode: parentSourceNode,
                        previousNodeOutput: parentOutputIndex,
                        previousNodeRun: parentRunIndex,
                    },
                },
            },
        ],
    ];
}
function initializeNodeRunData(runData, nodeName, parentNode, parentOutputIndex, parentRunIndex, actionMetadata, runIndex, displayOutputData) {
    runData[nodeName] ||= [];
    const nodeRunData = runData[nodeName];
    const nodeRunIndex = nodeRunData.length;
    const sourceData = {
        previousNode: parentNode,
        previousNodeOutput: parentOutputIndex,
        previousNodeRun: actionMetadata?.parentNodeName ? parentRunIndex : runIndex,
    };
    nodeRunData.push({
        inputOverride: { ai_tool: displayOutputData },
        source: [sourceData],
        executionIndex: 0,
        executionTime: 0,
        startTime: 0,
    });
    return nodeRunIndex;
}
function prepareRequestedNodesForExecution(workflow, currentNode, request, runIndex, runData, executionData) {
    const nodesToBeExecuted = [];
    const subNodeExecutionData = {
        actions: [],
        metadata: request.metadata,
    };
    const parentSourceData = executionData.source?.main?.[runIndex];
    const defaultParentOutputIndex = parentSourceData?.previousNodeOutput ?? 0;
    const defaultParentSourceNode = parentSourceData?.previousNode ?? currentNode.name;
    for (const action of request.actions) {
        const node = workflow.getNode(action.nodeName);
        if (!node) {
            throw new n8n_workflow_1.UnexpectedError(`Workflow does not contain a node with the name of "${action.nodeName}".`);
        }
        node.rewireOutputLogTo = action.type;
        const actionMetadata = action.metadata;
        const parentNode = actionMetadata?.parentNodeName ?? currentNode.name;
        const parentRunIndex = actionMetadata?.parentNodeName
            ? (runData[parentNode]?.length ?? 1) - 1
            : (parentSourceData?.previousNodeRun ?? 0);
        const itemIndex = actionMetadata?.itemIndex ?? 0;
        const agentInputData = executionData.data.main?.[0]?.[itemIndex];
        const mergedJson = {
            ...(agentInputData?.json ?? {}),
            ...action.input,
            toolCallId: action.id,
        };
        const displayJson = {
            ...action.input,
        };
        const parentOutputData = buildParentOutputData(mergedJson, parentRunIndex, defaultParentOutputIndex, defaultParentSourceNode);
        const displayOutputData = buildParentOutputData(displayJson, parentRunIndex, defaultParentOutputIndex, defaultParentSourceNode);
        const nodeRunIndex = initializeNodeRunData(runData, node.name, parentNode, defaultParentOutputIndex, parentRunIndex, actionMetadata, runIndex, displayOutputData);
        nodesToBeExecuted.push({
            inputConnectionData: { type: action.type, node: action.nodeName, index: 0 },
            parentOutputIndex: 0,
            parentNode,
            parentOutputData,
            runIndex: actionMetadata?.parentNodeName ? parentRunIndex : runIndex,
            nodeRunIndex,
            metadata: {
                preserveSourceOverwrite: true,
                preservedSourceOverwrite: executionData.metadata?.preservedSourceOverwrite ?? {
                    previousNode: defaultParentSourceNode,
                    previousNodeOutput: defaultParentOutputIndex,
                    previousNodeRun: parentRunIndex,
                },
            },
        });
        subNodeExecutionData.actions.push({
            action,
            nodeName: action.nodeName,
            runIndex: nodeRunIndex,
        });
    }
    return { nodesToBeExecuted, subNodeExecutionData };
}
function prepareRequestingNodeForResuming(workflow, request, executionData) {
    const parentNode = executionData.source?.main?.[0]?.previousNode;
    if (!parentNode) {
        di_1.Container.get(error_reporter_1.ErrorReporter).error(new n8n_workflow_1.UnexpectedError('Cannot find parent node for subnode execution - request will be ignored'), {
            extra: {
                executionNode: executionData.node.name,
                sourceData: executionData.source,
                workflowId: workflow.id,
                requestActions: request.actions.map((a) => ({
                    nodeName: a.nodeName,
                    actionType: a.actionType,
                    id: a.id,
                })),
            },
        });
        return undefined;
    }
    const metadata = executionData.metadata?.preservedSourceOverwrite &&
        executionData.metadata?.preserveSourceOverwrite
        ? {
            preserveSourceOverwrite: true,
            preservedSourceOverwrite: executionData.metadata.preservedSourceOverwrite,
        }
        : {};
    const connectionData = {
        type: 'ai_tool',
        node: executionData.node.name,
        index: 0,
    };
    return { connectionData, parentNode, metadata };
}
function handleRequest({ workflow, currentNode, request, runIndex, executionData, runData, }) {
    const { nodesToBeExecuted, subNodeExecutionData } = prepareRequestedNodesForExecution(workflow, currentNode, request, runIndex, runData, executionData);
    const result = prepareRequestingNodeForResuming(workflow, request, executionData);
    if (!result) {
        return { nodesToBeExecuted: [] };
    }
    nodesToBeExecuted.unshift({
        inputConnectionData: result.connectionData,
        parentOutputIndex: 0,
        parentNode: result.parentNode,
        parentOutputData: executionData.data.main,
        runIndex,
        nodeRunIndex: runIndex,
        metadata: { nodeWasResumed: true, subNodeExecutionData, ...result.metadata },
    });
    return { nodesToBeExecuted };
}
function isEngineRequest(responseOrRequest) {
    return !!responseOrRequest && 'actions' in responseOrRequest;
}
function makeEngineResponse() {
    return { actionResponses: [], metadata: {} };
}
