// n8n-core/dist/execution-engine/routing-node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoutingNode = void 0;
const get_1 = __importDefault(require("lodash/get"));
const merge_1 = __importDefault(require("lodash/merge"));
const set_1 = __importDefault(require("lodash/set"));
const n8n_workflow_1 = require("n8n-workflow");
const node_url_1 = __importDefault(require("node:url"));
const node_execution_context_1 = require("./node-execution-context");
class RoutingNode {
    constructor(context, nodeType, credentialsDecrypted) {
        this.context = context;
        this.nodeType = nodeType;
        this.credentialsDecrypted = credentialsDecrypted;
    }
    async runNode() {
        const { context, nodeType, credentialsDecrypted } = this;
        const { additionalData, executeData, inputData, node, workflow, mode, runIndex, connectionInputData, runExecutionData, } = context;
        const abortSignal = context.getExecutionCancelSignal();
        const items = (inputData[n8n_workflow_1.NodeConnectionTypes.Main] ??
            inputData[n8n_workflow_1.NodeConnectionTypes.AiTool])[0];
        const returnData = [];
        const { credentials, credentialDescription } = await this.prepareCredentials();
        const { batching } = context.getNodeParameter('requestOptions', 0, {});
        const batchSize = batching?.batch?.batchSize > 0 ? batching?.batch?.batchSize : 1;
        const batchInterval = batching?.batch.batchInterval;
        const requestPromises = [];
        const itemContext = [];
        for (let itemIndex = 0; itemIndex < items.length; itemIndex++) {
            if (itemIndex > 0 && batchSize >= 0 && batchInterval > 0) {
                if (itemIndex % batchSize === 0) {
                    await (0, n8n_workflow_1.sleep)(batchInterval);
                }
            }
            const thisArgs = new node_execution_context_1.ExecuteSingleContext(workflow, node, additionalData, mode, runExecutionData, runIndex, connectionInputData, inputData, itemIndex, executeData, abortSignal);
            itemContext.push({
                thisArgs,
                requestData: {
                    options: {
                        qs: {},
                        body: {},
                        headers: {},
                    },
                    preSend: [],
                    postReceive: [],
                    requestOperations: {},
                },
            });
            const { proxy, timeout, allowUnauthorizedCerts } = itemContext[itemIndex].thisArgs.getNodeParameter('requestOptions', 0, {});
            if (nodeType.description.requestOperations) {
                itemContext[itemIndex].requestData.requestOperations = {
                    ...nodeType.description.requestOperations,
                };
            }
            if (nodeType.description.requestDefaults) {
                for (const key of Object.keys(nodeType.description.requestDefaults)) {
                    let value = nodeType.description.requestDefaults[key];
                    value = this.getParameterValue(value, itemIndex, runIndex, executeData, { $credentials: credentials, $version: node.typeVersion }, false);
                    itemContext[itemIndex].requestData.options[key] = value;
                }
            }
            for (const property of nodeType.description.properties) {
                let value = (0, get_1.default)(node.parameters, property.name, []);
                value = this.getParameterValue(value, itemIndex, runIndex, executeData, { $credentials: credentials, $version: node.typeVersion }, false);
                const tempOptions = this.getRequestOptionsFromParameters(itemContext[itemIndex].thisArgs, property, itemIndex, runIndex, '', { $credentials: credentials, $value: value, $version: node.typeVersion });
                this.mergeOptions(itemContext[itemIndex].requestData, tempOptions);
            }
            if (proxy) {
                const proxyParsed = node_url_1.default.parse(proxy);
                const proxyProperties = ['host', 'port'];
                for (const property of proxyProperties) {
                    if (!(property in proxyParsed) ||
                        proxyParsed[property] === null) {
                        throw new n8n_workflow_1.NodeOperationError(node, 'The proxy is not value', {
                            runIndex,
                            itemIndex,
                            description: `The proxy URL does not contain a valid value for "${property}"`,
                        });
                    }
                }
                itemContext[itemIndex].requestData.options.proxy = {
                    host: proxyParsed.hostname,
                    port: parseInt(proxyParsed.port),
                    protocol: proxyParsed.protocol?.replace(/:$/, '') || undefined,
                };
                if (proxyParsed.auth) {
                    const [username, password] = proxyParsed.auth.split(':');
                    itemContext[itemIndex].requestData.options.proxy.auth = {
                        username,
                        password,
                    };
                }
            }
            if (allowUnauthorizedCerts) {
                itemContext[itemIndex].requestData.options.skipSslCertificateValidation =
                    allowUnauthorizedCerts;
            }
            if (timeout) {
                itemContext[itemIndex].requestData.options.timeout = timeout;
            }
            else {
                itemContext[itemIndex].requestData.options.timeout = 300_000;
            }
            requestPromises.push(this.makeRequest(itemContext[itemIndex].requestData, itemContext[itemIndex].thisArgs, itemIndex, runIndex, credentialDescription?.name, itemContext[itemIndex].requestData.requestOperations, credentialsDecrypted));
        }
        const promisesResponses = await Promise.allSettled(requestPromises);
        let responseData;
        for (let itemIndex = 0; itemIndex < items.length; itemIndex++) {
            responseData = promisesResponses.shift();
            if (responseData.status !== 'fulfilled') {
                if (responseData.reason.statusCode === 429) {
                    responseData.reason.message =
                        "Try spacing your requests out using the batching settings under 'Options'";
                }
                const error = responseData.reason;
                if (itemContext[itemIndex].thisArgs?.continueOnFail()) {
                    returnData.push({ json: {}, error: error });
                    continue;
                }
                if (error instanceof n8n_workflow_1.NodeApiError) {
                    (0, set_1.default)(error, 'context.itemIndex', itemIndex);
                    (0, set_1.default)(error, 'context.runIndex', runIndex);
                    throw error;
                }
                throw new n8n_workflow_1.NodeApiError(node, error, {
                    runIndex,
                    itemIndex,
                    message: error?.message,
                    description: error?.description,
                    httpCode: error.isAxiosError && error.response ? String(error.response?.status) : 'none',
                });
            }
            if (itemContext[itemIndex].requestData.maxResults) {
                responseData.value.splice(itemContext[itemIndex].requestData.maxResults);
            }
            returnData.push(...responseData.value);
        }
        return [returnData];
    }
    mergeOptions(destinationOptions, sourceOptions) {
        if (sourceOptions) {
            destinationOptions.paginate = destinationOptions.paginate ?? sourceOptions.paginate;
            destinationOptions.maxResults = sourceOptions.maxResults
                ? sourceOptions.maxResults
                : destinationOptions.maxResults;
            (0, merge_1.default)(destinationOptions.options, sourceOptions.options);
            destinationOptions.preSend.push(...sourceOptions.preSend);
            destinationOptions.postReceive.push(...sourceOptions.postReceive);
            if (sourceOptions.requestOperations && destinationOptions.requestOperations) {
                destinationOptions.requestOperations = Object.assign(destinationOptions.requestOperations, sourceOptions.requestOperations);
            }
        }
    }
    async runPostReceiveAction(executeSingleFunctions, action, inputData, responseData, parameterValue, itemIndex, runIndex) {
        if (typeof action === 'function') {
            return await action.call(executeSingleFunctions, inputData, responseData);
        }
        const { node } = this.context;
        if (action.type === 'rootProperty') {
            try {
                return inputData.flatMap((item) => {
                    let itemContent = (0, get_1.default)(item.json, action.properties.property);
                    if (!Array.isArray(itemContent)) {
                        itemContent = [itemContent];
                    }
                    return itemContent.map((json) => {
                        return {
                            json,
                        };
                    });
                });
            }
            catch (error) {
                throw new n8n_workflow_1.NodeOperationError(node, error, {
                    runIndex,
                    itemIndex,
                    description: `The rootProperty "${action.properties.property}" could not be found on item.`,
                });
            }
        }
        if (action.type === 'filter') {
            const passValue = action.properties.pass;
            const { credentials } = await this.prepareCredentials();
            inputData = inputData.filter((item) => {
                return this.getParameterValue(passValue, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), {
                    $credentials: credentials,
                    $response: responseData,
                    $responseItem: item.json,
                    $value: parameterValue,
                    $version: node.typeVersion,
                }, false);
            });
            return inputData;
        }
        if (action.type === 'limit') {
            const maxResults = this.getParameterValue(action.properties.maxResults, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), { $response: responseData, $value: parameterValue, $version: node.typeVersion }, false);
            return inputData.slice(0, parseInt(maxResults, 10));
        }
        if (action.type === 'set') {
            const { value } = action.properties;
            return [
                {
                    json: this.getParameterValue(value, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), { $response: responseData, $value: parameterValue, $version: node.typeVersion }, false),
                },
            ];
        }
        if (action.type === 'sort') {
            const sortKey = action.properties.key;
            inputData.sort((a, b) => {
                const aSortValue = a.json[sortKey]?.toString().toLowerCase() ?? '';
                const bSortValue = b.json[sortKey]?.toString().toLowerCase() ?? '';
                if (aSortValue < bSortValue) {
                    return -1;
                }
                if (aSortValue > bSortValue) {
                    return 1;
                }
                return 0;
            });
            return inputData;
        }
        if (action.type === 'setKeyValue') {
            const returnData = [];
            inputData.forEach((item) => {
                const returnItem = {};
                for (const key of Object.keys(action.properties)) {
                    let propertyValue = action.properties[key];
                    propertyValue = this.getParameterValue(propertyValue, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), {
                        $response: responseData,
                        $responseItem: item.json,
                        $value: parameterValue,
                        $version: node.typeVersion,
                    }, false);
                    returnItem[key] = propertyValue;
                }
                returnData.push({ json: returnItem });
            });
            return returnData;
        }
        if (action.type === 'binaryData') {
            const body = (responseData.body = Buffer.from(responseData.body));
            let { destinationProperty } = action.properties;
            destinationProperty = this.getParameterValue(destinationProperty, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), { $response: responseData, $value: parameterValue, $version: node.typeVersion }, false);
            const binaryData = await executeSingleFunctions.helpers.prepareBinaryData(body);
            return inputData.map((item) => {
                if (typeof item.json === 'string') {
                    item.json = {};
                }
                item.binary = {
                    [destinationProperty]: binaryData,
                };
                return item;
            });
        }
        return [];
    }
    async postProcessResponseData(executeSingleFunctions, responseData, requestData, itemIndex, runIndex) {
        let returnData = [
            {
                json: responseData.body,
            },
        ];
        if (requestData.postReceive.length) {
            for (const postReceiveMethod of requestData.postReceive) {
                for (const action of postReceiveMethod.actions) {
                    returnData = await this.runPostReceiveAction(executeSingleFunctions, action, returnData, responseData, postReceiveMethod.data.parameterValue, itemIndex, runIndex);
                }
            }
        }
        else {
            if (Array.isArray(responseData.body)) {
                returnData = responseData.body.map((json) => {
                    return {
                        json,
                    };
                });
            }
            else {
                returnData[0].json = responseData.body;
            }
        }
        return returnData;
    }
    async rawRoutingRequest(executeSingleFunctions, requestData, credentialType, credentialsDecrypted) {
        let responseData;
        requestData.options.returnFullResponse = true;
        if (credentialType) {
            responseData = (await executeSingleFunctions.helpers.httpRequestWithAuthentication.call(executeSingleFunctions, credentialType, requestData.options, { credentialsDecrypted }));
        }
        else {
            responseData = (await executeSingleFunctions.helpers.httpRequest(requestData.options));
        }
        return responseData;
    }
    async makeRequest(requestData, executeSingleFunctions, itemIndex, runIndex, credentialType, requestOperations, credentialsDecrypted) {
        let responseData;
        for (const preSendMethod of requestData.preSend) {
            requestData.options = await preSendMethod.call(executeSingleFunctions, requestData.options);
        }
        const makeRoutingRequest = async (requestOptions) => {
            return await this.rawRoutingRequest(executeSingleFunctions, requestOptions, credentialType, credentialsDecrypted).then(async (data) => await this.postProcessResponseData(executeSingleFunctions, data, requestData, itemIndex, runIndex));
        };
        const { node } = this.context;
        const executePaginationFunctions = Object.create(executeSingleFunctions, {
            makeRoutingRequest: { value: makeRoutingRequest },
        });
        if (requestData.paginate && requestOperations?.pagination) {
            if (typeof requestOperations.pagination === 'function') {
                responseData = await requestOperations.pagination.call(executePaginationFunctions, requestData);
            }
            else {
                responseData = [];
                if (!requestData.options.qs) {
                    requestData.options.qs = {};
                }
                if (requestOperations.pagination.type === 'generic') {
                    let tempResponseData;
                    let tempResponseItems;
                    let makeAdditionalRequest;
                    let paginateRequestData;
                    const additionalKeys = {
                        $request: requestData.options,
                        $response: {},
                        $version: node.typeVersion,
                    };
                    do {
                        additionalKeys.$request = requestData.options;
                        paginateRequestData = this.getParameterValue(requestOperations.pagination.properties.request, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), additionalKeys, false);
                        tempResponseData = await this.rawRoutingRequest(executeSingleFunctions, { ...requestData, options: { ...requestData.options, ...paginateRequestData } }, credentialType, credentialsDecrypted);
                        additionalKeys.$response = tempResponseData;
                        tempResponseItems = await this.postProcessResponseData(executeSingleFunctions, tempResponseData, requestData, itemIndex, runIndex);
                        responseData.push(...tempResponseItems);
                        makeAdditionalRequest = this.getParameterValue(requestOperations.pagination.properties.continue, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), additionalKeys, false);
                    } while (makeAdditionalRequest);
                }
                else if (requestOperations.pagination.type === 'offset') {
                    const { properties } = requestOperations.pagination;
                    const optionsType = properties.type === 'body' ? 'body' : 'qs';
                    if (properties.type === 'body' && !requestData.options.body) {
                        requestData.options.body = {};
                    }
                    requestData.options[optionsType][properties.limitParameter] =
                        properties.pageSize;
                    requestData.options[optionsType][properties.offsetParameter] = 0;
                    let tempResponseData;
                    do {
                        if (requestData?.maxResults) {
                            const resultsMissing = requestData?.maxResults - responseData.length;
                            if (resultsMissing < 1) {
                                break;
                            }
                            requestData.options[optionsType][properties.limitParameter] =
                                Math.min(properties.pageSize, resultsMissing);
                        }
                        tempResponseData = await this.rawRoutingRequest(executeSingleFunctions, requestData, credentialType, credentialsDecrypted).then(async (data) => await this.postProcessResponseData(executeSingleFunctions, data, requestData, itemIndex, runIndex));
                        requestData.options[optionsType][properties.offsetParameter] =
                            requestData.options[optionsType][properties.offsetParameter] + properties.pageSize;
                        if (properties.rootProperty) {
                            const tempResponseValue = (0, get_1.default)(tempResponseData[0].json, properties.rootProperty);
                            if (tempResponseValue === undefined) {
                                throw new n8n_workflow_1.NodeOperationError(node, `The rootProperty "${properties.rootProperty}" could not be found on item.`, { runIndex, itemIndex });
                            }
                            tempResponseData = tempResponseValue.map((item) => {
                                return {
                                    json: item,
                                };
                            });
                        }
                        responseData.push(...tempResponseData);
                    } while (tempResponseData.length && tempResponseData.length === properties.pageSize);
                }
            }
        }
        else {
            responseData = await this.rawRoutingRequest(executeSingleFunctions, requestData, credentialType, credentialsDecrypted).then(async (data) => await this.postProcessResponseData(executeSingleFunctions, data, requestData, itemIndex, runIndex));
        }
        return responseData;
    }
    getParameterValue(parameterValue, itemIndex, runIndex, executeData, additionalKeys, returnObjectAsString = false) {
        if (typeof parameterValue === 'object' ||
            (typeof parameterValue === 'string' && parameterValue.charAt(0) === '=')) {
            const { node, workflow, mode, connectionInputData, runExecutionData } = this.context;
            return workflow.expression.getParameterValue(parameterValue, runExecutionData ?? null, runIndex, itemIndex, node.name, connectionInputData, mode, additionalKeys ?? {}, executeData, returnObjectAsString);
        }
        return parameterValue;
    }
    getRequestOptionsFromParameters(executeSingleFunctions, nodeProperties, itemIndex, runIndex, path, additionalKeys) {
        const returnData = {
            options: {
                qs: {},
                body: {},
                headers: {},
            },
            preSend: [],
            postReceive: [],
            requestOperations: {},
        };
        let basePath = path ? `${path}.` : '';
        const { node, nodeType } = this.context;
        if (!n8n_workflow_1.NodeHelpers.displayParameter(node.parameters, nodeProperties, node, nodeType.description, node.parameters)) {
            return undefined;
        }
        if (nodeProperties.routing) {
            let parameterValue;
            if (basePath + nodeProperties.name && 'type' in nodeProperties) {
                const shouldExtractValue = nodeProperties.extractValue !== undefined || nodeProperties.type === 'resourceLocator';
                parameterValue = executeSingleFunctions.getNodeParameter(basePath + nodeProperties.name, undefined, { extractValue: shouldExtractValue });
            }
            if (nodeProperties.routing.operations) {
                returnData.requestOperations = { ...nodeProperties.routing.operations };
            }
            if (nodeProperties.routing.request) {
                for (const key of Object.keys(nodeProperties.routing.request)) {
                    let propertyValue = nodeProperties.routing.request[key];
                    propertyValue = this.getParameterValue(propertyValue, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), { ...additionalKeys, $value: parameterValue }, false);
                    returnData.options[key] = propertyValue;
                }
            }
            if (nodeProperties.routing.send) {
                let propertyName = nodeProperties.routing.send.property;
                if (propertyName !== undefined) {
                    propertyName = this.getParameterValue(propertyName, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), additionalKeys, true);
                    let value = parameterValue;
                    if (nodeProperties.routing.send.value) {
                        const valueString = nodeProperties.routing.send.value;
                        value = this.getParameterValue(valueString, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), { ...additionalKeys, $value: value }, false);
                    }
                    if (nodeProperties.routing.send.type === 'body') {
                        if (nodeProperties.routing.send.propertyInDotNotation === false) {
                            returnData.options.body[propertyName] = value;
                        }
                        else {
                            (0, set_1.default)(returnData.options.body, propertyName, value);
                        }
                    }
                    else {
                        if (nodeProperties.routing.send.propertyInDotNotation === false) {
                            returnData.options.qs[propertyName] = value;
                        }
                        else {
                            (0, set_1.default)(returnData.options.qs, propertyName, value);
                        }
                    }
                }
                if (nodeProperties.routing.send.paginate !== undefined) {
                    let paginateValue = nodeProperties.routing.send.paginate;
                    if (typeof paginateValue === 'string' && paginateValue.charAt(0) === '=') {
                        paginateValue = this.getParameterValue(paginateValue, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), { ...additionalKeys, $value: parameterValue }, true);
                    }
                    returnData.paginate = !!paginateValue;
                }
                if (nodeProperties.routing.send.preSend) {
                    returnData.preSend.push(...nodeProperties.routing.send.preSend);
                }
            }
            if (nodeProperties.routing.output) {
                if (nodeProperties.routing.output.maxResults !== undefined) {
                    let maxResultsValue = nodeProperties.routing.output.maxResults;
                    if (typeof maxResultsValue === 'string' && maxResultsValue.charAt(0) === '=') {
                        maxResultsValue = this.getParameterValue(maxResultsValue, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), { ...additionalKeys, $value: parameterValue }, true);
                    }
                    returnData.maxResults = maxResultsValue;
                }
                if (nodeProperties.routing.output.postReceive) {
                    const postReceiveActions = nodeProperties.routing.output.postReceive.filter((action) => {
                        if (typeof action === 'function') {
                            return true;
                        }
                        if (typeof action.enabled === 'string' && action.enabled.charAt(0) === '=') {
                            return this.getParameterValue(action.enabled, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), { ...additionalKeys, $value: parameterValue }, true);
                        }
                        return action.enabled !== false;
                    });
                    if (postReceiveActions.length) {
                        returnData.postReceive.push({
                            data: {
                                parameterValue,
                            },
                            actions: postReceiveActions,
                        });
                    }
                }
            }
        }
        if (!Object.prototype.hasOwnProperty.call(nodeProperties, 'options')) {
            return returnData;
        }
        nodeProperties = nodeProperties;
        let value;
        if (nodeProperties.type === 'options') {
            const optionValue = n8n_workflow_1.NodeHelpers.getParameterValueByPath(node.parameters, nodeProperties.name, basePath.slice(0, -1));
            const selectedOption = nodeProperties.options.filter((option) => option.value === optionValue);
            if (selectedOption.length) {
                const tempOptions = this.getRequestOptionsFromParameters(executeSingleFunctions, selectedOption[0], itemIndex, runIndex, `${basePath}${nodeProperties.name}`, { $value: optionValue, $version: node.typeVersion });
                this.mergeOptions(returnData, tempOptions);
            }
        }
        else if (nodeProperties.type === 'collection') {
            value = n8n_workflow_1.NodeHelpers.getParameterValueByPath(node.parameters, nodeProperties.name, basePath.slice(0, -1));
            for (const propertyOption of nodeProperties.options) {
                if (Object.keys(value).includes(propertyOption.name) &&
                    propertyOption.type !== undefined) {
                    const tempOptions = this.getRequestOptionsFromParameters(executeSingleFunctions, propertyOption, itemIndex, runIndex, `${basePath}${nodeProperties.name}`, { $version: node.typeVersion });
                    this.mergeOptions(returnData, tempOptions);
                }
            }
        }
        else if (nodeProperties.type === 'fixedCollection') {
            basePath = `${basePath}${nodeProperties.name}.`;
            for (const propertyOptions of nodeProperties.options) {
                value = n8n_workflow_1.NodeHelpers.getParameterValueByPath(node.parameters, propertyOptions.name, basePath.slice(0, -1));
                if (value === undefined) {
                    continue;
                }
                if (!Array.isArray(value)) {
                    value = [value];
                }
                value = this.getParameterValue(value, itemIndex, runIndex, executeSingleFunctions.getExecuteData(), { ...additionalKeys }, false);
                const loopBasePath = `${basePath}${propertyOptions.name}`;
                for (let i = 0; i < value.length; i++) {
                    for (const option of propertyOptions.values) {
                        const tempOptions = this.getRequestOptionsFromParameters(executeSingleFunctions, option, itemIndex, runIndex, nodeProperties.typeOptions?.multipleValues ? `${loopBasePath}[${i}]` : loopBasePath, { ...(additionalKeys || {}), $index: i, $parent: value[i] });
                        this.mergeOptions(returnData, tempOptions);
                    }
                }
            }
        }
        return returnData;
    }
    async prepareCredentials() {
        const { context, nodeType, credentialsDecrypted } = this;
        const { node } = context;
        let credentialDescription;
        if (nodeType.description.credentials?.length) {
            if (nodeType.description.credentials.length === 1) {
                credentialDescription = nodeType.description.credentials[0];
            }
            else {
                const authenticationMethod = context.getNodeParameter('authentication', 0);
                credentialDescription = nodeType.description.credentials.find((x) => x.displayOptions?.show?.authentication?.includes(authenticationMethod));
                if (!credentialDescription) {
                    throw new n8n_workflow_1.NodeOperationError(node, `Node type "${node.type}" does not have any credentials of type "${authenticationMethod}" defined`, { level: 'warning' });
                }
            }
        }
        let credentials;
        if (credentialsDecrypted) {
            credentials = credentialsDecrypted.data;
        }
        else if (credentialDescription) {
            try {
                credentials =
                    (await context.getCredentials(credentialDescription.name, 0)) || {};
            }
            catch (error) {
                if (credentialDescription.required) {
                    throw error;
                }
                else {
                    credentialDescription = undefined;
                }
            }
        }
        return { credentials, credentialDescription };
    }
}
exports.RoutingNode = RoutingNode;
