// n8n-core/dist/binary-data/binary-data.service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BinaryDataService = void 0;
const backend_common_1 = require("@n8n/backend-common");
const di_1 = require("@n8n/di");
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
const n8n_workflow_1 = require("n8n-workflow");
const promises_1 = require("node:fs/promises");
const pretty_bytes_1 = __importDefault(require("pretty-bytes"));
const errors_1 = require("../errors");
const binary_data_config_1 = require("./binary-data.config");
const utils_1 = require("./utils");
const invalid_manager_error_1 = require("../errors/invalid-manager.error");
let BinaryDataService = class BinaryDataService {
    constructor(config, errorReporter, logger) {
        this.config = config;
        this.errorReporter = errorReporter;
        this.logger = logger;
        this.mode = 'filesystem-v2';
        this.managers = {};
    }
    setManager(mode, manager) {
        this.managers[mode] = manager;
    }
    async init() {
        const { config } = this;
        this.mode = config.mode === 'filesystem' ? 'filesystem-v2' : config.mode;
        const { FileSystemManager } = await Promise.resolve().then(() => __importStar(require('./file-system.manager')));
        this.managers.filesystem = new FileSystemManager(config.localStoragePath, this.errorReporter);
        this.managers['filesystem-v2'] = this.managers.filesystem;
        await this.managers.filesystem.init();
    }
    createSignedToken(binaryData, expiresIn = '1 day') {
        if (!binaryData.id) {
            throw new n8n_workflow_1.UnexpectedError('URL signing is not available in memory mode');
        }
        const signingPayload = {
            id: binaryData.id,
        };
        const { signingSecret } = this.config;
        return jsonwebtoken_1.default.sign(signingPayload, signingSecret, { expiresIn });
    }
    validateSignedToken(token) {
        const { signingSecret } = this.config;
        const signedPayload = jsonwebtoken_1.default.verify(token, signingSecret);
        return signedPayload.id;
    }
    async copyBinaryFile(location, binaryData, filePath) {
        const manager = this.managers[this.mode];
        if (!manager) {
            const { size } = await (0, promises_1.stat)(filePath);
            binaryData.fileSize = (0, pretty_bytes_1.default)(size);
            binaryData.bytes = size;
            binaryData.data = await (0, promises_1.readFile)(filePath, { encoding: n8n_workflow_1.BINARY_ENCODING });
            return binaryData;
        }
        const metadata = {
            fileName: binaryData.fileName,
            mimeType: binaryData.mimeType,
        };
        const { fileId, fileSize } = await manager.copyByFilePath(location, filePath, metadata);
        binaryData.id = this.createBinaryDataId(fileId);
        binaryData.fileSize = (0, pretty_bytes_1.default)(fileSize);
        binaryData.bytes = fileSize;
        binaryData.data = this.mode;
        return binaryData;
    }
    async store(location, bufferOrStream, binaryData) {
        const manager = this.managers[this.mode];
        if (!manager) {
            const buffer = await (0, utils_1.binaryToBuffer)(bufferOrStream);
            binaryData.data = buffer.toString(n8n_workflow_1.BINARY_ENCODING);
            binaryData.fileSize = (0, pretty_bytes_1.default)(buffer.length);
            binaryData.bytes = buffer.length;
            return binaryData;
        }
        const metadata = {
            fileName: binaryData.fileName,
            mimeType: binaryData.mimeType,
        };
        const { fileId, fileSize } = await manager.store(location, bufferOrStream, metadata);
        binaryData.id = this.createBinaryDataId(fileId);
        binaryData.fileSize = (0, pretty_bytes_1.default)(fileSize);
        binaryData.bytes = fileSize;
        binaryData.data = this.mode;
        return binaryData;
    }
    async getAsStream(binaryDataId, chunkSize) {
        const [mode, fileId] = binaryDataId.split(':');
        return await this.getManager(mode).getAsStream(fileId, chunkSize);
    }
    async getAsBuffer(binaryData) {
        if (binaryData.id) {
            const [mode, fileId] = binaryData.id.split(':');
            return await this.getManager(mode).getAsBuffer(fileId);
        }
        return Buffer.from(binaryData.data, n8n_workflow_1.BINARY_ENCODING);
    }
    getPath(binaryDataId) {
        const [mode, fileId] = binaryDataId.split(':');
        return this.getManager(mode).getPath(fileId);
    }
    async getMetadata(binaryDataId) {
        const [mode, fileId] = binaryDataId.split(':');
        return await this.getManager(mode).getMetadata(fileId);
    }
    async deleteMany(locations) {
        const manager = this.managers[this.mode];
        if (!manager)
            return;
        if (manager.deleteMany)
            await manager.deleteMany(locations);
    }
    async deleteManyByBinaryDataId(ids) {
        const fileIdsByMode = new Map();
        for (const attachmentId of ids) {
            const [mode, fileId] = attachmentId.split(':');
            if (!fileId) {
                continue;
            }
            const entry = fileIdsByMode.get(mode) ?? [];
            fileIdsByMode.set(mode, entry.concat([fileId]));
        }
        for (const [mode, fileIds] of fileIdsByMode) {
            const manager = this.managers[mode];
            if (!manager) {
                this.logger.info(`File manager of mode ${mode} is missing. Skip deleting these files: ${fileIds.join(', ')}`);
                continue;
            }
            await manager.deleteManyByFileId?.(fileIds);
        }
    }
    async duplicateBinaryData(location, inputData) {
        if (inputData && this.managers[this.mode]) {
            const returnInputData = inputData.map(async (executionDataArray) => {
                if (executionDataArray) {
                    return await Promise.all(executionDataArray.map(async (executionData) => {
                        if (executionData.binary) {
                            return await this.duplicateBinaryDataInExecData(location, executionData);
                        }
                        return executionData;
                    }));
                }
                return executionDataArray;
            });
            return await Promise.all(returnInputData);
        }
        return inputData;
    }
    async rename(oldFileId, newFileId) {
        const manager = this.getManager(this.mode);
        if (!manager)
            return;
        await manager.rename(oldFileId, newFileId);
    }
    createBinaryDataId(fileId) {
        return `${this.mode}:${fileId}`;
    }
    async duplicateBinaryDataInExecData(location, executionData) {
        const manager = this.managers[this.mode];
        if (executionData.binary) {
            const binaryDataKeys = Object.keys(executionData.binary);
            const bdPromises = binaryDataKeys.map(async (key) => {
                if (!executionData.binary) {
                    return { key, newId: undefined };
                }
                const binaryDataId = executionData.binary[key].id;
                if (!binaryDataId) {
                    return { key, newId: undefined };
                }
                const [_mode, fileId] = binaryDataId.split(':');
                return await manager?.copyByFileId(location, fileId).then((newFileId) => ({
                    newId: this.createBinaryDataId(newFileId),
                    key,
                }));
            });
            return await Promise.all(bdPromises).then((b) => {
                return b.reduce((acc, curr) => {
                    if (acc.binary && curr) {
                        acc.binary[curr.key].id = curr.newId;
                    }
                    return acc;
                }, executionData);
            });
        }
        return executionData;
    }
    getManager(mode) {
        const manager = this.managers[mode];
        if (manager)
            return manager;
        throw new invalid_manager_error_1.InvalidManagerError(mode);
    }
};
exports.BinaryDataService = BinaryDataService;
exports.BinaryDataService = BinaryDataService = __decorate([
    (0, di_1.Service)(),
    __metadata("design:paramtypes", [binary_data_config_1.BinaryDataConfig,
        errors_1.ErrorReporter,
        backend_common_1.Logger])
], BinaryDataService);
