// @n8n/workflow-sdk/dist/workflow-builder/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.START_X = exports.DEFAULT_Y = exports.NODE_SPACING_X = exports.STICKY_NODE_TYPE = exports.STICKY_BOTTOM_PADDING = exports.AI_Y_SPACING = exports.AI_X_SPACING = exports.SUBGRAPH_SPACING = exports.NODE_Y_SPACING = exports.NODE_X_SPACING = exports.NODE_MIN_INPUT_ITEMS_COUNT = exports.CONFIGURABLE_NODE_SIZE = exports.CONFIGURATION_NODE_SIZE = exports.CONFIGURATION_NODE_RADIUS = exports.DEFAULT_NODE_SIZE = exports.GRID_SIZE = void 0;
exports.GRID_SIZE = 16;
exports.DEFAULT_NODE_SIZE = [exports.GRID_SIZE * 6, exports.GRID_SIZE * 6];
exports.CONFIGURATION_NODE_RADIUS = (exports.GRID_SIZE * 5) / 2;
exports.CONFIGURATION_NODE_SIZE = [
    exports.CONFIGURATION_NODE_RADIUS * 2,
    exports.CONFIGURATION_NODE_RADIUS * 2,
];
exports.CONFIGURABLE_NODE_SIZE = [exports.GRID_SIZE * 16, exports.GRID_SIZE * 6];
exports.NODE_MIN_INPUT_ITEMS_COUNT = 4;
exports.NODE_X_SPACING = exports.GRID_SIZE * 8;
exports.NODE_Y_SPACING = exports.GRID_SIZE * 6;
exports.SUBGRAPH_SPACING = exports.GRID_SIZE * 8;
exports.AI_X_SPACING = exports.GRID_SIZE * 3;
exports.AI_Y_SPACING = exports.GRID_SIZE * 8;
exports.STICKY_BOTTOM_PADDING = exports.GRID_SIZE * 4;
exports.STICKY_NODE_TYPE = 'n8n-nodes-base.stickyNote';
exports.NODE_SPACING_X = 200;
exports.DEFAULT_Y = 300;
exports.START_X = 100;
