// @n8n/workflow-sdk/dist/workflow-builder/layout-utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateNodePositions = calculateNodePositions;
exports.getNodeDimensions = getNodeDimensions;
exports.calculateNodePositionsDagre = calculateNodePositionsDagre;
exports.layoutWorkflowJSON = layoutWorkflowJSON;
const dagre_1 = __importDefault(require("@dagrejs/dagre"));
const constants_1 = require("./constants");
function calculateNodePositions(nodes) {
    const positions = new Map();
    const hasIncoming = new Set();
    for (const graphNode of nodes.values()) {
        for (const typeConns of graphNode.connections.values()) {
            for (const targets of typeConns.values()) {
                for (const target of targets) {
                    hasIncoming.add(target.node);
                }
            }
        }
    }
    const rootNodes = [...nodes.keys()].filter((name) => !hasIncoming.has(name));
    const visited = new Set();
    const queue = [];
    let y = constants_1.DEFAULT_Y;
    for (const rootName of rootNodes) {
        queue.push({ name: rootName, x: constants_1.START_X, y });
        y += 150;
    }
    while (queue.length > 0) {
        const { name, x, y: nodeY } = queue.shift();
        if (visited.has(name))
            continue;
        visited.add(name);
        const node = nodes.get(name);
        if (node && !node.instance.config?.position) {
            positions.set(name, [x, nodeY]);
        }
        if (node) {
            let branchOffset = 0;
            for (const typeConns of node.connections.values()) {
                for (const targets of typeConns.values()) {
                    for (const target of targets) {
                        if (!visited.has(target.node)) {
                            queue.push({
                                name: target.node,
                                x: x + constants_1.NODE_SPACING_X,
                                y: nodeY + branchOffset * 150,
                            });
                            branchOffset++;
                        }
                    }
                }
            }
        }
    }
    return positions;
}
function isAiConnectionType(type) {
    return type.startsWith('ai_');
}
function getAiParentNames(nodes) {
    const parents = new Set();
    for (const graphNode of nodes.values()) {
        for (const [connType, outputMap] of graphNode.connections) {
            if (!isAiConnectionType(connType))
                continue;
            for (const targets of outputMap.values()) {
                for (const target of targets) {
                    parents.add(target.node);
                }
            }
        }
    }
    return parents;
}
function getAiConfigNames(nodes) {
    const configs = new Set();
    for (const [name, graphNode] of nodes) {
        for (const connType of graphNode.connections.keys()) {
            if (isAiConnectionType(connType)) {
                configs.add(name);
                break;
            }
        }
    }
    return configs;
}
function getAllConnectedAiConfigNodes(graph, rootId, aiConfigNames) {
    const predecessors = graph.predecessors(rootId) ?? [];
    return predecessors
        .filter((id) => aiConfigNames.has(id))
        .flatMap((id) => [id, ...getAllConnectedAiConfigNodes(graph, id, aiConfigNames)]);
}
function getMainOutputCount(nodeName, nodes) {
    const graphNode = nodes.get(nodeName);
    if (!graphNode)
        return 1;
    const mainConns = graphNode.connections.get('main');
    if (!mainConns || mainConns.size === 0)
        return 1;
    return Math.max(...mainConns.keys()) + 1;
}
function getMainInputCount(nodeName, nodes) {
    let maxIndex = 0;
    for (const graphNode of nodes.values()) {
        const mainConns = graphNode.connections.get('main');
        if (!mainConns)
            continue;
        for (const targets of mainConns.values()) {
            for (const target of targets) {
                if (target.node === nodeName) {
                    maxIndex = Math.max(maxIndex, target.index + 1);
                }
            }
        }
    }
    return Math.max(1, maxIndex);
}
function calculateNodeHeight(mainInputCount, mainOutputCount) {
    const maxVerticalHandles = Math.max(mainInputCount, mainOutputCount, 1);
    return constants_1.DEFAULT_NODE_SIZE[1] + Math.max(0, maxVerticalHandles - 2) * constants_1.GRID_SIZE * 2;
}
function getNodeDimensions(nodeName, aiParentNames, aiConfigNames, nodes) {
    if (aiConfigNames.has(nodeName)) {
        return { width: constants_1.CONFIGURATION_NODE_SIZE[0], height: constants_1.CONFIGURATION_NODE_SIZE[1] };
    }
    if (aiParentNames.has(nodeName)) {
        const aiInputTypes = new Set();
        for (const graphNode of nodes.values()) {
            for (const [connType, outputMap] of graphNode.connections) {
                if (!isAiConnectionType(connType))
                    continue;
                for (const targets of outputMap.values()) {
                    for (const target of targets) {
                        if (target.node === nodeName) {
                            aiInputTypes.add(connType);
                        }
                    }
                }
            }
        }
        const portCount = Math.max(constants_1.NODE_MIN_INPUT_ITEMS_COUNT, aiInputTypes.size);
        const width = constants_1.CONFIGURATION_NODE_RADIUS * 2 + constants_1.GRID_SIZE * (portCount - 1) * 3;
        return { width, height: constants_1.CONFIGURABLE_NODE_SIZE[1] };
    }
    const mainInputCount = getMainInputCount(nodeName, nodes);
    const mainOutputCount = getMainOutputCount(nodeName, nodes);
    return {
        width: constants_1.DEFAULT_NODE_SIZE[0],
        height: calculateNodeHeight(mainInputCount, mainOutputCount),
    };
}
function snapToGrid(value) {
    return Math.round(value / constants_1.GRID_SIZE) * constants_1.GRID_SIZE;
}
function compositeBoundingBox(boxes) {
    const { minX, minY, maxX, maxY } = boxes.reduce((bbox, node) => ({
        minX: Math.min(bbox.minX, node.x),
        maxX: Math.max(bbox.maxX, node.x + node.width),
        minY: Math.min(bbox.minY, node.y),
        maxY: Math.max(bbox.maxY, node.y + node.height),
    }), { minX: Infinity, minY: Infinity, maxX: -Infinity, maxY: -Infinity });
    return { x: minX, y: minY, width: maxX - minX, height: maxY - minY };
}
function boundingBoxFromDagreNode(node) {
    return {
        x: node.x - node.width / 2,
        y: node.y - node.height / 2,
        width: node.width,
        height: node.height,
    };
}
function boundingBoxFromGraph(graph) {
    const nodeIds = graph.nodes();
    if (nodeIds.length === 0) {
        return { x: 0, y: 0, width: 0, height: 0 };
    }
    return compositeBoundingBox(nodeIds.map((nodeId) => boundingBoxFromDagreNode(graph.node(nodeId))));
}
function intersects(container, target, padding = 0) {
    const t = {
        x: target.x - padding,
        y: target.y - padding,
        width: target.width + padding * 2,
        height: target.height + padding * 2,
    };
    return !(t.x + t.width < container.x ||
        t.x > container.x + container.width ||
        t.y + t.height < container.y ||
        t.y > container.y + container.height);
}
function isCoveredBy(parent, child) {
    return (child.x >= parent.x &&
        child.y >= parent.y &&
        child.x + child.width <= parent.x + parent.width &&
        child.y + child.height <= parent.y + parent.height);
}
function centerHorizontally(container, target) {
    return container.x + container.width / 2 - target.width / 2;
}
function createSubGraph(nodeIds, parent) {
    const subGraph = new dagre_1.default.graphlib.Graph();
    subGraph.setGraph({
        rankdir: 'LR',
        edgesep: constants_1.NODE_Y_SPACING,
        nodesep: constants_1.NODE_Y_SPACING,
        ranksep: constants_1.NODE_X_SPACING,
    });
    subGraph.setDefaultEdgeLabel(() => ({}));
    const nodeIdSet = new Set(nodeIds);
    parent
        .nodes()
        .filter((id) => nodeIdSet.has(id))
        .forEach((id) => subGraph.setNode(id, parent.node(id)));
    parent
        .edges()
        .filter((edge) => nodeIdSet.has(edge.v) && nodeIdSet.has(edge.w))
        .forEach((edge) => subGraph.setEdge(edge.v, edge.w, parent.edge(edge)));
    return subGraph;
}
function createVerticalGraph(items) {
    const graph = new dagre_1.default.graphlib.Graph();
    graph.setGraph({
        rankdir: 'TB',
        align: 'UL',
        edgesep: constants_1.SUBGRAPH_SPACING,
        nodesep: constants_1.SUBGRAPH_SPACING,
        ranksep: constants_1.SUBGRAPH_SPACING,
    });
    graph.setDefaultEdgeLabel(() => ({}));
    items.forEach(({ id, box: { x, y, width, height } }) => graph.setNode(id, { x, y, width, height }));
    items.forEach(({ id }, index) => {
        if (items[index + 1]) {
            graph.setEdge(id, items[index + 1].id);
        }
    });
    return graph;
}
function createAiSubGraph(parent, nodeIds) {
    const graph = new dagre_1.default.graphlib.Graph();
    graph.setGraph({
        rankdir: 'TB',
        edgesep: constants_1.AI_X_SPACING,
        nodesep: constants_1.AI_X_SPACING,
        ranksep: constants_1.AI_Y_SPACING,
    });
    graph.setDefaultEdgeLabel(() => ({}));
    const nodeIdSet = new Set(nodeIds);
    parent
        .nodes()
        .filter((id) => nodeIdSet.has(id))
        .forEach((id) => graph.setNode(id, parent.node(id)));
    parent
        .edges()
        .filter((edge) => nodeIdSet.has(edge.v) && nodeIdSet.has(edge.w))
        .forEach((edge) => graph.setEdge(edge.w, edge.v));
    return graph;
}
function repositionStickyNotes(stickyNames, nonStickyNames, positionsBefore, positionsAfter, result) {
    for (const stickyName of stickyNames) {
        const stickyBoxBefore = positionsBefore.get(stickyName);
        if (!stickyBoxBefore)
            continue;
        const coveredNames = nonStickyNames.filter((name) => {
            const nodeBox = positionsBefore.get(name);
            return nodeBox && isCoveredBy(stickyBoxBefore, nodeBox);
        });
        if (coveredNames.length === 0)
            continue;
        const coveredBoxesAfter = coveredNames
            .map((name) => positionsAfter.get(name))
            .filter((box) => box !== undefined);
        if (coveredBoxesAfter.length === 0)
            continue;
        const coveredAfter = compositeBoundingBox(coveredBoxesAfter);
        const newX = centerHorizontally(coveredAfter, stickyBoxBefore);
        const newY = coveredAfter.y + coveredAfter.height - stickyBoxBefore.height + constants_1.STICKY_BOTTOM_PADDING;
        result.set(stickyName, [snapToGrid(newX), snapToGrid(newY)]);
    }
}
function calculateNodePositionsDagre(nodes) {
    const positions = new Map();
    if (nodes.size === 0)
        return positions;
    const aiParentNames = getAiParentNames(nodes);
    const aiConfigNames = getAiConfigNames(nodes);
    const stickyNames = [];
    const nonStickyNames = [];
    for (const [name, graphNode] of nodes) {
        if (graphNode.instance.type === constants_1.STICKY_NODE_TYPE) {
            stickyNames.push(name);
        }
        else {
            nonStickyNames.push(name);
        }
    }
    if (nonStickyNames.length === 0)
        return positions;
    const needsLayout = nonStickyNames.some((name) => {
        const node = nodes.get(name);
        return node && !node.instance.config?.position;
    });
    if (!needsLayout)
        return positions;
    const parentGraph = new dagre_1.default.graphlib.Graph();
    parentGraph.setGraph({});
    parentGraph.setDefaultEdgeLabel(() => ({}));
    for (const name of nonStickyNames) {
        const { width, height } = getNodeDimensions(name, aiParentNames, aiConfigNames, nodes);
        parentGraph.setNode(name, { width, height });
    }
    const nonStickySet = new Set(nonStickyNames);
    for (const name of nonStickyNames) {
        const graphNode = nodes.get(name);
        for (const [, outputMap] of graphNode.connections) {
            for (const targets of outputMap.values()) {
                for (const target of targets) {
                    if (nonStickySet.has(target.node)) {
                        parentGraph.setEdge(name, target.node);
                    }
                }
            }
        }
    }
    const components = dagre_1.default.graphlib.alg.components(parentGraph);
    const subgraphs = components.map((nodeIds) => {
        const subgraph = createSubGraph(nodeIds, parentGraph);
        const aiParentsInSubgraph = subgraph.nodes().filter((id) => aiParentNames.has(id));
        const aiGraphs = aiParentsInSubgraph.map((aiParentId) => {
            const configNodeIds = getAllConnectedAiConfigNodes(subgraph, aiParentId, aiConfigNames);
            const allAiNodeIds = configNodeIds.concat(aiParentId);
            const aiGraph = createAiSubGraph(subgraph, allAiNodeIds);
            const configNodeIdSet = new Set(configNodeIds);
            const rootEdges = subgraph
                .edges()
                .filter((edge) => (edge.v === aiParentId || edge.w === aiParentId) &&
                !configNodeIdSet.has(edge.v) &&
                !configNodeIdSet.has(edge.w));
            configNodeIds.forEach((id) => subgraph.removeNode(id));
            dagre_1.default.layout(aiGraph, { disableOptimalOrderHeuristic: true });
            const aiBoundingBox = boundingBoxFromGraph(aiGraph);
            subgraph.setNode(aiParentId, {
                width: aiBoundingBox.width,
                height: aiBoundingBox.height,
            });
            rootEdges.forEach((edge) => subgraph.setEdge(edge));
            return { graph: aiGraph, boundingBox: aiBoundingBox, aiParentId };
        });
        dagre_1.default.layout(subgraph, { disableOptimalOrderHeuristic: true });
        return { graph: subgraph, aiGraphs, boundingBox: boundingBoxFromGraph(subgraph) };
    });
    let compositeGraph;
    if (subgraphs.length > 1) {
        compositeGraph = createVerticalGraph(subgraphs.map(({ boundingBox }, index) => ({
            box: boundingBox,
            id: index.toString(),
        })));
        dagre_1.default.layout(compositeGraph);
    }
    const boundingBoxByNodeId = {};
    subgraphs.forEach(({ graph, aiGraphs }, index) => {
        let offset = { x: 0, y: 0 };
        if (compositeGraph) {
            const subgraphPosition = compositeGraph.node(index.toString());
            offset = {
                x: 0,
                y: subgraphPosition.y - subgraphPosition.height / 2,
            };
        }
        const aiParentIds = new Set(aiGraphs.map(({ aiParentId }) => aiParentId));
        for (const nodeId of graph.nodes()) {
            const { x, y, width, height } = graph.node(nodeId);
            const box = {
                x: x + offset.x - width / 2,
                y: y + offset.y - height / 2,
                width,
                height,
            };
            if (aiParentIds.has(nodeId)) {
                const aiGraphInfo = aiGraphs.find(({ aiParentId }) => aiParentId === nodeId);
                if (!aiGraphInfo)
                    continue;
                const parentOffset = { x: box.x, y: box.y };
                for (const aiNodeId of aiGraphInfo.graph.nodes()) {
                    const aiNode = aiGraphInfo.graph.node(aiNodeId);
                    boundingBoxByNodeId[aiNodeId] = {
                        x: aiNode.x + parentOffset.x - aiNode.width / 2,
                        y: aiNode.y + parentOffset.y - aiNode.height / 2,
                        width: aiNode.width,
                        height: aiNode.height,
                    };
                }
            }
            else {
                boundingBoxByNodeId[nodeId] = box;
            }
        }
    });
    subgraphs
        .flatMap(({ aiGraphs }) => aiGraphs)
        .forEach(({ graph }) => {
        const aiNodes = graph.nodes();
        const boxes = aiNodes
            .map((id) => boundingBoxByNodeId[id])
            .filter((b) => b !== undefined);
        if (boxes.length === 0)
            return;
        const aiGraphBoundingBox = compositeBoundingBox(boxes);
        const aiNodeVerticalCorrection = aiGraphBoundingBox.height / 2 - constants_1.DEFAULT_NODE_SIZE[0] / 2;
        aiGraphBoundingBox.y += aiNodeVerticalCorrection;
        const hasConflictingNodes = Object.entries(boundingBoxByNodeId)
            .filter(([id]) => !graph.hasNode(id))
            .some(([, nodeBoundingBox]) => intersects(aiGraphBoundingBox, nodeBoundingBox, constants_1.NODE_Y_SPACING));
        if (!hasConflictingNodes) {
            for (const aiNode of aiNodes) {
                if (boundingBoxByNodeId[aiNode]) {
                    boundingBoxByNodeId[aiNode].y += aiNodeVerticalCorrection;
                }
            }
        }
    });
    for (const [name, box] of Object.entries(boundingBoxByNodeId)) {
        const node = nodes.get(name);
        if (node && !node.instance.config?.position) {
            positions.set(name, [snapToGrid(box.x), snapToGrid(box.y)]);
        }
    }
    if (stickyNames.length > 0) {
        const positionsBefore = new Map();
        for (const [name, graphNode] of nodes) {
            const pos = graphNode.instance.config?.position;
            const { width, height } = getNodeDimensions(name, aiParentNames, aiConfigNames, nodes);
            positionsBefore.set(name, {
                x: pos ? pos[0] : 0,
                y: pos ? pos[1] : 0,
                width,
                height,
            });
        }
        const positionsAfter = new Map();
        for (const [name, box] of Object.entries(boundingBoxByNodeId)) {
            positionsAfter.set(name, box);
        }
        repositionStickyNotes(stickyNames, nonStickyNames, positionsBefore, positionsAfter, positions);
    }
    return positions;
}
function layoutWorkflowJSON(json) {
    const jsonNodes = json.nodes;
    if (!jsonNodes || jsonNodes.length === 0)
        return json;
    const connections = json.connections ?? {};
    const graphNodes = new Map();
    for (const node of jsonNodes) {
        if (!node.name)
            continue;
        const connectionsMap = new Map();
        connectionsMap.set('main', new Map());
        graphNodes.set(node.name, {
            instance: {
                type: node.type,
                name: node.name,
                version: node.typeVersion,
                config: {},
            },
            connections: connectionsMap,
        });
    }
    for (const [sourceName, nodeConns] of Object.entries(connections)) {
        const graphNode = graphNodes.get(sourceName);
        if (!graphNode)
            continue;
        for (const [connType, outputs] of Object.entries(nodeConns)) {
            if (!Array.isArray(outputs))
                continue;
            let outputMap = graphNode.connections.get(connType);
            if (!outputMap) {
                outputMap = new Map();
                graphNode.connections.set(connType, outputMap);
            }
            for (let outputIdx = 0; outputIdx < outputs.length; outputIdx++) {
                const slot = outputs[outputIdx];
                if (!Array.isArray(slot))
                    continue;
                const targets = slot
                    .filter((t) => !!t?.node)
                    .map((t) => ({ node: t.node, type: t.type, index: t.index }));
                if (targets.length > 0) {
                    outputMap.set(outputIdx, targets);
                }
            }
        }
    }
    const positions = calculateNodePositionsDagre(graphNodes);
    return {
        ...json,
        nodes: jsonNodes.map((node) => {
            const pos = node.name ? positions.get(node.name) : undefined;
            return pos ? { ...node, position: pos } : node;
        }),
    };
}
