// @n8n/workflow-sdk/dist/workflow-builder/string-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JS_METHODS = void 0;
exports.filterMethodsFromPath = filterMethodsFromPath;
exports.parseVersion = parseVersion;
exports.isPlaceholderValue = isPlaceholderValue;
exports.isResourceLocatorLike = isResourceLocatorLike;
exports.normalizeResourceLocators = normalizeResourceLocators;
exports.escapeNewlinesInStringLiterals = escapeNewlinesInStringLiterals;
exports.escapeNewlinesInExpressionStrings = escapeNewlinesInExpressionStrings;
exports.generateDeterministicNodeId = generateDeterministicNodeId;
const crypto_1 = require("crypto");
exports.JS_METHODS = new Set([
    'includes',
    'indexOf',
    'slice',
    'substring',
    'toLowerCase',
    'toUpperCase',
    'trim',
    'split',
    'replace',
    'match',
    'startsWith',
    'endsWith',
    'filter',
    'map',
    'reduce',
    'find',
    'findIndex',
    'some',
    'every',
    'forEach',
    'join',
    'sort',
    'push',
    'pop',
    'length',
    'toString',
]);
function filterMethodsFromPath(fieldPath) {
    const result = [...fieldPath];
    while (result.length > 0 && exports.JS_METHODS.has(result[result.length - 1])) {
        result.pop();
    }
    return result;
}
function parseVersion(version) {
    if (typeof version === 'number')
        return version;
    if (!version)
        return 1;
    const match = version.match(/v?(\d+(?:\.\d+)?)/);
    return match ? parseFloat(match[1]) : 1;
}
function isPlaceholderValue(value) {
    if (typeof value !== 'string')
        return false;
    return value.startsWith('<__PLACEHOLDER_VALUE__') && value.endsWith('__>');
}
function isResourceLocatorLike(obj) {
    if (typeof obj !== 'object' || obj === null || Array.isArray(obj)) {
        return false;
    }
    const record = obj;
    if (!('mode' in record)) {
        return false;
    }
    if (!('value' in record)) {
        return false;
    }
    return true;
}
function normalizeResourceLocators(params) {
    if (typeof params !== 'object' || params === null) {
        return params;
    }
    if (Array.isArray(params)) {
        return params.map((item) => normalizeResourceLocators(item));
    }
    const result = {};
    const record = params;
    for (const [key, value] of Object.entries(record)) {
        if (isResourceLocatorLike(value)) {
            const rlValue = value;
            const normalizedInner = normalizeResourceLocators(rlValue);
            if (rlValue.mode === 'list' && isPlaceholderValue(rlValue.value)) {
                result[key] = {
                    __rl: true,
                    ...normalizedInner,
                    value: '',
                };
            }
            else {
                result[key] = {
                    __rl: true,
                    ...normalizedInner,
                };
            }
        }
        else if (typeof value === 'object' && value !== null) {
            result[key] = normalizeResourceLocators(value);
        }
        else {
            result[key] = value;
        }
    }
    return result;
}
function couldBeRegexStart(code, i) {
    let j = i - 1;
    while (j >= 0 && /\s/.test(code[j])) {
        j--;
    }
    if (j < 0)
        return true;
    const prevChar = code[j];
    const regexPreceders = '(,=:[!&|?;{}><%+-*/^~';
    return regexPreceders.includes(prevChar);
}
function escapeNewlinesInStringLiterals(code) {
    let result = '';
    let i = 0;
    while (i < code.length) {
        const char = code[i];
        if (char === '`') {
            const start = i;
            i++;
            while (i < code.length) {
                if (code[i] === '\\' && i + 1 < code.length) {
                    i += 2;
                }
                else if (code[i] === '`') {
                    i++;
                    break;
                }
                else {
                    i++;
                }
            }
            result += code.slice(start, i);
            continue;
        }
        if (char === '/' && couldBeRegexStart(code, i)) {
            const next = code[i + 1];
            if (next !== '/' && next !== '*') {
                const start = i;
                i++;
                while (i < code.length) {
                    if (code[i] === '\\' && i + 1 < code.length) {
                        i += 2;
                    }
                    else if (code[i] === '/') {
                        i++;
                        while (i < code.length && /[gimsuy]/.test(code[i])) {
                            i++;
                        }
                        break;
                    }
                    else if (code[i] === '\n') {
                        break;
                    }
                    else {
                        i++;
                    }
                }
                result += code.slice(start, i);
                continue;
            }
        }
        if (char === '"' || char === "'") {
            const quote = char;
            result += char;
            i++;
            while (i < code.length) {
                const c = code[i];
                if (c === '\\' && i + 1 < code.length) {
                    result += c + code[i + 1];
                    i += 2;
                }
                else if (c === quote) {
                    result += c;
                    i++;
                    break;
                }
                else if (c === '\n') {
                    result += '\\n';
                    i++;
                }
                else {
                    result += c;
                    i++;
                }
            }
            continue;
        }
        result += char;
        i++;
    }
    return result;
}
function escapeNewlinesInExpressionStrings(value) {
    if (typeof value === 'string') {
        if (!value.startsWith('=')) {
            return value;
        }
        return value.replace(/\{\{([\s\S]*?)\}\}/g, (_match, inner) => {
            const escaped = escapeNewlinesInStringLiterals(inner);
            return `{{${escaped}}}`;
        });
    }
    if (Array.isArray(value)) {
        return value.map(escapeNewlinesInExpressionStrings);
    }
    if (typeof value === 'object' && value !== null) {
        const result = {};
        for (const [key, val] of Object.entries(value)) {
            result[key] = escapeNewlinesInExpressionStrings(val);
        }
        return result;
    }
    return value;
}
function generateDeterministicNodeId(workflowId, nodeType, nodeName) {
    const hash = (0, crypto_1.createHash)('sha256')
        .update(`${workflowId}:${nodeType}:${nodeName}`)
        .digest('hex')
        .slice(0, 32);
    return [
        hash.slice(0, 8),
        hash.slice(8, 12),
        '4' + hash.slice(13, 16),
        ((parseInt(hash[16], 16) & 0x3) | 0x8).toString(16) + hash.slice(17, 20),
        hash.slice(20, 32),
    ].join('-');
}
