// @n8n/workflow-sdk/dist/workflow-builder/validation-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TOOLS_WITHOUT_PARAMETERS = void 0;
exports.containsExpression = containsExpression;
exports.containsMalformedExpression = containsMalformedExpression;
exports.isSensitiveHeader = isSensitiveHeader;
exports.isCredentialFieldName = isCredentialFieldName;
exports.isToolNode = isToolNode;
exports.containsFromAI = containsFromAI;
exports.isTriggerNode = isTriggerNode;
exports.findMissingExpressionPrefixes = findMissingExpressionPrefixes;
exports.hasLuxonToISOStringMisuse = hasLuxonToISOStringMisuse;
exports.findInvalidDateMethods = findInvalidDateMethods;
exports.extractExpressions = extractExpressions;
exports.parseExpression = parseExpression;
exports.hasPath = hasPath;
const trigger_detection_1 = require("../utils/trigger-detection");
function containsExpression(value) {
    if (typeof value !== 'string') {
        return false;
    }
    return value.includes('={{') || value.startsWith('=');
}
function containsMalformedExpression(value) {
    if (typeof value !== 'string') {
        return false;
    }
    return !value.startsWith('=') && value.includes('{{ $');
}
function isSensitiveHeader(name) {
    const sensitiveHeaders = new Set([
        'authorization',
        'x-api-key',
        'x-auth-token',
        'x-access-token',
        'api-key',
        'apikey',
    ]);
    return sensitiveHeaders.has(name.toLowerCase());
}
function isCredentialFieldName(name) {
    const patterns = [
        /api[_-]?key/i,
        /access[_-]?token/i,
        /auth[_-]?token/i,
        /bearer[_-]?token/i,
        /secret[_-]?key/i,
        /private[_-]?key/i,
        /client[_-]?secret/i,
        /password/i,
        /credentials?/i,
        /^token$/i,
        /^secret$/i,
        /^auth$/i,
    ];
    return patterns.some((pattern) => pattern.test(name));
}
function isToolNode(type) {
    return type.includes('tool') || type.includes('Tool');
}
exports.TOOLS_WITHOUT_PARAMETERS = new Set([
    '@n8n/n8n-nodes-langchain.toolCalculator',
    '@n8n/n8n-nodes-langchain.toolVectorStore',
    '@n8n/n8n-nodes-langchain.vectorStoreInMemory',
    '@n8n/n8n-nodes-langchain.mcpClientTool',
    '@n8n/n8n-nodes-langchain.toolWikipedia',
    '@n8n/n8n-nodes-langchain.toolSerpApi',
]);
function containsFromAI(value) {
    if (typeof value === 'string') {
        return value.includes('$fromAI');
    }
    if (Array.isArray(value)) {
        return value.some((item) => containsFromAI(item));
    }
    if (typeof value === 'object' && value !== null) {
        return Object.values(value).some((v) => containsFromAI(v));
    }
    return false;
}
function isTriggerNode(type) {
    return (0, trigger_detection_1.isTriggerNodeType)(type);
}
function findMissingExpressionPrefixes(value, path = '') {
    const issues = [];
    if (typeof value === 'string') {
        if (!value.startsWith('=') && value.includes('{{ $')) {
            issues.push({ path, value });
        }
    }
    else if (Array.isArray(value)) {
        value.forEach((item, index) => {
            issues.push(...findMissingExpressionPrefixes(item, `${path}[${index}]`));
        });
    }
    else if (value && typeof value === 'object') {
        if ('__placeholder' in value && value.__placeholder) {
            return issues;
        }
        for (const [key, val] of Object.entries(value)) {
            const newPath = path ? `${path}.${key}` : key;
            issues.push(...findMissingExpressionPrefixes(val, newPath));
        }
    }
    return issues;
}
function hasLuxonToISOStringMisuse(value) {
    if (!value.includes('.toISOString()')) {
        return false;
    }
    const luxonPatterns = [
        /\$now\b[^;]*\.toISOString\(\)/,
        /\$today\b[^;]*\.toISOString\(\)/,
        /DateTime\s*\.\s*(now|local|utc|fromISO|fromJSDate)\s*\([^)]*\)[^;]*\.toISOString\(\)/,
    ];
    return luxonPatterns.some((pattern) => pattern.test(value));
}
function findInvalidDateMethods(value, path = '') {
    const issues = [];
    if (typeof value === 'string') {
        if (hasLuxonToISOStringMisuse(value)) {
            issues.push({ path, value });
        }
    }
    else if (Array.isArray(value)) {
        value.forEach((item, index) => {
            issues.push(...findInvalidDateMethods(item, `${path}[${index}]`));
        });
    }
    else if (value && typeof value === 'object') {
        if ('__placeholder' in value && value.__placeholder) {
            return issues;
        }
        for (const [key, val] of Object.entries(value)) {
            const newPath = path ? `${path}.${key}` : key;
            issues.push(...findInvalidDateMethods(val, newPath));
        }
    }
    return issues;
}
function extractExpressions(params) {
    const results = [];
    const recurse = (value, path) => {
        if (typeof value === 'string') {
            if (value.startsWith('=')) {
                results.push({ expression: value, path });
            }
        }
        else if (Array.isArray(value)) {
            value.forEach((item, index) => {
                recurse(item, `${path}[${index}]`);
            });
        }
        else if (value && typeof value === 'object') {
            for (const [key, val] of Object.entries(value)) {
                const newPath = path ? `${path}.${key}` : key;
                recurse(val, newPath);
            }
        }
    };
    recurse(params, '');
    return results;
}
function parseExpression(expr) {
    const jsonMatch = expr.match(/\$json\.([a-zA-Z_][a-zA-Z0-9_]*(?:\.[a-zA-Z_][a-zA-Z0-9_]*)*)/);
    if (jsonMatch) {
        return {
            type: '$json',
            fieldPath: jsonMatch[1].split('.'),
        };
    }
    const nodeMatch = expr.match(/\$\(['"]([^'"]+)['"]\)\.item\.json\.([a-zA-Z_][a-zA-Z0-9_]*(?:\.[a-zA-Z_][a-zA-Z0-9_]*)*)/);
    if (nodeMatch) {
        return {
            type: '$node',
            nodeName: nodeMatch[1],
            fieldPath: nodeMatch[2].split('.'),
        };
    }
    return { type: 'other', fieldPath: [] };
}
function hasPath(shape, path) {
    let current = shape;
    for (const key of path) {
        if (current === null || current === undefined || typeof current !== 'object') {
            return false;
        }
        if (!(key in current)) {
            return false;
        }
        current = current[key];
    }
    return true;
}
