// @n8n/workflow-sdk/dist/workflow-builder/workflow-import.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseWorkflowJSON = parseWorkflowJSON;
const n8n_workflow_1 = require("n8n-workflow");
const base_1 = require("../types/base");
function parseWorkflowJSON(json) {
    const nodes = new Map();
    const nameToKey = new Map();
    let unnamedCounter = 0;
    for (const rawNode of json.nodes) {
        const n8nNode = { ...rawNode };
        if (typeof n8nNode.typeVersion === 'string') {
            n8nNode.typeVersion = Number(n8nNode.typeVersion);
        }
        const version = `v${n8nNode.typeVersion}`;
        const credentials = n8nNode.credentials
            ? (0, n8n_workflow_1.deepCopy)(n8nNode.credentials)
            : undefined;
        const nodeName = n8nNode.name ?? n8nNode.id;
        const instance = {
            type: n8nNode.type,
            version,
            id: n8nNode.id,
            name: nodeName,
            config: {
                name: nodeName,
                parameters: n8nNode.parameters,
                credentials,
                ...{ _originalName: n8nNode.name },
                position: n8nNode.position,
                webhookId: n8nNode.webhookId,
                disabled: n8nNode.disabled,
                notes: n8nNode.notes,
                notesInFlow: n8nNode.notesInFlow,
                executeOnce: n8nNode.executeOnce,
                retryOnFail: n8nNode.retryOnFail,
                alwaysOutputData: n8nNode.alwaysOutputData,
                onError: n8nNode.onError,
            },
            update(config) {
                return { ...this, config: { ...this.config, ...config } };
            },
            to() {
                throw new Error('Nodes from fromJSON() do not support to()');
            },
            input() {
                throw new Error('Nodes from fromJSON() do not support input()');
            },
            output() {
                throw new Error('Nodes from fromJSON() do not support output()');
            },
            onError() {
                throw new Error('Nodes from fromJSON() do not support onError()');
            },
            getConnections() {
                return [];
            },
        };
        const connectionsMap = new Map();
        let mapKey = nodeName || `__unnamed_${unnamedCounter++}`;
        if (nodes.has(mapKey)) {
            mapKey = (0, base_1.generateUniqueName)(nodeName, (n) => nodes.has(n));
        }
        else {
            nameToKey.set(nodeName, mapKey);
        }
        nodes.set(mapKey, {
            instance,
            connections: connectionsMap,
        });
    }
    if (json.connections) {
        const connections = (0, n8n_workflow_1.deepCopy)(json.connections);
        (0, base_1.normalizeConnections)(connections);
        for (const [sourceName, nodeConns] of Object.entries(connections)) {
            const mapKey = nameToKey.get(sourceName);
            const graphNode = mapKey ? nodes.get(mapKey) : undefined;
            if (!graphNode)
                continue;
            for (const [connType, outputs] of Object.entries(nodeConns)) {
                if (!outputs || !Array.isArray(outputs))
                    continue;
                const typeMap = graphNode.connections.get(connType) ?? new Map();
                for (let outputIndex = 0; outputIndex < outputs.length; outputIndex++) {
                    const targets = outputs[outputIndex];
                    if (targets && targets.length > 0) {
                        typeMap.set(outputIndex, targets.map((conn) => ({
                            node: conn.node,
                            type: conn.type,
                            index: conn.index,
                        })));
                    }
                    else {
                        typeMap.set(outputIndex, []);
                    }
                }
                graphNode.connections.set(connType, typeMap);
            }
        }
    }
    let lastNode = null;
    for (const name of nodes.keys()) {
        lastNode = name;
    }
    return {
        id: json.id ?? '',
        name: json.name,
        settings: json.settings,
        nodes,
        lastNode,
        pinData: json.pinData,
        meta: json.meta,
    };
}
