// n8n-workflow/dist/cjs/native-methods/boolean.methods.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.booleanMethods = void 0;
    exports.booleanMethods = {
        typeName: 'Boolean',
        functions: {
            toString: {
                doc: {
                    name: 'toString',
                    description: "Converts <code>true</code> to the string <code>'true'</code> and <code>false</code> to the string <code>'false'</code>.",
                    examples: [
                        { example: 'true.toString()', evaluated: "'true'" },
                        { example: 'false.toString()', evaluated: "'false'" },
                    ],
                    docURL: 'https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean/toString',
                    returnType: 'string',
                },
            },
        },
    };
});
