// @n8n/permissions/dist/roles/all-roles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isBuiltInRole = exports.ALL_ROLES = void 0;
const constants_ee_1 = require("../constants.ee");
const role_maps_ee_1 = require("./role-maps.ee");
const get_role_scopes_ee_1 = require("../utilities/get-role-scopes.ee");
const ROLE_NAMES = {
    'global:owner': 'Owner',
    'global:admin': 'Admin',
    'global:member': 'Member',
    'global:chatUser': 'Chat User',
    [constants_ee_1.PROJECT_OWNER_ROLE_SLUG]: 'Project Owner',
    [constants_ee_1.PROJECT_ADMIN_ROLE_SLUG]: 'Project Admin',
    [constants_ee_1.PROJECT_EDITOR_ROLE_SLUG]: 'Project Editor',
    [constants_ee_1.PROJECT_VIEWER_ROLE_SLUG]: 'Project Viewer',
    [constants_ee_1.PROJECT_CHAT_USER_ROLE_SLUG]: 'Project Chat User',
    'credential:user': 'Credential User',
    'credential:owner': 'Credential Owner',
    'workflow:owner': 'Workflow Owner',
    'workflow:editor': 'Workflow Editor',
    'secretsProviderConnection:owner': 'Secrets Provider Connection Owner',
    'secretsProviderConnection:user': 'Secrets Provider Connection User',
};
const ROLE_DESCRIPTIONS = {
    'global:owner': 'Owner',
    'global:admin': 'Admin',
    'global:member': 'Member',
    'global:chatUser': 'Chat User',
    [constants_ee_1.PROJECT_OWNER_ROLE_SLUG]: 'Project Owner',
    [constants_ee_1.PROJECT_ADMIN_ROLE_SLUG]: 'Full control of settings, members, workflows, credentials and executions',
    [constants_ee_1.PROJECT_EDITOR_ROLE_SLUG]: 'Create, edit, and delete workflows, credentials, and executions',
    [constants_ee_1.PROJECT_VIEWER_ROLE_SLUG]: 'Read-only access to workflows, credentials, and executions',
    [constants_ee_1.PROJECT_CHAT_USER_ROLE_SLUG]: 'Chat-only access to chatting with workflows that have n8n Chat enabled',
    'credential:user': 'Credential User',
    'credential:owner': 'Credential Owner',
    'workflow:owner': 'Workflow Owner',
    'workflow:editor': 'Workflow Editor',
    'secretsProviderConnection:owner': 'Full control of secrets provider connection settings and secrets',
    'secretsProviderConnection:user': 'Read-only access to use secrets from the connection',
};
const mapToRoleObject = (roles, roleType) => Object.keys(roles).map((role) => ({
    slug: role,
    displayName: ROLE_NAMES[role],
    scopes: (0, get_role_scopes_ee_1.getRoleScopes)(role),
    description: ROLE_DESCRIPTIONS[role],
    licensed: false,
    systemRole: true,
    roleType,
}));
exports.ALL_ROLES = Object.freeze({
    global: mapToRoleObject(role_maps_ee_1.GLOBAL_SCOPE_MAP, 'global'),
    project: mapToRoleObject(role_maps_ee_1.PROJECT_SCOPE_MAP, 'project'),
    credential: mapToRoleObject(role_maps_ee_1.CREDENTIALS_SHARING_SCOPE_MAP, 'credential'),
    workflow: mapToRoleObject(role_maps_ee_1.WORKFLOW_SHARING_SCOPE_MAP, 'workflow'),
    secretsProviderConnection: mapToRoleObject(role_maps_ee_1.SECRETS_PROVIDER_CONNECTION_SHARING_SCOPE_MAP, 'secretsProviderConnection'),
});
const isBuiltInRole = (role) => {
    return Object.prototype.hasOwnProperty.call(ROLE_NAMES, role);
};
exports.isBuiltInRole = isBuiltInRole;
