// @n8n/config/dist/configs/instance-ai.config.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstanceAiConfig = void 0;
const decorators_1 = require("../decorators");
let InstanceAiConfig = class InstanceAiConfig {
    constructor() {
        this.model = 'anthropic/claude-sonnet-4-6';
        this.modelUrl = '';
        this.modelApiKey = '';
        this.maxContextWindowTokens = 500_000;
        this.mcpServers = '';
        this.lastMessages = 20;
        this.embedderModel = '';
        this.semanticRecallTopK = 5;
        this.subAgentMaxSteps = 100;
        this.localGatewayDisabled = false;
        this.browserMcp = false;
        this.sandboxEnabled = false;
        this.sandboxProvider = 'daytona';
        this.daytonaApiUrl = '';
        this.daytonaApiKey = '';
        this.n8nSandboxServiceUrl = '';
        this.n8nSandboxServiceApiKey = '';
        this.sandboxImage = 'daytonaio/sandbox:0.5.0';
        this.sandboxTimeout = 300_000;
        this.braveSearchApiKey = '';
        this.searxngUrl = '';
        this.filesystemPath = '';
        this.gatewayApiKey = '';
        this.threadTtlDays = 90;
        this.snapshotPruneInterval = 60 * 60 * 1000;
        this.snapshotRetention = 24 * 60 * 60 * 1000;
        this.confirmationTimeout = 10 * 60 * 1000;
    }
};
exports.InstanceAiConfig = InstanceAiConfig;
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_MODEL'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "model", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_MODEL_URL'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "modelUrl", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_MODEL_API_KEY'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "modelApiKey", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_MAX_CONTEXT_WINDOW_TOKENS'),
    __metadata("design:type", Number)
], InstanceAiConfig.prototype, "maxContextWindowTokens", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_MCP_SERVERS'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "mcpServers", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_LAST_MESSAGES'),
    __metadata("design:type", Number)
], InstanceAiConfig.prototype, "lastMessages", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_EMBEDDER_MODEL'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "embedderModel", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_SEMANTIC_RECALL_TOP_K'),
    __metadata("design:type", Number)
], InstanceAiConfig.prototype, "semanticRecallTopK", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_SUB_AGENT_MAX_STEPS'),
    __metadata("design:type", Number)
], InstanceAiConfig.prototype, "subAgentMaxSteps", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_LOCAL_GATEWAY_DISABLED'),
    __metadata("design:type", Boolean)
], InstanceAiConfig.prototype, "localGatewayDisabled", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_BROWSER_MCP'),
    __metadata("design:type", Boolean)
], InstanceAiConfig.prototype, "browserMcp", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_SANDBOX_ENABLED'),
    __metadata("design:type", Boolean)
], InstanceAiConfig.prototype, "sandboxEnabled", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_SANDBOX_PROVIDER'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "sandboxProvider", void 0);
__decorate([
    (0, decorators_1.Env)('DAYTONA_API_URL'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "daytonaApiUrl", void 0);
__decorate([
    (0, decorators_1.Env)('DAYTONA_API_KEY'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "daytonaApiKey", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SANDBOX_SERVICE_URL'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "n8nSandboxServiceUrl", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SANDBOX_SERVICE_API_KEY'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "n8nSandboxServiceApiKey", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_SANDBOX_IMAGE'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "sandboxImage", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_SANDBOX_TIMEOUT'),
    __metadata("design:type", Number)
], InstanceAiConfig.prototype, "sandboxTimeout", void 0);
__decorate([
    (0, decorators_1.Env)('INSTANCE_AI_BRAVE_SEARCH_API_KEY'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "braveSearchApiKey", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_SEARXNG_URL'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "searxngUrl", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_FILESYSTEM_PATH'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "filesystemPath", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_GATEWAY_API_KEY'),
    __metadata("design:type", String)
], InstanceAiConfig.prototype, "gatewayApiKey", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_THREAD_TTL_DAYS'),
    __metadata("design:type", Number)
], InstanceAiConfig.prototype, "threadTtlDays", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_SNAPSHOT_PRUNE_INTERVAL'),
    __metadata("design:type", Number)
], InstanceAiConfig.prototype, "snapshotPruneInterval", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_SNAPSHOT_RETENTION'),
    __metadata("design:type", Number)
], InstanceAiConfig.prototype, "snapshotRetention", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_AI_CONFIRMATION_TIMEOUT'),
    __metadata("design:type", Number)
], InstanceAiConfig.prototype, "confirmationTimeout", void 0);
exports.InstanceAiConfig = InstanceAiConfig = __decorate([
    decorators_1.Config
], InstanceAiConfig);
