// @n8n/config/dist/configs/security.config.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityConfig = void 0;
const zod_1 = __importDefault(require("zod"));
const decorators_1 = require("../decorators");
const crossOriginOpenerPolicySchema = zod_1.default.enum(['same-origin', 'same-origin-allow-popups']);
let SecurityConfig = class SecurityConfig {
    constructor() {
        this.restrictFileAccessTo = '~/.n8n-files';
        this.blockFileAccessToN8nFiles = true;
        this.blockFilePatterns = '^(.*\\/)*\\.git(\\/.*)*$';
        this.daysAbandonedWorkflow = 90;
        this.contentSecurityPolicy = '{}';
        this.contentSecurityPolicyReportOnly = false;
        this.crossOriginOpenerPolicy = 'same-origin';
        this.disableWebhookHtmlSandboxing = false;
        this.disableFormHtmlSandboxing = false;
        this.disableBareRepos = true;
        this.awsSystemCredentialsAccess = false;
        this.enableGitNodeHooks = false;
        this.enableGitNodeAllConfigKeys = false;
    }
};
exports.SecurityConfig = SecurityConfig;
__decorate([
    (0, decorators_1.Env)('N8N_RESTRICT_FILE_ACCESS_TO'),
    __metadata("design:type", String)
], SecurityConfig.prototype, "restrictFileAccessTo", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_BLOCK_FILE_ACCESS_TO_N8N_FILES'),
    __metadata("design:type", Boolean)
], SecurityConfig.prototype, "blockFileAccessToN8nFiles", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_BLOCK_FILE_PATTERNS'),
    __metadata("design:type", String)
], SecurityConfig.prototype, "blockFilePatterns", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SECURITY_AUDIT_DAYS_ABANDONED_WORKFLOW'),
    __metadata("design:type", Number)
], SecurityConfig.prototype, "daysAbandonedWorkflow", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_CONTENT_SECURITY_POLICY'),
    __metadata("design:type", String)
], SecurityConfig.prototype, "contentSecurityPolicy", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_CONTENT_SECURITY_POLICY_REPORT_ONLY'),
    __metadata("design:type", Boolean)
], SecurityConfig.prototype, "contentSecurityPolicyReportOnly", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_CROSS_ORIGIN_OPENER_POLICY', crossOriginOpenerPolicySchema),
    __metadata("design:type", void 0)
], SecurityConfig.prototype, "crossOriginOpenerPolicy", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSECURE_DISABLE_WEBHOOK_IFRAME_SANDBOX'),
    __metadata("design:type", Boolean)
], SecurityConfig.prototype, "disableWebhookHtmlSandboxing", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSECURE_DISABLE_FORM_HTML_SANDBOX'),
    __metadata("design:type", Boolean)
], SecurityConfig.prototype, "disableFormHtmlSandboxing", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_GIT_NODE_DISABLE_BARE_REPOS'),
    __metadata("design:type", Boolean)
], SecurityConfig.prototype, "disableBareRepos", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_AWS_SYSTEM_CREDENTIALS_ACCESS_ENABLED'),
    __metadata("design:type", Boolean)
], SecurityConfig.prototype, "awsSystemCredentialsAccess", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_GIT_NODE_ENABLE_HOOKS'),
    __metadata("design:type", Boolean)
], SecurityConfig.prototype, "enableGitNodeHooks", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_GIT_NODE_ENABLE_ALL_CONFIG_KEYS'),
    __metadata("design:type", Boolean)
], SecurityConfig.prototype, "enableGitNodeAllConfigKeys", void 0);
exports.SecurityConfig = SecurityConfig = __decorate([
    decorators_1.Config
], SecurityConfig);
