// n8n-core/dist/observability/tracing/noop-tracing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoopTracing = exports.EmptySpan = void 0;
class EmptySpan {
    constructor() {
        this.emptyContextData = Object.freeze({
            traceId: '00000000000000000000000000000000',
            spanId: '0000000000000000',
            traceFlags: 0,
        });
    }
    spanContext() {
        return this.emptyContextData;
    }
    end() {
    }
    setAttribute() {
        return this;
    }
    setAttributes() {
        return this;
    }
    setStatus() {
        return this;
    }
    updateName() {
        return this;
    }
    isRecording() {
        return false;
    }
    addEvent() {
        return this;
    }
    addLink() {
        return this;
    }
    addLinks() {
        return this;
    }
    recordException() {
    }
}
exports.EmptySpan = EmptySpan;
class NoopTracing {
    constructor() {
        this.emptySpan = new EmptySpan();
    }
    async startSpan(_options, spanCb) {
        return await spanCb(this.emptySpan);
    }
}
exports.NoopTracing = NoopTracing;
