// n8n-core/dist/observability/tracing/tracing.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tracing = exports.SpanStatus = void 0;
const di_1 = require("@n8n/di");
const noop_tracing_1 = require("./noop-tracing");
const COMMON_TRACE_ATTRIBUTES = {
    workflow: {
        id: 'n8n.workflow.id',
        name: 'n8n.workflow.name',
    },
    node: {
        id: 'n8n.node.id',
        name: 'n8n.node.name',
        type: 'n8n.node.type',
        typeVersion: 'n8n.node.type_version',
    },
};
var SpanStatus;
(function (SpanStatus) {
    SpanStatus[SpanStatus["ok"] = 1] = "ok";
    SpanStatus[SpanStatus["error"] = 2] = "error";
})(SpanStatus || (exports.SpanStatus = SpanStatus = {}));
let Tracing = class Tracing {
    constructor() {
        this.tracer = new noop_tracing_1.NoopTracing();
        this.commonAttrs = COMMON_TRACE_ATTRIBUTES;
    }
    setTracingImplementation(tracing) {
        this.tracer = tracing;
    }
    async startSpan(options, spanCb) {
        return await this.tracer.startSpan(options, spanCb);
    }
    pickWorkflowAttributes(workflow) {
        return {
            [this.commonAttrs.workflow.id]: workflow.id,
            [this.commonAttrs.workflow.name]: workflow.name,
        };
    }
    pickNodeAttributes(node) {
        return {
            [this.commonAttrs.node.id]: node.id,
            [this.commonAttrs.node.name]: node.name,
            [this.commonAttrs.node.type]: node.type,
            [this.commonAttrs.node.typeVersion]: node.typeVersion,
        };
    }
};
exports.Tracing = Tracing;
exports.Tracing = Tracing = __decorate([
    (0, di_1.Service)()
], Tracing);
