// @n8n/constants/dist/logstreaming.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOGSTREAMING_CB_DEFAULT_CONCURRENT_HALF_OPEN_REQUESTS = exports.LOGSTREAMING_CB_DEFAULT_FAILURE_WINDOW_MS = exports.LOGSTREAMING_CB_DEFAULT_HALF_OPEN_REQUESTS = exports.LOGSTREAMING_CB_DEFAULT_MAX_FAILURES = exports.LOGSTREAMING_CB_DEFAULT_MAX_DURATION_MS = exports.LOGSTREAMING_DEFAULT_SOCKET_TIMEOUT_MS = exports.LOGSTREAMING_DEFAULT_MAX_TOTAL_SOCKETS = exports.LOGSTREAMING_DEFAULT_MAX_SOCKETS = exports.LOGSTREAMING_DEFAULT_MAX_FREE_SOCKETS = void 0;
const time_1 = require("./time");
exports.LOGSTREAMING_DEFAULT_MAX_FREE_SOCKETS = 5;
exports.LOGSTREAMING_DEFAULT_MAX_SOCKETS = 50;
exports.LOGSTREAMING_DEFAULT_MAX_TOTAL_SOCKETS = 100;
exports.LOGSTREAMING_DEFAULT_SOCKET_TIMEOUT_MS = 5 * time_1.Time.seconds.toMilliseconds;
exports.LOGSTREAMING_CB_DEFAULT_MAX_DURATION_MS = 3 * time_1.Time.minutes.toMilliseconds;
exports.LOGSTREAMING_CB_DEFAULT_MAX_FAILURES = 5;
exports.LOGSTREAMING_CB_DEFAULT_HALF_OPEN_REQUESTS = 2;
exports.LOGSTREAMING_CB_DEFAULT_FAILURE_WINDOW_MS = 1 * time_1.Time.minutes.toMilliseconds;
exports.LOGSTREAMING_CB_DEFAULT_CONCURRENT_HALF_OPEN_REQUESTS = 1;
