// @n8n/workflow-sdk/dist/validation/display-options.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkConditions = checkConditions;
exports.getPropertyValue = getPropertyValue;
exports.matchesDisplayOptions = matchesDisplayOptions;
const get_1 = __importDefault(require("lodash/get"));
const isEqual_1 = __importDefault(require("lodash/isEqual"));
function isDisplayCondition(value) {
    return (value !== null &&
        typeof value === 'object' &&
        '_cnd' in value &&
        Object.keys(value).length === 1);
}
function isRegexPath(path) {
    return path.includes('|') || path.includes('(');
}
function getAllPaths(obj, prefix = '') {
    if (obj === null || obj === undefined || typeof obj !== 'object') {
        return prefix ? [prefix] : [];
    }
    const paths = [];
    if (prefix)
        paths.push(prefix);
    for (const key of Object.keys(obj)) {
        const newPrefix = prefix ? `${prefix}.${key}` : key;
        paths.push(...getAllPaths(obj[key], newPrefix));
    }
    return paths;
}
function getMatchingPathValues(context, regexPath) {
    const rootParams = context.rootParameters ?? context.parameters;
    const pattern = regexPath.startsWith('/') ? regexPath.slice(1) : regexPath;
    const regexPattern = '^' + pattern.replace(/\./g, '\\.').replace(/\\\.\(/g, '.(') + '$';
    const regex = new RegExp(regexPattern);
    const allPaths = getAllPaths(rootParams);
    const matchingValues = [];
    for (const path of allPaths) {
        if (regex.test(path)) {
            const value = (0, get_1.default)(rootParams, path);
            if (value !== undefined) {
                matchingValues.push(Array.isArray(value) ? value : [value]);
            }
        }
    }
    return matchingValues;
}
function checkConditions(conditions, actualValues) {
    return conditions.some((condition) => {
        if (isDisplayCondition(condition)) {
            const [key, targetValue] = Object.entries(condition._cnd)[0];
            if (actualValues.length === 0) {
                if (key === 'not')
                    return true;
                return false;
            }
            return actualValues.every((propertyValue) => {
                if (key === 'eq') {
                    return (0, isEqual_1.default)(propertyValue, targetValue);
                }
                if (key === 'not') {
                    return !(0, isEqual_1.default)(propertyValue, targetValue);
                }
                if (key === 'gte') {
                    return propertyValue >= targetValue;
                }
                if (key === 'lte') {
                    return propertyValue <= targetValue;
                }
                if (key === 'gt') {
                    return propertyValue > targetValue;
                }
                if (key === 'lt') {
                    return propertyValue < targetValue;
                }
                if (key === 'between') {
                    const { from, to } = targetValue;
                    return propertyValue >= from && propertyValue <= to;
                }
                if (key === 'includes') {
                    return propertyValue.includes(targetValue);
                }
                if (key === 'startsWith') {
                    return propertyValue.startsWith(targetValue);
                }
                if (key === 'endsWith') {
                    return propertyValue.endsWith(targetValue);
                }
                if (key === 'regex') {
                    return new RegExp(targetValue).test(propertyValue);
                }
                if (key === 'exists') {
                    return propertyValue !== null && propertyValue !== undefined && propertyValue !== '';
                }
                return false;
            });
        }
        return actualValues.includes(condition);
    });
}
function getPropertyValue(context, propertyName) {
    let value;
    if (propertyName.charAt(0) === '/') {
        const rootParams = context.rootParameters ?? context.parameters;
        value = (0, get_1.default)(rootParams, propertyName.slice(1));
        if (value === undefined && context.defaults) {
            value = (0, get_1.default)(context.defaults, propertyName.slice(1));
        }
    }
    else if (propertyName === '@version') {
        value = context.nodeVersion ?? 0;
    }
    else if (propertyName === '@tool') {
        value = context.isToolNode ?? false;
    }
    else {
        value = (0, get_1.default)(context.parameters, propertyName);
        if (value === undefined && context.defaults) {
            value = (0, get_1.default)(context.defaults, propertyName);
        }
    }
    if (value && typeof value === 'object' && '__rl' in value) {
        const rlValue = value;
        if (rlValue.__rl === true && 'value' in rlValue) {
            value = rlValue.value;
        }
    }
    if (!Array.isArray(value)) {
        return [value];
    }
    return value;
}
function getRawPropertyValue(context, propertyName) {
    if (propertyName.charAt(0) === '/') {
        const rootParams = context.rootParameters ?? context.parameters;
        return (0, get_1.default)(rootParams, propertyName.slice(1));
    }
    if (propertyName === '@version' || propertyName === '@tool') {
        return undefined;
    }
    return (0, get_1.default)(context.parameters, propertyName);
}
function isUnselectedResourceLocator(value) {
    if (!value || typeof value !== 'object' || !('__rl' in value))
        return false;
    const rl = value;
    return rl.__rl === true && rl.value === '';
}
function matchesDisplayOptions(context, displayOptions) {
    const { show, hide } = displayOptions;
    if (show) {
        for (const propertyName of Object.keys(show)) {
            const conditions = show[propertyName];
            if (isRegexPath(propertyName)) {
                const matchingPathValues = getMatchingPathValues(context, propertyName);
                if (matchingPathValues.length === 0) {
                    return false;
                }
                const anyMatch = matchingPathValues.some((values) => {
                    if (values.some((v) => typeof v === 'string' && v.charAt(0) === '=')) {
                        return true;
                    }
                    return checkConditions(conditions, values);
                });
                if (!anyMatch) {
                    return false;
                }
                continue;
            }
            const values = getPropertyValue(context, propertyName);
            if (values.some((v) => typeof v === 'string' && v.charAt(0) === '=')) {
                return true;
            }
            if (!checkConditions(conditions, values)) {
                return false;
            }
        }
    }
    if (hide) {
        for (const propertyName of Object.keys(hide)) {
            if (isUnselectedResourceLocator(getRawPropertyValue(context, propertyName))) {
                continue;
            }
            const values = getPropertyValue(context, propertyName);
            if (values.length !== 0 && checkConditions(hide[propertyName], values)) {
                return false;
            }
        }
    }
    return true;
}
