// @n8n/workflow-sdk/dist/validation/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationWarning = exports.ValidationError = exports.setSchemaBaseDirs = void 0;
exports.validateWorkflow = validateWorkflow;
const n8n_workflow_1 = require("n8n-workflow");
const display_options_1 = require("./display-options");
const input_resolver_1 = require("./input-resolver");
const schema_validator_1 = require("./schema-validator");
const node_types_1 = require("../constants/node-types");
var schema_validator_2 = require("./schema-validator");
Object.defineProperty(exports, "setSchemaBaseDirs", { enumerable: true, get: function () { return schema_validator_2.setSchemaBaseDirs; } });
class ValidationError {
    constructor(code, message, nodeName, parameterName, violationLevel) {
        this.code = code;
        this.message = message;
        this.nodeName = nodeName;
        this.parameterName = parameterName;
        this.violationLevel = violationLevel;
    }
}
exports.ValidationError = ValidationError;
class ValidationWarning {
    constructor(code, message, nodeName, parameterPath, originalName, violationLevel) {
        this.code = code;
        this.message = message;
        this.nodeName = nodeName;
        this.parameterPath = parameterPath;
        this.originalName = originalName;
        this.violationLevel = violationLevel;
    }
}
exports.ValidationWarning = ValidationWarning;
function isTriggerNode(type) {
    return (type.includes('Trigger') ||
        type.includes('trigger') ||
        type.includes('Webhook') ||
        type.includes('webhook') ||
        type.includes('Schedule') ||
        type.includes('schedule') ||
        type.includes('Poll') ||
        type.includes('poll'));
}
const AI_CONNECTION_TYPES = [
    'ai_languageModel',
    'ai_memory',
    'ai_tool',
    'ai_outputParser',
    'ai_embedding',
    'ai_vectorStore',
    'ai_retriever',
    'ai_document',
    'ai_textSplitter',
    'ai_reranker',
];
const AI_CONNECTION_TO_SUBNODE_FIELD = {
    ai_languageModel: 'model',
    ai_memory: 'memory',
    ai_tool: 'tools',
    ai_outputParser: 'outputParser',
    ai_embedding: 'embedding',
    ai_vectorStore: 'vectorStore',
    ai_retriever: 'retriever',
    ai_document: 'documentLoader',
    ai_textSplitter: 'textSplitter',
    ai_reranker: 'reranker',
};
const AI_ARRAY_TYPES = new Set(['ai_tool']);
function reconstructSubnodesFromConnections(targetNodeName, json) {
    const subnodes = {};
    const nodesByName = new Map();
    for (const node of json.nodes) {
        if (node.name) {
            nodesByName.set(node.name, node);
        }
    }
    for (const [sourceNodeName, nodeConnections] of Object.entries(json.connections)) {
        for (const connType of AI_CONNECTION_TYPES) {
            const aiConns = nodeConnections[connType];
            if (!aiConns || !Array.isArray(aiConns))
                continue;
            for (const outputs of aiConns) {
                if (!outputs)
                    continue;
                for (const conn of outputs) {
                    if (conn.node === targetNodeName) {
                        const subnodeField = AI_CONNECTION_TO_SUBNODE_FIELD[connType];
                        if (!subnodeField)
                            continue;
                        const sourceNode = nodesByName.get(sourceNodeName);
                        if (!sourceNode)
                            continue;
                        const subnodeConfig = {
                            type: sourceNode.type,
                            version: sourceNode.typeVersion,
                            parameters: sourceNode.parameters ?? {},
                        };
                        if (AI_ARRAY_TYPES.has(connType)) {
                            const existing = subnodes[subnodeField];
                            if (Array.isArray(existing)) {
                                existing.push(subnodeConfig);
                            }
                            else {
                                subnodes[subnodeField] = [subnodeConfig];
                            }
                        }
                        else {
                            subnodes[subnodeField] = subnodeConfig;
                        }
                    }
                }
            }
        }
    }
    return Object.keys(subnodes).length > 0 ? subnodes : undefined;
}
function hasAiConnectionToParent(nodeName, json) {
    const nodeConnections = json.connections[nodeName];
    if (!nodeConnections)
        return false;
    for (const connType of AI_CONNECTION_TYPES) {
        const aiConns = nodeConnections[connType];
        if (aiConns && Array.isArray(aiConns)) {
            for (const outputs of aiConns) {
                if (outputs && outputs.length > 0) {
                    return true;
                }
            }
        }
    }
    return false;
}
function isToolSubnode(nodeName, json) {
    const nodeConnections = json.connections[nodeName];
    if (!nodeConnections)
        return false;
    const toolConns = nodeConnections.ai_tool;
    if (toolConns && Array.isArray(toolConns)) {
        for (const outputs of toolConns) {
            if (outputs && outputs.length > 0) {
                return true;
            }
        }
    }
    return false;
}
function findDisconnectedNodes(json) {
    const hasIncoming = new Set();
    for (const [_sourceName, nodeConnections] of Object.entries(json.connections)) {
        if (nodeConnections.main) {
            for (const outputs of nodeConnections.main) {
                if (outputs) {
                    for (const connection of outputs) {
                        hasIncoming.add(connection.node);
                    }
                }
            }
        }
    }
    const disconnected = [];
    for (const node of json.nodes) {
        if (!node.name)
            continue;
        if (hasIncoming.has(node.name))
            continue;
        if (isTriggerNode(node.type))
            continue;
        if ((0, node_types_1.isStickyNoteType)(node.type))
            continue;
        if (hasAiConnectionToParent(node.name, json))
            continue;
        disconnected.push(node.name);
    }
    return disconnected;
}
function validateWorkflow(workflowOrJson, options = {}) {
    const json = 'toJSON' in workflowOrJson ? workflowOrJson.toJSON() : workflowOrJson;
    const errors = [];
    const warnings = [];
    if (!options.allowNoTrigger) {
        const hasTrigger = json.nodes.some((node) => isTriggerNode(node.type));
        if (!hasTrigger) {
            warnings.push(new ValidationWarning('MISSING_TRIGGER', 'Workflow has no trigger node. It will need to be started manually.'));
        }
    }
    if (!options.allowDisconnectedNodes) {
        const disconnected = findDisconnectedNodes(json);
        for (const nodeName of disconnected) {
            warnings.push(new ValidationWarning('DISCONNECTED_NODE', `Node '${nodeName}' is not connected to any input. It will not receive data.`, nodeName));
        }
    }
    if (options.strictMode) {
        for (const node of json.nodes) {
            if ((0, node_types_1.isHttpRequestType)(node.type)) {
                if (!node.parameters?.url && !node.parameters?.requestUrl) {
                    warnings.push(new ValidationWarning('MISSING_PARAMETER', `HTTP Request node '${node.name}' may be missing URL parameter`, node.name));
                }
            }
        }
    }
    if (options.validateSchema !== false) {
        for (const node of json.nodes) {
            const version = typeof node.typeVersion === 'string'
                ? parseFloat(node.typeVersion)
                : (node.typeVersion ?? 1);
            const config = {};
            if (node.parameters !== undefined) {
                config.parameters = node.parameters;
            }
            const nodeWithSubnodes = node;
            if (nodeWithSubnodes.subnodes !== undefined) {
                config.subnodes = nodeWithSubnodes.subnodes;
            }
            else if (node.name) {
                const reconstructed = reconstructSubnodesFromConnections(node.name, json);
                if (reconstructed) {
                    config.subnodes = reconstructed;
                }
            }
            const isToolNode = node.name ? isToolSubnode(node.name, json) : false;
            const schemaResult = (0, schema_validator_1.validateNodeConfig)(node.type, version, config, { isToolNode });
            if (!schemaResult.valid) {
                for (const error of schemaResult.errors) {
                    let message = error.message;
                    if (error.path === 'subnodes' &&
                        message.includes('Unknown field') &&
                        options.nodeTypesProvider) {
                        const nodeType = options.nodeTypesProvider.getByNameAndVersion(node.type, version);
                        const validInputs = nodeType?.description?.builderHint?.inputs;
                        if (validInputs) {
                            const validSubnodes = Object.keys(validInputs)
                                .map((k) => AI_CONNECTION_TO_SUBNODE_FIELD[k])
                                .filter(Boolean);
                            if (validSubnodes.length > 0) {
                                message = message.replace(/Unknown field\(s\) at "subnodes": (.+)\./, `Invalid subnode(s) $1. This node only accepts: ${validSubnodes.join(', ')}.`);
                            }
                        }
                    }
                    warnings.push(new ValidationWarning('INVALID_PARAMETER', `Node "${node.name}": ${message}`, node.name));
                }
            }
        }
    }
    if (options.nodeTypesProvider) {
        checkNodeInputIndices(json, options.nodeTypesProvider, warnings);
        validateSubnodeParameters(json, options.nodeTypesProvider, warnings);
        validateParentSupportsInputs(json, options.nodeTypesProvider, warnings);
    }
    return {
        valid: errors.length === 0,
        errors,
        warnings,
    };
}
const AI_CONNECTION_TO_SDK_FUNCTION = {
    ai_languageModel: 'languageModel()',
    ai_memory: 'memory()',
    ai_tool: 'tool()',
    ai_outputParser: 'outputParser()',
    ai_embedding: 'embeddings()',
    ai_vectorStore: 'vectorStore()',
    ai_retriever: 'retriever()',
    ai_document: 'documentLoader()',
    ai_textSplitter: 'textSplitter()',
    ai_reranker: 'reranker()',
};
function checkDisplayOptionsMatch(subnodeParams, displayOptions) {
    const mismatches = [];
    if (!displayOptions.show)
        return { matches: true, mismatches };
    for (const [paramName, expectedValues] of Object.entries(displayOptions.show)) {
        if (!expectedValues)
            continue;
        const actualValue = subnodeParams[paramName];
        if (!expectedValues.includes(actualValue)) {
            mismatches.push({
                param: paramName,
                expected: expectedValues,
                actual: actualValue,
            });
        }
    }
    return { matches: mismatches.length === 0, mismatches };
}
function validateSubnodeParameters(json, nodeTypesProvider, warnings) {
    const nodesByName = new Map();
    for (const node of json.nodes) {
        if (node.name) {
            nodesByName.set(node.name, node);
        }
    }
    const connectionsByDest = (0, n8n_workflow_1.mapConnectionsByDestination)(json.connections);
    for (const parentNode of json.nodes) {
        if (!parentNode.name)
            continue;
        const parentNodeType = nodeTypesProvider.getByNameAndVersion(parentNode.type, typeof parentNode.typeVersion === 'string'
            ? parseFloat(parentNode.typeVersion)
            : (parentNode.typeVersion ?? 1));
        const builderHintInputs = parentNodeType?.description?.builderHint?.inputs;
        if (!builderHintInputs)
            continue;
        for (const [connectionType, inputConfig] of Object.entries(builderHintInputs)) {
            if (!connectionType.startsWith('ai_'))
                continue;
            if (!inputConfig?.displayOptions?.show)
                continue;
            const incomingConnections = connectionsByDest[parentNode.name]?.[connectionType];
            if (!incomingConnections)
                continue;
            for (const connList of incomingConnections) {
                if (!connList)
                    continue;
                for (const conn of connList) {
                    const subnodeName = conn.node;
                    const subnode = nodesByName.get(subnodeName);
                    if (!subnode?.parameters)
                        continue;
                    const { matches, mismatches } = checkDisplayOptionsMatch(subnode.parameters, inputConfig.displayOptions);
                    if (!matches) {
                        const sdkFn = AI_CONNECTION_TO_SDK_FUNCTION[connectionType] || connectionType;
                        const mismatchDetails = mismatches
                            .map((m) => `${m.param}='${String(m.actual)}' (expected: ${m.expected.map((v) => `'${String(v)}'`).join(' or ')})`)
                            .join(', ');
                        warnings.push(new ValidationWarning('SUBNODE_PARAMETER_MISMATCH', `'${subnodeName}' is connected to '${parentNode.name}' using ${sdkFn} but has ${mismatchDetails}. Update parameters to match the SDK function used.`, subnodeName, mismatches[0]?.param));
                    }
                }
            }
        }
    }
}
function buildConditionSummary(displayOptions, parentParams) {
    if (!displayOptions.show)
        return '';
    const parts = [];
    for (const [paramName, expectedValues] of Object.entries(displayOptions.show)) {
        if (!expectedValues)
            continue;
        const actual = parentParams[paramName];
        const expectedStr = expectedValues.map((v) => `'${String(v)}'`).join(' or ');
        parts.push(`${paramName} should be ${expectedStr} (currently '${String(actual)}')`);
    }
    return parts.length > 0 ? `Required: ${parts.join(', ')}.` : '';
}
function validateParentSupportsInputs(json, nodeTypesProvider, warnings) {
    const nodesByName = new Map();
    for (const node of json.nodes) {
        if (node.name) {
            nodesByName.set(node.name, node);
        }
    }
    const connectionsByDest = (0, n8n_workflow_1.mapConnectionsByDestination)(json.connections);
    for (const parentNode of json.nodes) {
        if (!parentNode.name)
            continue;
        const version = typeof parentNode.typeVersion === 'string'
            ? parseFloat(parentNode.typeVersion)
            : (parentNode.typeVersion ?? 1);
        const parentNodeType = nodeTypesProvider.getByNameAndVersion(parentNode.type, version);
        const builderHintInputs = parentNodeType?.description?.builderHint?.inputs;
        if (!builderHintInputs)
            continue;
        const parentContext = {
            parameters: (parentNode.parameters ?? {}),
            nodeVersion: version,
            rootParameters: (parentNode.parameters ?? {}),
        };
        for (const [connectionType, inputConfig] of Object.entries(builderHintInputs)) {
            if (!connectionType.startsWith('ai_'))
                continue;
            if (!inputConfig?.displayOptions)
                continue;
            const parentSupportsInput = (0, display_options_1.matchesDisplayOptions)(parentContext, inputConfig.displayOptions);
            if (parentSupportsInput)
                continue;
            const incomingConnections = connectionsByDest[parentNode.name]?.[connectionType];
            if (!incomingConnections)
                continue;
            for (const connList of incomingConnections) {
                if (!connList)
                    continue;
                for (const conn of connList) {
                    const subnodeName = conn.node;
                    const subnodeField = AI_CONNECTION_TO_SUBNODE_FIELD[connectionType] || connectionType;
                    const conditionDetails = buildConditionSummary(inputConfig.displayOptions, (parentNode.parameters ?? {}));
                    warnings.push(new ValidationWarning('UNSUPPORTED_SUBNODE_INPUT', `'${subnodeName}' is connected to '${parentNode.name}' as ${subnodeField}, but '${parentNode.name}' does not support ${subnodeField} in its current configuration. ${conditionDetails}`, parentNode.name, undefined, undefined, 'major'));
                }
            }
        }
    }
}
function checkNodeInputIndices(json, nodeTypesProvider, warnings) {
    const nodesByName = new Map();
    for (const node of json.nodes) {
        if (node.name) {
            nodesByName.set(node.name, node);
        }
    }
    const warnedInputs = new Set();
    for (const [_sourceName, nodeConnections] of Object.entries(json.connections)) {
        const mainConnections = nodeConnections.main;
        if (!mainConnections || !Array.isArray(mainConnections))
            continue;
        for (const outputs of mainConnections) {
            if (!outputs)
                continue;
            for (const conn of outputs) {
                const targetNodeName = conn.node;
                const targetInputIndex = conn.index;
                const targetNode = nodesByName.get(targetNodeName);
                if (!targetNode)
                    continue;
                const version = typeof targetNode.typeVersion === 'string'
                    ? parseFloat(targetNode.typeVersion)
                    : (targetNode.typeVersion ?? 1);
                const mainInputCount = (0, input_resolver_1.resolveMainInputCount)(nodeTypesProvider, targetNode.type, version);
                if (mainInputCount === undefined)
                    continue;
                if (targetInputIndex >= mainInputCount) {
                    const warnKey = `${targetNodeName}:${targetInputIndex}`;
                    if (!warnedInputs.has(warnKey)) {
                        warnedInputs.add(warnKey);
                        warnings.push(new ValidationWarning('INVALID_INPUT_INDEX', `Connection to '${targetNodeName}' uses input index ${targetInputIndex}, but node only has ${mainInputCount} input(s) (indices 0-${mainInputCount - 1})`, targetNodeName));
                    }
                }
            }
        }
    }
}
