// @n8n/workflow-sdk/dist/validation/resolve-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.matchesDisplayOptions = matchesDisplayOptions;
exports.resolveSchema = resolveSchema;
const zod_1 = require("zod");
const display_options_1 = require("./display-options");
function formatValue(val) {
    if (typeof val === 'boolean')
        return String(val);
    if (typeof val === 'string')
        return `"${val}"`;
    if (typeof val === 'number')
        return String(val);
    if (val !== null && typeof val === 'object' && '_cnd' in val) {
        const cnd = val._cnd;
        const [operator, operand] = Object.entries(cnd)[0];
        switch (operator) {
            case 'exists':
                return 'exists';
            case 'eq':
                return `equals ${formatValue(operand)}`;
            case 'not':
                return `not ${formatValue(operand)}`;
            case 'gte':
                return `>= ${String(operand)}`;
            case 'lte':
                return `<= ${String(operand)}`;
            case 'gt':
                return `> ${String(operand)}`;
            case 'lt':
                return `< ${String(operand)}`;
            case 'includes':
                return `includes "${String(operand)}"`;
            case 'startsWith':
                return `starts with "${String(operand)}"`;
            case 'endsWith':
                return `ends with "${String(operand)}"`;
            case 'regex':
                return `matches /${String(operand)}/`;
            case 'between': {
                const { from, to } = operand;
                return `between ${from} and ${to}`;
            }
            default:
                return JSON.stringify(val);
        }
    }
    return JSON.stringify(val);
}
function formatRegexPath(path) {
    const cleanPath = path.startsWith('/') ? path.slice(1) : path;
    const match = cleanPath.match(/^(.+)\.\(([^)]+)\)$/);
    if (match) {
        const prefix = match[1];
        const alternatives = match[2].split('|');
        const paths = alternatives.map((alt) => `${prefix}.${alt}`);
        if (paths.length === 1)
            return paths[0];
        if (paths.length === 2)
            return `${paths[0]} or ${paths[1]}`;
        return paths.slice(0, -1).join(', ') + ', or ' + paths[paths.length - 1];
    }
    return cleanPath;
}
function isRegexPath(path) {
    return path.includes('|') || path.includes('(');
}
function formatDisplayOptionsRequirements(displayOptions) {
    const requirements = [];
    if (displayOptions.show) {
        for (const [key, values] of Object.entries(displayOptions.show)) {
            if (Array.isArray(values)) {
                const valStr = values.length === 1 ? formatValue(values[0]) : values.map(formatValue).join(' or ');
                if (isRegexPath(key)) {
                    const formattedPath = formatRegexPath(key);
                    requirements.push(`one of ${formattedPath} ${valStr}`);
                }
                else {
                    requirements.push(`${key}=${valStr}`);
                }
            }
        }
    }
    return requirements.join(', ');
}
function matchesDisplayOptions(parameters, displayOptions) {
    const context = { parameters };
    return (0, display_options_1.matchesDisplayOptions)(context, displayOptions);
}
function resolveSchema({ parameters, schema, required, displayOptions, defaults = {}, isToolNode, }) {
    const context = { parameters, defaults, isToolNode };
    const isVisible = (0, display_options_1.matchesDisplayOptions)(context, displayOptions);
    if (isVisible) {
        return required ? schema : schema.optional();
    }
    else {
        const requirements = formatDisplayOptionsRequirements(displayOptions);
        return zod_1.z.any().refine((val) => val === undefined, {
            message: requirements
                ? `This field is only allowed when: ${requirements}`
                : 'This field is not applicable for the current configuration',
        });
    }
}
