// @n8n/workflow-sdk/dist/validation/schema-validator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSchemaBaseDirs = getSchemaBaseDirs;
exports.setSchemaBaseDirs = setSchemaBaseDirs;
exports.loadSchema = loadSchema;
exports.validateNodeConfig = validateNodeConfig;
const path = __importStar(require("path"));
const schemaHelpers = __importStar(require("./schema-helpers"));
let schemaBaseDirs = [];
const schemaCache = new Map();
function getSchemaBaseDirs() {
    return schemaBaseDirs;
}
function setSchemaBaseDirs(dirs) {
    schemaBaseDirs = dirs;
    schemaCache.clear();
}
function nodeTypeToPathComponents(nodeType) {
    const normalized = nodeType.replace(/^@n8n\//, '');
    const [pkg, ...rest] = normalized.split('.');
    const nodeName = rest.join('.');
    return { pkg, nodeName };
}
function versionToString(version) {
    const versionStr = String(version).replace('.', '');
    return `v${versionStr}`;
}
function buildExpectedSchemaName(nodeName, versionStr, isLangchain) {
    const pascalName = nodeName.charAt(0).toUpperCase() + nodeName.slice(1);
    const pascalVersion = versionStr.charAt(0).toUpperCase() + versionStr.slice(1);
    const prefix = isLangchain ? 'Lc' : '';
    return `${prefix}${pascalName}${pascalVersion}ConfigSchema`;
}
function buildExpectedFactoryName(nodeName, versionStr, isLangchain) {
    const pascalName = nodeName.charAt(0).toUpperCase() + nodeName.slice(1);
    const pascalVersion = versionStr.charAt(0).toUpperCase() + versionStr.slice(1);
    const prefix = isLangchain ? 'Lc' : '';
    return `get${prefix}${pascalName}${pascalVersion}ConfigSchema`;
}
function tryLoadSchemaModule(schemaPath) {
    try {
        return require(schemaPath);
    }
    catch {
        return null;
    }
}
function resolveSchemaFromModule(schemaModule, nodeName, versionStr, isLangchain) {
    const expectedSchemaName = buildExpectedSchemaName(nodeName, versionStr, isLangchain);
    const expectedFactoryName = buildExpectedFactoryName(nodeName, versionStr, isLangchain);
    if (typeof schemaModule === 'function') {
        return schemaModule;
    }
    if (typeof schemaModule.default === 'function') {
        return schemaModule.default;
    }
    if (schemaModule[expectedSchemaName]) {
        return schemaModule[expectedSchemaName];
    }
    if (typeof schemaModule[expectedFactoryName] === 'function') {
        return schemaModule[expectedFactoryName];
    }
    const schemaName = Object.keys(schemaModule).find((k) => k.endsWith('ConfigSchema') && !k.includes('Subnode'));
    if (schemaName && schemaModule[schemaName]) {
        return schemaModule[schemaName];
    }
    const factoryName = Object.keys(schemaModule).find((k) => k.startsWith('get') &&
        k.endsWith('ConfigSchema') &&
        !k.includes('Subnode') &&
        typeof schemaModule[k] === 'function');
    if (factoryName && schemaModule[factoryName]) {
        return schemaModule[factoryName];
    }
    return null;
}
function tryLoadSchemaForNodeType(nodeType, version) {
    const { pkg, nodeName } = nodeTypeToPathComponents(nodeType);
    const versionStr = versionToString(version);
    const isLangchain = pkg === 'n8n-nodes-langchain';
    for (const baseDir of schemaBaseDirs) {
        const flatSchemaPath = path.join(baseDir, 'nodes', pkg, nodeName, `${versionStr}.schema`);
        const splitSchemaPath = path.join(baseDir, 'nodes', pkg, nodeName, versionStr, 'index.schema');
        const schemaModule = tryLoadSchemaModule(flatSchemaPath) ?? tryLoadSchemaModule(splitSchemaPath);
        if (!schemaModule) {
            continue;
        }
        const result = resolveSchemaFromModule(schemaModule, nodeName, versionStr, isLangchain);
        if (result) {
            return result;
        }
    }
    return null;
}
function loadSchema(nodeType, version) {
    const cacheKey = `${nodeType}@${version}`;
    if (schemaCache.has(cacheKey)) {
        return schemaCache.get(cacheKey) ?? null;
    }
    let schema = tryLoadSchemaForNodeType(nodeType, version);
    if (!schema && nodeType.endsWith('Tool')) {
        const baseNodeType = nodeType.slice(0, -4);
        schema = tryLoadSchemaForNodeType(baseNodeType, version);
    }
    schemaCache.set(cacheKey, schema);
    return schema;
}
function scoreUnionVariant(issues) {
    const discriminatorFields = ['mode', 'resource', 'operation'];
    let discriminatorMismatches = 0;
    for (const iss of issues) {
        if (iss.code === 'invalid_literal') {
            const lastPart = iss.path[iss.path.length - 1];
            if (typeof lastPart === 'string' && discriminatorFields.includes(lastPart)) {
                discriminatorMismatches++;
            }
        }
        else if (iss.code === 'invalid_union' &&
            'unionErrors' in iss &&
            Array.isArray(iss.unionErrors)) {
            const nestedScores = iss.unionErrors.map((ue) => scoreUnionVariant(ue.issues));
            discriminatorMismatches += Math.min(...nestedScores);
        }
    }
    return discriminatorMismatches;
}
function findBestMatchingVariant(unionErrors) {
    if (unionErrors.length === 0)
        return null;
    let bestVariant = unionErrors[0];
    let bestScore = scoreUnionVariant(bestVariant.issues);
    for (let i = 1; i < unionErrors.length; i++) {
        const score = scoreUnionVariant(unionErrors[i].issues);
        if (score < bestScore) {
            bestScore = score;
            bestVariant = unionErrors[i];
        }
    }
    return bestVariant;
}
function collectIssuesFromBestPath(unionErrors) {
    const bestVariant = findBestMatchingVariant(unionErrors);
    if (!bestVariant)
        return [];
    const result = [];
    for (const iss of bestVariant.issues) {
        if (iss.code === 'invalid_union' && 'unionErrors' in iss && Array.isArray(iss.unionErrors)) {
            result.push(...collectIssuesFromBestPath(iss.unionErrors));
        }
        else {
            result.push(iss);
        }
    }
    return result;
}
function collectAllDiscriminatorValues(unionErrors, discriminatorPath) {
    const values = [];
    for (const unionError of unionErrors) {
        for (const iss of unionError.issues) {
            if (iss.code === 'invalid_literal' && iss.path.join('.') === discriminatorPath) {
                values.push(iss.expected);
            }
            else if (iss.code === 'invalid_union' &&
                'unionErrors' in iss &&
                Array.isArray(iss.unionErrors)) {
                values.push(...collectAllDiscriminatorValues(iss.unionErrors, discriminatorPath));
            }
        }
    }
    return [...new Set(values)];
}
function extractUnionErrorSummary(unionErrors) {
    const bestPathIssues = collectIssuesFromBestPath(unionErrors);
    if (bestPathIssues.length === 0) {
        return null;
    }
    const issuesByPath = new Map();
    for (const iss of bestPathIssues) {
        const path = iss.path.join('.');
        if (!issuesByPath.has(path)) {
            issuesByPath.set(path, []);
        }
        issuesByPath.get(path).push(iss);
    }
    const discriminatorFields = ['mode', 'resource', 'operation'];
    for (const field of discriminatorFields) {
        for (const [path, issues] of issuesByPath) {
            if (path.endsWith(field)) {
                const literalIssues = issues.filter((i) => i.code === 'invalid_literal');
                if (literalIssues.length > 0) {
                    const receivedValue = literalIssues[0].received;
                    const expectedValues = receivedValue === undefined
                        ? collectAllDiscriminatorValues(unionErrors, path)
                        : [...new Set(literalIssues.map((i) => i.expected))];
                    const expectedStr = expectedValues.map((v) => `"${String(v)}"`).join(', ');
                    if (receivedValue === undefined) {
                        return `Missing discriminator "${path}". Expected one of: ${expectedStr}. Make sure "${field}" is inside "parameters".`;
                    }
                    let receivedStr;
                    if (typeof receivedValue === 'object') {
                        receivedStr = JSON.stringify(receivedValue);
                    }
                    else {
                        receivedStr = String(receivedValue);
                    }
                    return `Invalid value for "${path}": got "${receivedStr}", expected one of: ${expectedStr}.`;
                }
            }
        }
    }
    const typeMismatches = [];
    for (const [path, issues] of issuesByPath) {
        for (const iss of issues) {
            if (iss.code === 'invalid_type') {
                typeMismatches.push({
                    path,
                    expected: iss.expected ?? 'unknown',
                    received: iss.received ?? 'unknown',
                });
            }
        }
    }
    if (typeMismatches.length > 0) {
        const uniqueMismatches = new Map();
        for (const m of typeMismatches) {
            if (!uniqueMismatches.has(m.path)) {
                uniqueMismatches.set(m.path, { expected: m.expected, received: m.received });
            }
        }
        if (uniqueMismatches.size === 1) {
            const [path, { expected, received }] = [...uniqueMismatches.entries()][0];
            return `Field "${path}" has wrong type: expected ${expected}, got ${received}.`;
        }
        const paths = [...uniqueMismatches.keys()].slice(0, 3);
        const summaries = paths.map((p) => {
            const { expected, received } = uniqueMismatches.get(p);
            return `"${p}" (expected ${expected}, got ${received})`;
        });
        const more = uniqueMismatches.size > 3 ? ` and ${uniqueMismatches.size - 3} more` : '';
        return `Type mismatches: ${summaries.join(', ')}${more}.`;
    }
    const specificIssues = bestPathIssues
        .filter((i) => i.code !== 'invalid_union')
        .slice(0, 3);
    if (specificIssues.length > 0) {
        const summaries = specificIssues.map((iss) => {
            const path = iss.path.join('.') || 'config';
            if (iss.code === 'invalid_literal') {
                return `"${path}" must be "${String(iss.expected)}"`;
            }
            return `"${path}": ${iss.message}`;
        });
        return `Validation failed: ${summaries.join('; ')}.`;
    }
    return null;
}
function formatInvalidType(issue, path) {
    const received = issue.received;
    const expected = issue.expected;
    if (received === 'undefined') {
        return `Required field "${path}" is missing. Expected ${expected}.`;
    }
    return `Field "${path}" has wrong type. Expected ${expected}, but got ${received}.`;
}
function formatInvalidUnion(issue, path) {
    if ('unionErrors' in issue && Array.isArray(issue.unionErrors)) {
        const allMissing = issue.unionErrors.every((ue) => ue.issues.some((i) => i.code === 'invalid_type' && i.received === 'undefined'));
        if (allMissing) {
            return `Required field "${path}" is missing.`;
        }
        const errorSummary = extractUnionErrorSummary(issue.unionErrors);
        if (errorSummary) {
            return errorSummary;
        }
    }
    return `Field "${path}" has invalid value. None of the expected types matched.`;
}
function formatInvalidLiteral(issue, path) {
    const expected = issue.expected;
    const received = issue.received;
    return `Field "${path}" must be exactly "${String(expected)}", but got "${String(received)}".`;
}
function formatInvalidEnum(issue, path) {
    if ('options' in issue && Array.isArray(issue.options)) {
        return `Field "${path}" must be one of: ${issue.options.map((o) => `"${String(o)}"`).join(', ')}.`;
    }
    return `Field "${path}" has invalid enum value.`;
}
function formatTooSmall(issue, path) {
    const issueType = issue.type;
    const minimum = issue.minimum;
    if (issueType === 'string') {
        return `Field "${path}" is too short. Minimum length is ${minimum}.`;
    }
    if (issueType === 'array') {
        return `Field "${path}" must have at least ${minimum} item(s).`;
    }
    return `Field "${path}" is too small.`;
}
function formatTooBig(issue, path) {
    const issueType = issue.type;
    const maximum = issue.maximum;
    if (issueType === 'string') {
        return `Field "${path}" is too long. Maximum length is ${maximum}.`;
    }
    if (issueType === 'array') {
        return `Field "${path}" must have at most ${maximum} item(s).`;
    }
    return `Field "${path}" is too large.`;
}
function formatUnrecognizedKeys(issue, path) {
    if ('keys' in issue && Array.isArray(issue.keys)) {
        return `Unknown field(s) at "${path}": ${issue.keys.map((k) => `"${String(k)}"`).join(', ')}.`;
    }
    return `Unknown fields at "${path}".`;
}
const ISSUE_FORMATTERS = {
    invalid_type: formatInvalidType,
    invalid_union: formatInvalidUnion,
    invalid_literal: formatInvalidLiteral,
    invalid_enum_value: formatInvalidEnum,
    too_small: formatTooSmall,
    too_big: formatTooBig,
    unrecognized_keys: formatUnrecognizedKeys,
};
function formatZodIssue(issue) {
    const path = issue.path.length > 0 ? issue.path.join('.') : 'config';
    const formatter = ISSUE_FORMATTERS[issue.code];
    if (formatter) {
        return formatter(issue, path);
    }
    return `Field "${path}": ${issue.message}`;
}
function formatZodErrors(issues) {
    return issues.map((issue) => ({
        path: issue.path.join('.'),
        message: formatZodIssue(issue),
    }));
}
function isZodSchema(value) {
    return typeof value === 'object' && value !== null && 'safeParse' in value;
}
function validateNodeConfig(nodeType, version, config, options) {
    const schemaOrFactory = loadSchema(nodeType, version);
    if (!schemaOrFactory) {
        return { valid: true, errors: [] };
    }
    let schema;
    if (isZodSchema(schemaOrFactory)) {
        schema = schemaOrFactory;
    }
    else {
        const parameters = (config.parameters ?? {});
        const isToolNode = options?.isToolNode ?? false;
        schema = schemaOrFactory({
            ...schemaHelpers,
            parameters,
            resolveSchema: (cfg) => schemaHelpers.resolveSchema({ ...cfg, isToolNode }),
        });
    }
    const result = schema.safeParse(config);
    if (result.success) {
        return { valid: true, errors: [] };
    }
    return {
        valid: false,
        errors: formatZodErrors(result.error.issues),
    };
}
