// @n8n/workflow-sdk/dist/validation/schema-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rerankerInstanceSchema = exports.textSplitterInstanceSchema = exports.documentLoaderInstanceSchema = exports.retrieverInstanceSchema = exports.vectorStoreInstanceSchema = exports.embeddingInstanceSchema = exports.outputParserInstanceSchema = exports.toolInstanceSchema = exports.memoryInstanceSchema = exports.languageModelInstanceSchema = exports.subnodeInstanceBaseSchema = exports.resourceMapperValueSchema = exports.resolveSchema = exports.multiOptionsSchema = exports.optionsWithExpression = exports.literalUnion = exports.iDataObjectSchema = exports.assignmentCollectionValueSchema = exports.assignmentSchema = exports.filterValueSchema = exports.filterConditionSchema = exports.filterOperatorSchema = exports.resourceLocatorValueSchema = exports.booleanOrExpression = exports.numberOrExpression = exports.stringOrExpression = exports.expressionSchema = exports.expressionPattern = exports.z = void 0;
const zod_1 = require("zod");
Object.defineProperty(exports, "z", { enumerable: true, get: function () { return zod_1.z; } });
var zod_helpers_1 = require("../generate-types/zod-helpers");
Object.defineProperty(exports, "expressionPattern", { enumerable: true, get: function () { return zod_helpers_1.expressionPattern; } });
Object.defineProperty(exports, "expressionSchema", { enumerable: true, get: function () { return zod_helpers_1.expressionSchema; } });
Object.defineProperty(exports, "stringOrExpression", { enumerable: true, get: function () { return zod_helpers_1.stringOrExpression; } });
Object.defineProperty(exports, "numberOrExpression", { enumerable: true, get: function () { return zod_helpers_1.numberOrExpression; } });
Object.defineProperty(exports, "booleanOrExpression", { enumerable: true, get: function () { return zod_helpers_1.booleanOrExpression; } });
Object.defineProperty(exports, "resourceLocatorValueSchema", { enumerable: true, get: function () { return zod_helpers_1.resourceLocatorValueSchema; } });
Object.defineProperty(exports, "filterOperatorSchema", { enumerable: true, get: function () { return zod_helpers_1.filterOperatorSchema; } });
Object.defineProperty(exports, "filterConditionSchema", { enumerable: true, get: function () { return zod_helpers_1.filterConditionSchema; } });
Object.defineProperty(exports, "filterValueSchema", { enumerable: true, get: function () { return zod_helpers_1.filterValueSchema; } });
Object.defineProperty(exports, "assignmentSchema", { enumerable: true, get: function () { return zod_helpers_1.assignmentSchema; } });
Object.defineProperty(exports, "assignmentCollectionValueSchema", { enumerable: true, get: function () { return zod_helpers_1.assignmentCollectionValueSchema; } });
Object.defineProperty(exports, "iDataObjectSchema", { enumerable: true, get: function () { return zod_helpers_1.iDataObjectSchema; } });
Object.defineProperty(exports, "literalUnion", { enumerable: true, get: function () { return zod_helpers_1.literalUnion; } });
Object.defineProperty(exports, "optionsWithExpression", { enumerable: true, get: function () { return zod_helpers_1.optionsWithExpression; } });
Object.defineProperty(exports, "multiOptionsSchema", { enumerable: true, get: function () { return zod_helpers_1.multiOptionsSchema; } });
var resolve_schema_1 = require("./resolve-schema");
Object.defineProperty(exports, "resolveSchema", { enumerable: true, get: function () { return resolve_schema_1.resolveSchema; } });
const resourceMapperObjectSchema = zod_1.z
    .object({
    mappingMode: zod_1.z.string(),
    value: zod_1.z.unknown().optional(),
    schema: zod_1.z.array(zod_1.z.unknown()).optional(),
    cachedResultName: zod_1.z.string().optional(),
})
    .passthrough();
exports.resourceMapperValueSchema = zod_1.z.union([
    resourceMapperObjectSchema,
    zod_1.z.string().regex(/^={{.*}}$/s, 'Must be an n8n expression (={{...}})'),
]);
exports.subnodeInstanceBaseSchema = zod_1.z.object({
    type: zod_1.z.string(),
    version: zod_1.z.number(),
    config: zod_1.z.record(zod_1.z.string(), zod_1.z.unknown()).optional(),
});
exports.languageModelInstanceSchema = exports.subnodeInstanceBaseSchema;
exports.memoryInstanceSchema = exports.subnodeInstanceBaseSchema;
exports.toolInstanceSchema = exports.subnodeInstanceBaseSchema;
exports.outputParserInstanceSchema = exports.subnodeInstanceBaseSchema;
exports.embeddingInstanceSchema = exports.subnodeInstanceBaseSchema;
exports.vectorStoreInstanceSchema = exports.subnodeInstanceBaseSchema;
exports.retrieverInstanceSchema = exports.subnodeInstanceBaseSchema;
exports.documentLoaderInstanceSchema = exports.subnodeInstanceBaseSchema;
exports.textSplitterInstanceSchema = exports.subnodeInstanceBaseSchema;
exports.rerankerInstanceSchema = exports.subnodeInstanceBaseSchema;
