// @n8n/workflow-sdk/dist/validation/test-schema-setup.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupTestSchemas = setupTestSchemas;
exports.teardownTestSchemas = teardownTestSchemas;
const crypto = __importStar(require("crypto"));
const fs = __importStar(require("fs"));
const os = __importStar(require("os"));
const path = __importStar(require("path"));
const schema_validator_1 = require("./schema-validator");
const generate_node_defs_cli_1 = require("../generate-types/generate-node-defs-cli");
const WORKER_ID = process.env.JEST_WORKER_ID ?? '0';
const SCHEMA_TEST_DIR = path.join(os.tmpdir(), `n8n-schema-tests-${WORKER_ID}`);
const STAMP_FILE = path.join(SCHEMA_TEST_DIR, '.generator-hash');
let originalBaseDirs;
function computeGeneratorHash() {
    const generatorPath = path.resolve(__dirname, '../generate-types/generate-zod-schemas.ts');
    try {
        const content = fs.readFileSync(generatorPath, 'utf-8');
        return crypto.createHash('md5').update(content).digest('hex');
    }
    catch {
        return '';
    }
}
function isCacheValid() {
    try {
        if (!fs.existsSync(STAMP_FILE))
            return false;
        const storedHash = fs.readFileSync(STAMP_FILE, 'utf-8').trim();
        const currentHash = computeGeneratorHash();
        return storedHash === currentHash && currentHash !== '';
    }
    catch {
        return false;
    }
}
async function setupTestSchemas() {
    originalBaseDirs = (0, schema_validator_1.getSchemaBaseDirs)();
    const nodesDir = path.join(SCHEMA_TEST_DIR, 'nodes');
    const nodesDirExists = fs.existsSync(nodesDir);
    const cacheValid = isCacheValid();
    if (!nodesDirExists || !cacheValid) {
        if (fs.existsSync(SCHEMA_TEST_DIR)) {
            fs.rmSync(SCHEMA_TEST_DIR, { recursive: true, force: true });
        }
        const repoRoot = path.resolve(__dirname, '../../../../..');
        const nodesBaseJson = path.join(repoRoot, 'packages/nodes-base/dist/types/nodes.json');
        if (fs.existsSync(nodesBaseJson)) {
            await (0, generate_node_defs_cli_1.generateNodeDefinitions)({
                nodesJsonPath: nodesBaseJson,
                outputDir: SCHEMA_TEST_DIR,
                packageName: 'n8n-nodes-base',
            });
        }
        const langchainJson = path.join(repoRoot, 'packages/@n8n/nodes-langchain/dist/types/nodes.json');
        if (fs.existsSync(langchainJson)) {
            await (0, generate_node_defs_cli_1.generateNodeDefinitions)({
                nodesJsonPath: langchainJson,
                outputDir: SCHEMA_TEST_DIR,
                packageName: '@n8n/n8n-nodes-langchain',
            });
        }
        const hash = computeGeneratorHash();
        if (hash) {
            fs.mkdirSync(SCHEMA_TEST_DIR, { recursive: true });
            fs.writeFileSync(STAMP_FILE, hash);
        }
    }
    (0, schema_validator_1.setSchemaBaseDirs)([SCHEMA_TEST_DIR]);
}
function teardownTestSchemas() {
    if (originalBaseDirs !== undefined) {
        (0, schema_validator_1.setSchemaBaseDirs)(originalBaseDirs);
    }
}
