// n8n-core/dist/execution-engine/node-execution-context/base-execute-context.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseExecuteContext = void 0;
const get_1 = __importDefault(require("lodash/get"));
const n8n_workflow_1 = require("n8n-workflow");
const node_execution_context_1 = require("./node-execution-context");
class BaseExecuteContext extends node_execution_context_1.NodeExecutionContext {
    constructor(workflow, node, additionalData, mode, runExecutionData, runIndex, connectionInputData, inputData, executeData, abortSignal) {
        super(workflow, node, additionalData, mode, runExecutionData, runIndex);
        this.runExecutionData = runExecutionData;
        this.connectionInputData = connectionInputData;
        this.inputData = inputData;
        this.executeData = executeData;
        this.abortSignal = abortSignal;
    }
    getExecutionContext() {
        return this.runExecutionData.executionData?.runtimeData;
    }
    getExecutionCancelSignal() {
        return this.abortSignal;
    }
    onExecutionCancellation(handler) {
        const fn = () => {
            this.abortSignal?.removeEventListener('abort', fn);
            handler();
        };
        this.abortSignal?.addEventListener('abort', fn);
    }
    getExecuteData() {
        return this.executeData;
    }
    setMetadata(metadata) {
        this.executeData.metadata = {
            ...(this.executeData.metadata ?? {}),
            ...metadata,
        };
    }
    getContext(type) {
        return n8n_workflow_1.NodeHelpers.getContext(this.runExecutionData, type, this.node);
    }
    continueOnFail() {
        const onError = (0, get_1.default)(this.node, 'onError', undefined);
        if (onError === undefined) {
            return (0, get_1.default)(this.node, 'continueOnFail', false);
        }
        return ['continueRegularOutput', 'continueErrorOutput'].includes(onError);
    }
    async getCredentials(type, itemIndex) {
        return await this._getCredentials(type, this.executeData, this.connectionInputData, itemIndex);
    }
    async putExecutionToWait(waitTill) {
        this.runExecutionData.waitTill = waitTill;
        if (this.additionalData.setExecutionStatus) {
            this.additionalData.setExecutionStatus('waiting');
        }
    }
    async executeWorkflow(workflowInfo, inputData, parentCallbackManager, options) {
        if (options?.parentExecution) {
            if (!options.parentExecution.executionContext &&
                options.parentExecution.executionId === this.getExecutionId()) {
                options.parentExecution.executionContext = this.getExecutionContext();
            }
        }
        const result = await this.additionalData.executeWorkflow(workflowInfo, this.additionalData, {
            ...options,
            parentWorkflowId: this.workflow.id,
            inputData,
            parentWorkflowSettings: this.workflow.settings,
            node: this.node,
            parentCallbackManager,
        });
        if (result.waitTill) {
            await this.putExecutionToWait(n8n_workflow_1.WAIT_INDEFINITELY);
        }
        return result;
    }
    async getExecutionDataById(executionId) {
        return await this.additionalData.getRunExecutionData(executionId);
    }
    getInputItems(inputIndex, connectionType) {
        const inputData = this.inputData[connectionType];
        if (inputData.length < inputIndex) {
            throw new n8n_workflow_1.ApplicationError('Could not get input with given index', {
                extra: { inputIndex, connectionType },
            });
        }
        const allItems = inputData[inputIndex];
        if (allItems === null) {
            throw new n8n_workflow_1.ApplicationError('Input index was not set', {
                extra: { inputIndex, connectionType },
            });
        }
        return allItems;
    }
    getInputSourceData(inputIndex = 0, connectionType = n8n_workflow_1.NodeConnectionTypes.Main) {
        if (this.executeData?.source === null) {
            throw new n8n_workflow_1.ApplicationError('Source data is missing');
        }
        return this.executeData.source[connectionType][inputIndex];
    }
    getWorkflowDataProxy(itemIndex) {
        return new n8n_workflow_1.WorkflowDataProxy(this.workflow, this.runExecutionData, this.runIndex, itemIndex, this.node.name, this.connectionInputData, {}, this.mode, this.additionalKeys, this.executeData).getDataProxy();
    }
    sendMessageToUI(...args) {
        if (this.mode !== 'manual') {
            return;
        }
        try {
            if (this.additionalData.sendDataToUI) {
                args = args.map((arg) => {
                    if (arg.isLuxonDateTime && arg.invalidReason)
                        return { ...arg };
                    if (arg.isLuxonDateTime)
                        return new Date(arg.ts).toString();
                    if (arg instanceof Date)
                        return arg.toString();
                    return arg;
                });
                this.additionalData.sendDataToUI('sendConsoleMessage', {
                    source: `[Node: "${this.node.name}"]`,
                    messages: args,
                });
            }
        }
        catch (error) {
            this.logger.warn(`There was a problem sending message to UI: ${error.message}`);
        }
    }
    logAiEvent(eventName, msg) {
        return this.additionalData.logAiEvent(eventName, {
            executionId: this.additionalData.executionId ?? 'unsaved-execution',
            nodeName: this.node.name,
            workflowName: this.workflow.name ?? 'Unnamed workflow',
            nodeType: this.node.type,
            workflowId: this.workflow.id ?? 'unsaved-workflow',
            msg,
        });
    }
    async startJob(jobType, settings, itemIndex) {
        return await this.additionalData.startRunnerTask(this.additionalData, jobType, settings, this, this.inputData, this.node, this.workflow, this.runExecutionData, this.runIndex, itemIndex, this.node.name, this.connectionInputData, {}, this.mode, (0, n8n_workflow_1.createEnvProviderState)(), this.executeData);
    }
    getRunnerStatus(taskType) {
        return this.additionalData.getRunnerStatus?.(taskType) ?? { available: true };
    }
}
exports.BaseExecuteContext = BaseExecuteContext;
