// n8n-core/dist/execution-engine/node-execution-context/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveSourceOverwrite = exports.returnJsonArray = exports.parseRequestObject = exports.parseIncomingMessage = exports.normalizeItems = exports.getNonWorkflowAdditionalKeys = exports.getAdditionalKeys = exports.constructExecutionMetaData = exports.StructuredToolkit = exports.WebhookContext = exports.TriggerContext = exports.SupplyDataContext = exports.PollContext = exports.LocalLoadOptionsContext = exports.LoadOptionsContext = exports.HookContext = exports.ExecuteSingleContext = exports.ExecuteContext = exports.CredentialTestContext = void 0;
var credentials_test_context_1 = require("./credentials-test-context");
Object.defineProperty(exports, "CredentialTestContext", { enumerable: true, get: function () { return credentials_test_context_1.CredentialTestContext; } });
var execute_context_1 = require("./execute-context");
Object.defineProperty(exports, "ExecuteContext", { enumerable: true, get: function () { return execute_context_1.ExecuteContext; } });
var execute_single_context_1 = require("./execute-single-context");
Object.defineProperty(exports, "ExecuteSingleContext", { enumerable: true, get: function () { return execute_single_context_1.ExecuteSingleContext; } });
var hook_context_1 = require("./hook-context");
Object.defineProperty(exports, "HookContext", { enumerable: true, get: function () { return hook_context_1.HookContext; } });
var load_options_context_1 = require("./load-options-context");
Object.defineProperty(exports, "LoadOptionsContext", { enumerable: true, get: function () { return load_options_context_1.LoadOptionsContext; } });
var local_load_options_context_1 = require("./local-load-options-context");
Object.defineProperty(exports, "LocalLoadOptionsContext", { enumerable: true, get: function () { return local_load_options_context_1.LocalLoadOptionsContext; } });
var poll_context_1 = require("./poll-context");
Object.defineProperty(exports, "PollContext", { enumerable: true, get: function () { return poll_context_1.PollContext; } });
var supply_data_context_1 = require("./supply-data-context");
Object.defineProperty(exports, "SupplyDataContext", { enumerable: true, get: function () { return supply_data_context_1.SupplyDataContext; } });
var trigger_context_1 = require("./trigger-context");
Object.defineProperty(exports, "TriggerContext", { enumerable: true, get: function () { return trigger_context_1.TriggerContext; } });
var webhook_context_1 = require("./webhook-context");
Object.defineProperty(exports, "WebhookContext", { enumerable: true, get: function () { return webhook_context_1.WebhookContext; } });
var ai_tool_types_1 = require("./utils/ai-tool-types");
Object.defineProperty(exports, "StructuredToolkit", { enumerable: true, get: function () { return ai_tool_types_1.StructuredToolkit; } });
var construct_execution_metadata_1 = require("./utils/construct-execution-metadata");
Object.defineProperty(exports, "constructExecutionMetaData", { enumerable: true, get: function () { return construct_execution_metadata_1.constructExecutionMetaData; } });
var get_additional_keys_1 = require("./utils/get-additional-keys");
Object.defineProperty(exports, "getAdditionalKeys", { enumerable: true, get: function () { return get_additional_keys_1.getAdditionalKeys; } });
Object.defineProperty(exports, "getNonWorkflowAdditionalKeys", { enumerable: true, get: function () { return get_additional_keys_1.getNonWorkflowAdditionalKeys; } });
var normalize_items_1 = require("./utils/normalize-items");
Object.defineProperty(exports, "normalizeItems", { enumerable: true, get: function () { return normalize_items_1.normalizeItems; } });
var parse_incoming_message_1 = require("./utils/parse-incoming-message");
Object.defineProperty(exports, "parseIncomingMessage", { enumerable: true, get: function () { return parse_incoming_message_1.parseIncomingMessage; } });
var request_helper_functions_1 = require("./utils/request-helper-functions");
Object.defineProperty(exports, "parseRequestObject", { enumerable: true, get: function () { return request_helper_functions_1.parseRequestObject; } });
var return_json_array_1 = require("./utils/return-json-array");
Object.defineProperty(exports, "returnJsonArray", { enumerable: true, get: function () { return return_json_array_1.returnJsonArray; } });
var resolve_source_overwrite_1 = require("./utils/resolve-source-overwrite");
Object.defineProperty(exports, "resolveSourceOverwrite", { enumerable: true, get: function () { return resolve_source_overwrite_1.resolveSourceOverwrite; } });
__exportStar(require("./utils/binary-helper-functions"), exports);
