// n8n-core/dist/execution-engine/node-execution-context/node-execution-context.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeExecutionContext = void 0;
const backend_common_1 = require("@n8n/backend-common");
const decorators_1 = require("@n8n/decorators");
const di_1 = require("@n8n/di");
const get_1 = __importDefault(require("lodash/get"));
const n8n_workflow_1 = require("n8n-workflow");
const constants_1 = require("../../constants");
const instance_settings_1 = require("../../instance-settings");
const signature_helpers_1 = require("../../utils/signature-helpers");
const cleanup_parameter_data_1 = require("./utils/cleanup-parameter-data");
const custom_data_1 = require("./utils/custom-data");
const ensure_type_1 = require("./utils/ensure-type");
const extract_value_1 = require("./utils/extract-value");
const get_additional_keys_1 = require("./utils/get-additional-keys");
const validate_value_against_schema_1 = require("./utils/validate-value-against-schema");
class NodeExecutionContext {
    constructor(workflow, node, additionalData, mode, runExecutionData = null, runIndex = 0, connectionInputData = [], executeData) {
        this.workflow = workflow;
        this.node = node;
        this.additionalData = additionalData;
        this.mode = mode;
        this.runExecutionData = runExecutionData;
        this.runIndex = runIndex;
        this.connectionInputData = connectionInputData;
        this.executeData = executeData;
        this.instanceSettings = di_1.Container.get(instance_settings_1.InstanceSettings);
    }
    get logger() {
        return di_1.Container.get(backend_common_1.Logger);
    }
    get customData() {
        if (!this.runExecutionData) {
            throw new n8n_workflow_1.ApplicationError('Cannot access customData: runExecutionData is not available in this context');
        }
        return (0, custom_data_1.createExecutionCustomData)({
            runExecutionData: this.runExecutionData,
            mode: this.mode,
        });
    }
    getExecutionContext() {
        return this.runExecutionData?.executionData?.runtimeData;
    }
    getExecutionId() {
        return this.additionalData.executionId;
    }
    getNode() {
        return (0, n8n_workflow_1.deepCopy)(this.node);
    }
    getWorkflow() {
        const { id, name, active } = this.workflow;
        return { id, name, active };
    }
    getMode() {
        return this.mode;
    }
    getWorkflowStaticData(type) {
        return this.workflow.getStaticData(type, this.node);
    }
    getChildNodes(nodeName, options) {
        const output = [];
        const nodeNames = this.workflow.getChildNodes(nodeName);
        for (const n of nodeNames) {
            const node = this.workflow.nodes[n];
            const entry = {
                name: node.name,
                type: node.type,
                typeVersion: node.typeVersion,
                disabled: node.disabled ?? false,
            };
            if (options?.includeNodeParameters) {
                entry.parameters = node.parameters;
            }
            output.push(entry);
        }
        return output;
    }
    getParentNodes(nodeName, options) {
        const output = [];
        const nodeNames = this.workflow.getParentNodes(nodeName, options?.connectionType, options?.depth);
        for (const n of nodeNames) {
            const node = this.workflow.nodes[n];
            const entry = {
                name: node.name,
                type: node.type,
                typeVersion: node.typeVersion,
                disabled: node.disabled ?? false,
            };
            if (options?.includeNodeParameters) {
                entry.parameters = node.parameters;
            }
            output.push(entry);
        }
        return output;
    }
    getChatTrigger() {
        for (const node of Object.values(this.workflow.nodes)) {
            if (this.workflow.nodes[node.name].type === n8n_workflow_1.CHAT_TRIGGER_NODE_TYPE) {
                return this.workflow.nodes[node.name];
            }
        }
        return null;
    }
    get workflowSettings() {
        return Object.freeze(structuredClone(this.workflow.settings));
    }
    getWorkflowSettings() {
        return this.workflowSettings;
    }
    get nodeType() {
        const { type, typeVersion } = this.node;
        return this.workflow.nodeTypes.getByNameAndVersion(type, typeVersion);
    }
    get nodeFeatures() {
        const description = this.nodeType.description;
        return n8n_workflow_1.NodeHelpers.getNodeFeatures(description.features, this.node.typeVersion);
    }
    isNodeFeatureEnabled(featureName) {
        return this.nodeFeatures[featureName] ?? false;
    }
    get nodeInputs() {
        return n8n_workflow_1.NodeHelpers.getNodeInputs(this.workflow, this.node, this.nodeType.description).map((input) => (typeof input === 'string' ? { type: input } : input));
    }
    getNodeInputs() {
        return this.nodeInputs;
    }
    get nodeOutputs() {
        return n8n_workflow_1.NodeHelpers.getNodeOutputs(this.workflow, this.node, this.nodeType.description).map((output) => (typeof output === 'string' ? { type: output } : output));
    }
    getConnectedNodes(connectionType) {
        return this.workflow
            .getParentNodes(this.node.name, connectionType, 1)
            .map((nodeName) => this.workflow.getNode(nodeName))
            .filter((node) => !!node)
            .filter((node) => node.disabled !== true);
    }
    getConnections(destination, connectionType) {
        return this.workflow.connectionsByDestinationNode[destination.name]?.[connectionType] ?? [];
    }
    getNodeOutputs() {
        return this.nodeOutputs;
    }
    getKnownNodeTypes() {
        return this.workflow.nodeTypes.getKnownTypes();
    }
    getRestApiUrl() {
        return this.additionalData.restApiUrl;
    }
    getInstanceBaseUrl() {
        return this.additionalData.instanceBaseUrl;
    }
    getInstanceId() {
        return this.instanceSettings.instanceId;
    }
    getSignedResumeUrl(parameters = {}) {
        const { webhookWaitingBaseUrl, executionId } = this.additionalData;
        if (typeof executionId !== 'string') {
            throw new n8n_workflow_1.UnexpectedError('Execution id is missing');
        }
        const baseURL = new URL(`${webhookWaitingBaseUrl}/${executionId}/${this.node.id}`);
        for (const [key, value] of Object.entries(parameters)) {
            baseURL.searchParams.set(key, value);
        }
        const urlForSigning = (0, signature_helpers_1.prepareUrlForSigning)(baseURL);
        const signature = (0, signature_helpers_1.generateUrlSignature)(urlForSigning, this.instanceSettings.hmacSignatureSecret);
        baseURL.searchParams.set(constants_1.WAITING_TOKEN_QUERY_PARAM, signature);
        return baseURL.toString();
    }
    getTimezone() {
        return this.workflow.timezone;
    }
    getCredentialsProperties(type) {
        return this.additionalData.credentialsHelper.getCredentialsProperties(type);
    }
    async _getCredentials(type, executeData, connectionInputData, itemIndex) {
        const { workflow, node, additionalData, mode, runExecutionData, runIndex } = this;
        const nodeType = workflow.nodeTypes.getByNameAndVersion(node.type, node.typeVersion);
        const fullAccess = [
            constants_1.HTTP_REQUEST_NODE_TYPE,
            constants_1.HTTP_REQUEST_TOOL_NODE_TYPE,
            constants_1.HTTP_REQUEST_AS_TOOL_NODE_TYPE,
        ].includes(node.type);
        let nodeCredentialDescription;
        if (!fullAccess) {
            if (nodeType.description.credentials === undefined) {
                throw new n8n_workflow_1.NodeOperationError(node, `Node type "${node.type}" does not have any credentials defined`, { level: 'warning' });
            }
            nodeCredentialDescription = nodeType.description.credentials.find((credentialTypeDescription) => credentialTypeDescription.name === type);
            if (nodeCredentialDescription === undefined) {
                throw new n8n_workflow_1.NodeOperationError(node, `Node type "${node.type}" does not have any credentials of type "${type}" defined`, { level: 'warning' });
            }
            if (!n8n_workflow_1.NodeHelpers.displayParameter(additionalData.currentNodeParameters || node.parameters, nodeCredentialDescription, node, nodeType.description, node.parameters)) {
                throw new n8n_workflow_1.NodeOperationError(node, 'Credentials not found');
            }
        }
        if (!fullAccess && !node.credentials?.[type]) {
            if (nodeCredentialDescription?.required === true) {
                if (!node.credentials) {
                    throw new n8n_workflow_1.NodeOperationError(node, 'Node does not have any credentials set', {
                        level: 'warning',
                    });
                }
                if (!node.credentials[type]) {
                    throw new n8n_workflow_1.NodeOperationError(node, `Node does not have any credentials set for "${type}"`, {
                        level: 'warning',
                    });
                }
            }
            else {
                throw new n8n_workflow_1.NodeOperationError(node, 'Node does not require credentials');
            }
        }
        if (fullAccess && !node.credentials?.[type]) {
            throw new n8n_workflow_1.NodeOperationError(node, 'Credentials not found');
        }
        let expressionResolveValues;
        if (connectionInputData && runExecutionData && runIndex !== undefined) {
            expressionResolveValues = {
                connectionInputData,
                itemIndex: itemIndex || 0,
                node,
                runExecutionData,
                runIndex,
                workflow,
            };
        }
        const nodeCredentials = node.credentials
            ? node.credentials[type]
            : {};
        additionalData.executionContext = this.getExecutionContext();
        const decryptedDataObject = await additionalData.credentialsHelper.getDecrypted(additionalData, nodeCredentials, type, mode, executeData, false, expressionResolveValues);
        return decryptedDataObject;
    }
    get additionalKeys() {
        return (0, get_additional_keys_1.getAdditionalKeys)(this.additionalData, this.mode, this.runExecutionData);
    }
    getNodeParameter(parameterName, fallbackValue, options) {
        const itemIndex = 0;
        return this._getNodeParameter(parameterName, itemIndex, fallbackValue, options);
    }
    _getNodeParameter(parameterName, itemIndex, fallbackValue, options) {
        const { workflow, node, mode, runExecutionData, runIndex, connectionInputData, executeData } = this;
        const nodeType = workflow.nodeTypes.getByNameAndVersion(node.type, node.typeVersion);
        const value = (0, get_1.default)(node.parameters, parameterName, fallbackValue);
        if (value === undefined) {
            throw new n8n_workflow_1.ApplicationError('Could not get parameter', { extra: { parameterName } });
        }
        if (options?.rawExpressions) {
            return value;
        }
        const { additionalKeys } = this;
        let returnData;
        try {
            returnData = workflow.expression.getParameterValue(value, runExecutionData, runIndex, itemIndex, node.name, connectionInputData, mode, additionalKeys, executeData, false, {}, options?.contextNode?.name);
            (0, cleanup_parameter_data_1.cleanupParameterData)(returnData);
        }
        catch (e) {
            if (e instanceof n8n_workflow_1.ExpressionError &&
                node.continueOnFail &&
                node.type === 'n8n-nodes-base.set') {
                returnData = [{ name: undefined, value: undefined }];
            }
            else {
                if (e.context)
                    e.context.parameter = parameterName;
                e.cause = value;
                throw e;
            }
        }
        if (options?.extractValue) {
            returnData = (0, extract_value_1.extractValue)(returnData, parameterName, node, nodeType, itemIndex);
        }
        if (options?.ensureType) {
            returnData = (0, ensure_type_1.ensureType)(options.ensureType, returnData, parameterName, {
                itemIndex,
                runIndex,
                nodeCause: node.name,
            });
        }
        if (options?.skipValidation)
            return returnData;
        returnData = (0, validate_value_against_schema_1.validateValueAgainstSchema)(node, nodeType, returnData, parameterName, runIndex, itemIndex);
        return returnData;
    }
    evaluateExpression(expression, itemIndex = 0) {
        return this.workflow.expression.resolveSimpleParameterValue(`=${expression}`, {}, this.runExecutionData, this.runIndex, itemIndex, this.node.name, this.connectionInputData, this.mode, this.additionalKeys, this.executeData);
    }
    async prepareOutputData(outputData) {
        return [outputData];
    }
}
exports.NodeExecutionContext = NodeExecutionContext;
__decorate([
    decorators_1.Memoized,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NodeExecutionContext.prototype, "logger", null);
__decorate([
    decorators_1.Memoized,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NodeExecutionContext.prototype, "customData", null);
__decorate([
    decorators_1.Memoized,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NodeExecutionContext.prototype, "workflowSettings", null);
__decorate([
    decorators_1.Memoized,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NodeExecutionContext.prototype, "nodeType", null);
__decorate([
    decorators_1.Memoized,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NodeExecutionContext.prototype, "nodeFeatures", null);
__decorate([
    decorators_1.Memoized,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NodeExecutionContext.prototype, "nodeInputs", null);
__decorate([
    decorators_1.Memoized,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NodeExecutionContext.prototype, "nodeOutputs", null);
__decorate([
    decorators_1.Memoized,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NodeExecutionContext.prototype, "additionalKeys", null);
