// @n8n/workflow-sdk/dist/workflow-builder/control-flow-builders/split-in-batches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitInBatches = splitInBatches;
exports.isSplitInBatchesBuilder = isSplitInBatchesBuilder;
const uuid_1 = require("uuid");
const node_types_1 = require("../../constants/node-types");
const base_1 = require("../../types/base");
const node_builder_1 = require("../node-builders/node-builder");
class SplitInBatchesNodeInstance {
    constructor(input) {
        this.type = 'n8n-nodes-base.splitInBatches';
        const config = input.config ?? {};
        this.version = String(input.version);
        this.id = (0, uuid_1.v4)();
        this.name = config.name ?? 'Split In Batches';
        this.config = {
            ...config,
            parameters: config.parameters,
        };
    }
    update(config) {
        return new SplitInBatchesNodeInstance({
            version: this.version,
            config: {
                ...this.config,
                ...config,
            },
        });
    }
    input(_index) {
        throw new Error('SplitInBatches node input connections are managed by SplitInBatchesBuilder');
    }
    output(_index) {
        throw new Error('SplitInBatches node output connections are managed by SplitInBatchesBuilder');
    }
    to(_target, _outputIndex) {
        throw new Error('SplitInBatches node connections are managed by SplitInBatchesBuilder');
    }
    onError(_handler) {
        throw new Error('SplitInBatches node error handling is managed by SplitInBatchesBuilder');
    }
    getConnections() {
        return [];
    }
}
class SplitInBatchesBuilderImpl {
    constructor(input) {
        this._doneNodes = [];
        this._eachNodes = [];
        this._doneBatches = [];
        this._eachBatches = [];
        this._hasLoop = false;
        this.sibNode = new SplitInBatchesNodeInstance(input);
    }
    onEachBatch(target) {
        this._eachTarget = target;
        if (target !== null) {
            const nodes = extractNodesFromTarget(target);
            this._eachNodes = nodes;
            const firstNodes = getFirstNodes(target);
            if (firstNodes.length > 1) {
                this._eachBatches.push(firstNodes);
            }
            else if (firstNodes.length === 1) {
                this._eachBatches.push(firstNodes[0]);
            }
        }
        return this;
    }
    onDone(target) {
        this._doneTarget = target;
        if (target !== null) {
            const nodes = extractNodesFromTarget(target);
            this._doneNodes = nodes;
            const firstNodes = getFirstNodes(target);
            if (firstNodes.length > 1) {
                this._doneBatches.push(firstNodes);
            }
            else if (firstNodes.length === 1) {
                this._doneBatches.push(firstNodes[0]);
            }
        }
        return this;
    }
    getAllNodes() {
        return [this.sibNode, ...this._doneNodes, ...this._eachNodes];
    }
    getDoneNodes() {
        return this._doneNodes;
    }
    getEachNodes() {
        return this._eachNodes;
    }
    hasLoop() {
        return this._hasLoop;
    }
}
class SplitInBatchesBuilderWithExistingNode {
    constructor(existingNode) {
        this._doneNodes = [];
        this._eachNodes = [];
        this._doneBatches = [];
        this._eachBatches = [];
        this._hasLoop = false;
        this.sibNode = existingNode;
    }
    onEachBatch(target) {
        this._eachTarget = target;
        if (target !== null) {
            const nodes = extractNodesFromTarget(target);
            this._eachNodes = nodes;
            const firstNodes = getFirstNodes(target);
            if (firstNodes.length > 1) {
                this._eachBatches.push(firstNodes);
            }
            else if (firstNodes.length === 1) {
                this._eachBatches.push(firstNodes[0]);
            }
        }
        return this;
    }
    onDone(target) {
        this._doneTarget = target;
        if (target !== null) {
            const nodes = extractNodesFromTarget(target);
            this._doneNodes = nodes;
            const firstNodes = getFirstNodes(target);
            if (firstNodes.length > 1) {
                this._doneBatches.push(firstNodes);
            }
            else if (firstNodes.length === 1) {
                this._doneBatches.push(firstNodes[0]);
            }
        }
        return this;
    }
    getAllNodes() {
        return [this.sibNode, ...this._doneNodes, ...this._eachNodes];
    }
    getDoneNodes() {
        return this._doneNodes;
    }
    getEachNodes() {
        return this._eachNodes;
    }
    hasLoop() {
        return this._hasLoop;
    }
}
function splitInBatches(configOrNode, branches) {
    if (branches !== undefined &&
        isBranchesConfig(branches) &&
        (0, base_1.isNodeInstance)(configOrNode) &&
        (0, node_types_1.isSplitInBatchesType)(configOrNode.type)) {
        return new SplitInBatchesNamedSyntaxBuilder(configOrNode, branches);
    }
    if ((0, base_1.isNodeInstance)(configOrNode) && (0, node_types_1.isSplitInBatchesType)(configOrNode.type)) {
        return new SplitInBatchesBuilderWithExistingNode(configOrNode);
    }
    return new SplitInBatchesBuilderImpl(configOrNode);
}
function isSplitInBatchesBuilder(value) {
    return value instanceof SplitInBatchesBuilderImpl;
}
function extractNodesFromTarget(target) {
    if (target === null) {
        return [];
    }
    if (Array.isArray(target)) {
        return target.flatMap((t) => extractNodesFromTarget(t));
    }
    if ((0, base_1.isNodeChain)(target)) {
        return target.allNodes;
    }
    if ((0, node_builder_1.isIfElseBuilder)(target)) {
        const builder = target;
        const nodes = [builder.ifNode];
        nodes.push(...extractNodesFromTarget(builder.trueBranch));
        nodes.push(...extractNodesFromTarget(builder.falseBranch));
        return nodes;
    }
    if ((0, node_builder_1.isSwitchCaseBuilder)(target)) {
        const builder = target;
        const nodes = [builder.switchNode];
        for (const caseTarget of builder.caseMapping.values()) {
            nodes.push(...extractNodesFromTarget(caseTarget));
        }
        return nodes;
    }
    return [target];
}
function getFirstNodes(target) {
    if (target === null) {
        return [];
    }
    if (Array.isArray(target)) {
        return target.flatMap((t) => getFirstNodes(t));
    }
    if ((0, base_1.isNodeChain)(target)) {
        return [target.head];
    }
    if ((0, node_builder_1.isIfElseBuilder)(target)) {
        return [target.ifNode];
    }
    if ((0, node_builder_1.isSwitchCaseBuilder)(target)) {
        return [target.switchNode];
    }
    return [target];
}
class SplitInBatchesNamedSyntaxBuilder {
    constructor(sibNode, branches) {
        this._doneNodes = [];
        this._eachNodes = [];
        this._doneBatches = [];
        this._eachBatches = [];
        this._hasLoop = false;
        this.sibNode = sibNode;
        this._doneTarget = branches.done;
        this._eachTarget = branches.each;
        this._doneNodes = extractNodesFromTarget(branches.done);
        this._eachNodes = extractNodesFromTarget(branches.each);
        if (branches.done !== null) {
            const firstDoneNodes = getFirstNodes(branches.done);
            if (firstDoneNodes.length > 1) {
                this._doneBatches.push(firstDoneNodes);
            }
            else if (firstDoneNodes.length === 1) {
                this._doneBatches.push(firstDoneNodes[0]);
            }
        }
        if (branches.each !== null) {
            const firstEachNodes = getFirstNodes(branches.each);
            if (firstEachNodes.length > 1) {
                this._eachBatches.push(firstEachNodes);
            }
            else if (firstEachNodes.length === 1) {
                this._eachBatches.push(firstEachNodes[0]);
            }
        }
    }
    onEachBatch(_target) {
        throw new Error('Named object syntax does not support .onEachBatch() - branches are configured in the constructor');
    }
    onDone(_target) {
        throw new Error('Named object syntax does not support .onDone() - branches are configured in the constructor');
    }
    getAllNodes() {
        return [this.sibNode, ...this._doneNodes, ...this._eachNodes];
    }
    getDoneNodes() {
        return this._doneNodes;
    }
    getEachNodes() {
        return this._eachNodes;
    }
    hasLoop() {
        return this._hasLoop;
    }
}
function isBranchesConfig(arg) {
    return (arg !== null &&
        typeof arg === 'object' &&
        'done' in arg &&
        'each' in arg &&
        !('type' in arg) &&
        !('version' in arg));
}
