// n8n-core/dist/instance-settings/instance-settings.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstanceSettings = void 0;
const backend_common_1 = require("@n8n/backend-common");
const config_1 = require("@n8n/config");
const decorators_1 = require("@n8n/decorators");
const di_1 = require("@n8n/di");
const crypto_1 = require("crypto");
const n8n_workflow_1 = require("n8n-workflow");
const nanoid_1 = require("nanoid");
const node_fs_1 = require("node:fs");
const node_os_1 = __importDefault(require("node:os"));
const path_1 = __importDefault(require("path"));
const worker_missing_encryption_key_error_1 = require("./worker-missing-encryption-key.error");
const nanoid = (0, nanoid_1.customAlphabet)(n8n_workflow_1.ALPHABET, 16);
let InstanceSettings = class InstanceSettings {
    constructor(config, logger) {
        this.config = config;
        this.logger = logger;
        this.n8nFolder = this.config.n8nFolder;
        this.staticCacheDir = path_1.default.join(this.config.userHome, '.cache/n8n/public');
        this.customExtensionDir = path_1.default.join(this.n8nFolder, 'custom');
        this.nodesDownloadDir = path_1.default.join(this.n8nFolder, 'nodes');
        this.nodeDefinitionsDir = path_1.default.join(this.n8nFolder, 'node-definitions');
        this.settingsFile = path_1.default.join(this.n8nFolder, 'config');
        this.enforceSettingsFilePermissions = this.loadEnforceSettingsFilePermissionsFlag();
        this.instanceRole = 'unset';
        this.isMultiMainEnabled = false;
        this.isMultiMainLicensed = false;
        const command = process.argv[2];
        this.instanceType = ['webhook', 'worker'].includes(command) ? command : 'main';
        this.hostId = `${this.instanceType}-${this.isDocker ? node_os_1.default.hostname() : nanoid()}`;
        this.settings = this.loadOrCreate();
        this.instanceId = this.generateInstanceId();
        this.hmacSignatureSecret = this.getOrGenerateHmacSignatureSecret();
    }
    setMultiMainEnabled(newState) {
        this.isMultiMainEnabled = newState;
    }
    setMultiMainLicensed(newState) {
        this.isMultiMainLicensed = newState;
    }
    get isMultiMain() {
        return this.instanceType === 'main' && this.isMultiMainEnabled && this.isMultiMainLicensed;
    }
    get isSingleMain() {
        return !this.isMultiMain;
    }
    get isWorker() {
        return this.instanceType === 'worker';
    }
    get isLeader() {
        return this.instanceRole === 'leader';
    }
    markAsLeader() {
        this.instanceRole = 'leader';
    }
    get isFollower() {
        return this.instanceRole === 'follower';
    }
    markAsFollower() {
        this.instanceRole = 'follower';
    }
    get encryptionKey() {
        return this.settings.encryptionKey;
    }
    get tunnelSubdomain() {
        return this.settings.tunnelSubdomain;
    }
    get fsStorageMigrated() {
        return this.settings.fsStorageMigrated === true;
    }
    markFsStorageMigrated() {
        this.update({ fsStorageMigrated: true });
    }
    get isDocker() {
        if ((0, node_fs_1.existsSync)('/.dockerenv') || (0, node_fs_1.existsSync)('/run/.containerenv'))
            return true;
        try {
            const cgroupV1 = (0, node_fs_1.readFileSync)('/proc/self/cgroup', 'utf8');
            if (cgroupV1.includes('docker') ||
                cgroupV1.includes('kubepods') ||
                cgroupV1.includes('containerd'))
                return true;
        }
        catch { }
        try {
            const cgroupV2 = (0, node_fs_1.readFileSync)('/proc/self/mountinfo', 'utf8');
            if (cgroupV2.includes('docker') ||
                cgroupV2.includes('kubelet') ||
                cgroupV2.includes('containerd'))
                return true;
        }
        catch { }
        return false;
    }
    update(newSettings) {
        this.save({ ...this.settings, ...newSettings });
    }
    loadOrCreate() {
        const encryptionKeyFromEnv = this.config.encryptionKey || undefined;
        if ((0, node_fs_1.existsSync)(this.settingsFile)) {
            const content = (0, node_fs_1.readFileSync)(this.settingsFile, 'utf8');
            this.ensureSettingsFilePermissions();
            const settings = (0, n8n_workflow_1.jsonParse)(content, {
                errorMessage: `Error parsing n8n-config file "${this.settingsFile}". It does not seem to be valid JSON.`,
            });
            if (!backend_common_1.inTest)
                this.logger.debug(`User settings loaded from: ${this.settingsFile}`);
            const { encryptionKey, tunnelSubdomain, fsStorageMigrated } = settings;
            if (encryptionKeyFromEnv && encryptionKey !== encryptionKeyFromEnv) {
                throw new n8n_workflow_1.ApplicationError(`Mismatching encryption keys. The encryption key in the settings file ${this.settingsFile} does not match the N8N_ENCRYPTION_KEY env var. Please make sure both keys match. More information: https://docs.n8n.io/hosting/environment-variables/configuration-methods/#encryption-key`);
            }
            return { encryptionKey, tunnelSubdomain, fsStorageMigrated };
        }
        if (!encryptionKeyFromEnv) {
            if (this.instanceType === 'worker') {
                throw new worker_missing_encryption_key_error_1.WorkerMissingEncryptionKey();
            }
            if (!backend_common_1.inTest) {
                this.logger.info(`No encryption key found - Auto-generating and saving to: ${this.settingsFile}`);
            }
        }
        (0, node_fs_1.mkdirSync)(this.n8nFolder, { recursive: true });
        const encryptionKey = encryptionKeyFromEnv ?? (0, crypto_1.randomBytes)(24).toString('base64');
        const settings = { encryptionKey };
        this.save(settings);
        this.ensureSettingsFilePermissions();
        return settings;
    }
    generateInstanceId() {
        const { encryptionKey } = this;
        return (0, crypto_1.createHash)('sha256')
            .update(encryptionKey.slice(Math.round(encryptionKey.length / 2)))
            .digest('hex');
    }
    getOrGenerateHmacSignatureSecret() {
        const hmacSignatureSecretFromEnv = process.env.N8N_HMAC_SIGNATURE_SECRET;
        if (hmacSignatureSecretFromEnv)
            return hmacSignatureSecretFromEnv;
        const { encryptionKey } = this;
        return (0, crypto_1.createHash)('sha256').update(`hmac-signature:${encryptionKey}`).digest('hex');
    }
    save(settings) {
        this.settings = settings;
        (0, node_fs_1.writeFileSync)(this.settingsFile, JSON.stringify(this.settings, null, '\t'), {
            mode: this.enforceSettingsFilePermissions.enforce ? 0o600 : undefined,
            encoding: 'utf-8',
        });
    }
    loadEnforceSettingsFilePermissionsFlag() {
        const { enforceSettingsFilePermissions } = this.config;
        const isEnvVarSet = !!process.env.N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS;
        if (this.isWindows()) {
            if (isEnvVarSet) {
                console.warn('Ignoring N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS as it is not supported on Windows.');
            }
            return {
                isSet: isEnvVarSet,
                enforce: false,
            };
        }
        return {
            isSet: isEnvVarSet,
            enforce: enforceSettingsFilePermissions,
        };
    }
    ensureSettingsFilePermissions() {
        if (!this.enforceSettingsFilePermissions.enforce)
            return;
        if (this.isWindows())
            return;
        const permissionsResult = (0, n8n_workflow_1.toResult)(() => {
            const stats = (0, node_fs_1.statSync)(this.settingsFile);
            return stats?.mode & 0o777;
        });
        if (!permissionsResult.ok) {
            this.logger.warn(`Could not ensure settings file permissions: ${permissionsResult.error.message}. To skip this check, set N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=false.`);
            return;
        }
        const arePermissionsCorrect = permissionsResult.result === 0o600;
        if (arePermissionsCorrect)
            return;
        this.logger.error(`Permissions 0${permissionsResult.result.toString(8)} for n8n settings file ${this.settingsFile} are too wide. Changing permissions to 0600..`);
        const chmodResult = (0, n8n_workflow_1.toResult)(() => (0, node_fs_1.chmodSync)(this.settingsFile, 0o600));
        if (!chmodResult.ok) {
            this.logger.warn(`Could not enforce settings file permissions: ${chmodResult.error.message}. To skip this check, set N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=false.`);
        }
    }
    isWindows() {
        return process.platform === 'win32';
    }
};
exports.InstanceSettings = InstanceSettings;
__decorate([
    decorators_1.Memoized,
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], InstanceSettings.prototype, "isDocker", null);
exports.InstanceSettings = InstanceSettings = __decorate([
    (0, di_1.Service)(),
    __metadata("design:paramtypes", [config_1.InstanceSettingsConfig,
        backend_common_1.Logger])
], InstanceSettings);
