// n8n-workflow/dist/cjs/run-execution-data/run-execution-data.v1.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.runExecutionDataV0ToV1 = runExecutionDataV0ToV1;
    function runExecutionDataV0ToV1(data) {
        const destinationNodeV0 = data.startData?.destinationNode;
        const originalDestinationNodeV0 = data.startData?.originalDestinationNode;
        return {
            ...data,
            version: 1,
            startData: {
                ...data.startData,
                destinationNode: destinationNodeV0
                    ? {
                        nodeName: destinationNodeV0,
                        mode: 'inclusive',
                    }
                    : undefined,
                originalDestinationNode: originalDestinationNodeV0
                    ? {
                        nodeName: originalDestinationNodeV0,
                        mode: 'inclusive',
                    }
                    : undefined,
            },
        };
    }
});
